{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "This notebook is a companion to the paper published in Nature by Tao Ding & Pat Schloss entitled [\"Dynamics and associations of microbial community types across the human body\"](http://dx.doi.org/10.1038/nature13178). Dr. Ding was a postdoc in the lab of Dr Schloss within the Department of Microbiology & Immunology at the University of Michigan. If any questions should arise, please feel free to contact Dr. Schloss at pschloss@umich.edu. This notebook was originally developed using mothur ~v.1.30. Since then we have tweaked the running of trim.flows which may have a small effect on the number of samples with 1000 reads. Since then all commands have been re-validated using mothur v.1.33.3. You can find all of the raw data and derivative files generated below (except for the dbGaP data) in the [dingschlossnature bucket](http://dingschlossnature.s3-website-us-west-2.amazonaws.com) on the AWS S3 server\n",
    "\n",
    "\n",
    "## The HMP Clinical Sequencing Project\n",
    "\n",
    "The underlying data were developed as part of the Human Microbiome Project to understand the structure of the microbiome in healthy individuals [Peterson et al. 2009](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2792171/). The folks at PLoS have been very helpful in collecting the papers published as part of the overall HMP clinical study as well as the other projects associated with HMP funding as the [The Human Microbiome Project Collection](http://www.ploscollections.org/article/browseIssue.action?issue=info:doi/10.1371/issue.pcol.v01.i13) (nb: there are other papers from the HMP that don't appear in this collection). As mentioned in our paper, the individuals in the main clinical study were **very** healthy. A talmudic question for you to consider is whether heatlhy == normal. We contend that if 80% of American's have a cavity in their mouth, they are *normal*, but not healthy. In this study, people with cavities were excluded from the entire study. You can learn more about the exclusion criteria at the [dbGaP project site](http://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000228.v3.p1) and [Aagaard et al. ,2013](http://www.ncbi.nlm.nih.gov/pubmed/23165986).\n",
    "\n",
    "Overall, the study sampled 15 bodysites from 150 women and 18 bodysites from 150 men. Of the 300 subjects, 150 were enrolled in Houston, TX and 150 were enrolled in St. Louis, MO. The goal was to sample all subjects twice and 100 subjects a third time. All samples were to be characterized by 16S rRNA gene sequencing and a decent number were sequenced via metagenomics shotgun sequencing. The current study only concerned itself with the 16S rRNA gene sequencing. There were several stages to the study. The first consisted of a \"Pilot Production Study\" where a set of 24 subjects (12 men and 12 women) were recruited in Houston. The samples were processed there and then split with one replicate sent to one sequencing center and another replicate sent to another sequencing center. There were four sequencing centers in total: The Baylor College of Medicine (BCM), The Washington University Genome Sequencing Center (WUGSC), The Broad Institute (BI), and the J Craig Venter Institute (JCVI). These samples have already been described in earlier studies by us and others ([Schloss et al. 2011](http://www.ploscollections.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0027310), [JCHMPDGWG 2012](http://www.ploscollections.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0039315)).\n",
    "\n",
    "Next, the project scaled up to the full 300 subject pool. As we discuss in the Supplementary Information for the paper, the city of origin for each subject was a complicating varaible. The samples from the 150 subjects from St. Louis were extracted and sequenced at WUGSC. The samples from the 150 subjects in Houston were sent to BCM, JCVI, and BI for extraction and sequencing. Thus the city of origin was perfectly confounded with the location of sequencing. As was discussed elsewhere, the largest effect observed in the data was from city of origin ([HMPC 2012](http://www.nature.com/nature/journal/v486/n7402/full/nature11234.html). Although we would love to believe that people vary in their microbiome between Houston and St. Louis and can concoct many reasons why this would be so, the easiest is that there were small differences in how samples were obtained, handled, and processed that affected the microbiome structure. Regardless, prior to our paper, the only data that had been published was from the so-called \"May 1, 2010\" data freeze. The samples from this data freeze primarily only contained the first sampling point for each subject. Although the second sample was available for some of these subjects by that freeze, the papers published to date (e.g. [HMPC 2012](http://www.nature.com/nature/journal/v486/n7402/full/nature11234.html), [HMPC 2012](http://www.nature.com/nature/journal/v486/n7402/full/nature11209.html), and [others](http://www.ploscollections.org/article/browseIssue.action?issue=info:doi/10.1371/issue.pcol.v01.i13)). Thus, the goal of our paper was to present an analysis of all of the data generated by the HMP clinical study.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Data curation\n",
    "## The data\n",
    "**Sequence data.** The 16S rRNA gene sequencing data was generated by sequencing the V35 region of gene on the 454 Titanium platform. Initially the project sequenced the V13 and V69 regions; however, these were eventually dropped from the analysis plan. Thus, we focused our effort on the V35 region. All of the 16S rRNA gene sequence data are available at the NCBI Short Read Archive (SRA): the Clinical Pilot Project (accession [SRP002012](http://www.ncbi.nlm.nih.gov/Traces/sra/?study=SRP002012)), Phase I (the May 1 freeze; accession [SRP002395](http://www.ncbi.nlm.nih.gov/Traces/sra/?study=SRP002395)), and Phase II (accession [SRP002860](http://www.ncbi.nlm.nih.gov/Traces/sra/?study=SRP002860)). Because we were part of the original Data Analysis Working Group (DAWG), we originally obtained some of the data from Jonathan Crabtree at the [HMP Data Analsyis Coordination Center (DACC)](http://www.hmpdacc.org). Thus, there may be slight variations in the data that we used and the data available throgh the SRA.\n",
    "\n",
    "The 16S rRNA gene sequence analysis pipeline used in this study is based on the results of our previous work analyzing the mock community data generated as part of the PPS ([Schloss et al. 2011](http://www.ploscollections.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0027310)). We used the PyroNoise-based approach with *de novo* chimera checking using the UCHIME algorithm ([Quince et al. 2011](http://www.biomedcentral.com/1471-2105/12/38) and [Edgar et al. 2011](http://bioinformatics.oxfordjournals.org/content/27/16/2194.long)).  All analysis steps were performed within mothur ([Schloss et al. 2009](http://aem.asm.org/content/75/23/7537.long)). As we described there, this curation pipeline yields an error rate of ~0.02%. The mothur package can be obtained from the [mothur website](http://www.mothur.org) or our [GitHub repository](https://github.com/mothur/mothur). The wiki also has a shorter description of our [SOP for processing 454 sequencing data](http://www.mothur.org/wiki/454_SOP).\n",
    "\n",
    "**Clinical data.** The clinical data for this project is available through the [dbGaP project site](http://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000228.v3.p1). Unfortunately, because this is considered PHI, I am not at liberty to distribute these data. You may obtain the data from dbGaP by going through the appropriate channels. *Don't ask us to send the files to you!*\n",
    "\n",
    "\n",
    "## Sequence curation\n",
    "### Preliminaries\n",
    "To organize the files we created several files in our directory:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir scripts oligos_files references sff_files sffinfo trim.flows shhh.flows trim.seqs v13 v35 v69 dbGaP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can obtain the oligos_file files from the GitHub repository for this project. If you are trying to run this, you will need to go in to the file and decompress the files (see oligos_files.tgz):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd oligos_files\n",
    "tar xvzf oligos_files.tgz\n",
    "cd ../"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These oligos files will tell mothur the barcode sequence that corresponds to each sample so that we can map reads to samples. If you open one of these files you'll see something that looks like this:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "forward\tATTACCGCGGCTGCTGG\tv13\n",
    "forward\tCCGTCAATTCMTTTRAGT\tv35\n",
    "forward\tTACGGYTACCTTGTTAYGACTT\tv69\n",
    "barcode\tACGAGAAC\tphase1.SRR040576.BI.700034678"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first three lines are the \"forward\" PCR primer that were adjacent to the sequencing primer. By putting all three possible primers in the file we let mothur figure out which region the seuqences belong to. The barcode line gives the barcode sequence and then the sample name. The \"phase 1\" indicates that the sample was from before the May 1 data freeze. The \"SRR040576\" indicates the sequencing run that the sample came from. The \"BI\" indicates that the sample was sequenced at the Broad Institute\". Finally, the \"700034678\" corresponds to the anonymized sample id. Together this string represents a unique identifier for every sequencing collection (some samples were sequenced twice, some at two centers, and some in multiple phases of the study).\n",
    "\n",
    "You will also want to download the sequence data files from the SRA and convert them into sff format using the SRA's tools. Put the resulting sff files into the sff_files folder and return to the parent directory. Because we obtained several of the PPS datasets before they were publically available, the names of our files are slightly different from those on the SRA. At this point you'll need to double check that you have an oligos file for every sff file. Runing the following two commands should give the same number. The third command will tell you what you're missing. If you're using the same files that we used, there will be 3 files extra: F5672XE02, F5672XE02.v13, and F5672XE02.v35. There was somethign screwy about how these were sequenced and so later we will replace the F5672XE02 with the v13 and v35 versions and all will be well with the world."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "ls sff_files | wc -l\n",
    "ls oligos_files | wc -l\n",
    "ls oligos_files/* sff_files/* | cut -f 2 -d \"/\" | cut -f 1 -d \".\" | sort | uniq -u"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we want to get the reference files that we'll be using to process the data. We chose to use the SILVA-based reference alignment ([Pruesse et al. 2007](http://nar.oxfordjournals.org/lookup/pmid?view=long&pmid=17947321)) because of its superiority to the greengenes-based alignment. Previous work from our lab has demonstrated this ([Schloss 2010](http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1000844)). In addition, we used a modified version of the RDP's training set ([Wang et al. 2007](http://aem.asm.org/content/73/16/5261.long); v.9). Our modifications included adding several sequences from the Archaea, Eukarya, chloroplasts, and mitochondria."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd references\n",
    "wget http://www.mothur.org/w/images/5/59/Trainset9_032012.pds.zip\n",
    "unzip Trainset9_032012.pds.zip\n",
    "\n",
    "wget http://www.mothur.org/w/images/9/98/Silva.bacteria.zip\n",
    "unzip Silva.bacteria.zip\n",
    "mv silva.bacteria/silva.bacteria.fasta ./\n",
    "\n",
    "cd ../"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we want to extract the flowgram data from the sff files. Let's move to the sff_info folder. Then we'll want to run sffinfo from mothur on each of the sff files. When we actually ran this we submitted these jobs in parallel using pbs rather than in series and we are limited to about 100 jobs per user. You can find the actual sff files that we processed in sff_files.txt. We had 12943 sff files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd sffinfo\n",
    "\n",
    "> sffinfo.batch\n",
    "\n",
    "for SFF in $(ls ../sff_files/*sff)\n",
    "do\n",
    "    #make a big file with all of the sffinfo commands in the file\n",
    "    echo \"mothur \\\"#sffinfo(sff=$SFF, outputdir=./)\\\"\" >> sffinfo.batch    \n",
    "done\n",
    "\n",
    "#split sffinfo.batch in to files with 130 lines each and named with sffinfo_??\n",
    "split -l 130 sffinfo.batch sffinfo_\n",
    "\n",
    "cat > header.batch << \"_EOF_\"\n",
    "#!/bin/sh\n",
    "#PBS -l nodes=1:ppn=1\n",
    "#PBS -l walltime=500:00:00\n",
    "#PBS -l mem=1gb\n",
    "#PBS -j oe\n",
    "#PBS -m abe\n",
    "#PBS -V\n",
    "#PBS -M your@email.here\n",
    "#PBS -q first\n",
    "\n",
    "echo \"ncpus-2.pbs\"\n",
    "cat $PBS_NODEFILE\n",
    "qstat -f $PBS_JOBID\n",
    "\n",
    "cd $PBS_O_WORKDIR\n",
    "\n",
    "NCPUS=`wc -l $PBS_NODEFILE | awk '{print $1}'`\n",
    "\n",
    "_EOF_\n",
    "\n",
    "\n",
    "cat > tail.batch << \"_EOF_\"\n",
    "\n",
    "echo \"qsub working directory absolute is\"\n",
    "echo $PBS_O_WORKDIR\n",
    "exit\n",
    "_EOF_\n",
    "\n",
    "for BATCH in $(ls sffinfo_*)\n",
    "do\n",
    "    cat header.batch $BATCH tail.batch > $BATCH.batch\n",
    "    qsub $BATCH.batch\n",
    "done\n",
    "\n",
    "#there were those screwy files above that needed to be copied for v13 and v35 and removed\n",
    "#be sure to run these only after the scripts are done\n",
    "cp F5672XE02.flow F5672XE02.v13.flow\n",
    "cp F5672XE02.flow F5672XE02.v35.flow\n",
    "rm F5672XE02.flow\n",
    "\n",
    "#check to make sure we have the right number of files. If not, some baby sitting may be needed\n",
    "ls ../sff_files/*sff | wc -l    #should equal 12943\n",
    "ls *flow | wc -l                #should equal 12944"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As mentioned above there was a PPS stage of the project and then the actual production stages. Accidentally, some of the data from WUGSC was uploaded in the production stage even though it had already been uploaded in the PPS stage. To keep track of what was from the PPS stage and what was from the production stage, we ran the following Perl script (modPPSFlowFiles.pl) on the flow gram data from the PPS stage go ahead and save a copy to the scripts folder:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#!perl\n",
    "use strict;\n",
    "use warnings;\n",
    "\n",
    "foreach my $file (@ARGV){\n",
    "\t\n",
    "\topen(INPUT, $file) or die;\n",
    "\t\n",
    "\t$file =~ /(.*).flow/;\n",
    "\tmy $outfile = \"$1.mod.flow\";\n",
    "\t\n",
    "\topen(OUTPUT, \">$outfile\") or die;\n",
    "\t\n",
    "\tmy $header = <INPUT>;\n",
    "\tprint OUTPUT $header;\n",
    "\t\n",
    "\twhile(<INPUT>){\n",
    "\t\t$_ =~ s/^(\\S*)/$1p/;\n",
    "\t\tprint OUTPUT $_;\n",
    "\t}\n",
    "\tclose(INPUT);\n",
    "\tclose(OUTPUT);\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This generates new files ending in \"mod.flow\" where the sequence name has a \"p\" appended to the end. The SRA files all start with \"SRR\" and since the PPS files weren't in the SRA, we'll use that to our advantage to pick those files out..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "perl ../scripts/modPPSFlowFiles.pl `ls *flow | grep -v \"^SRR\"`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we want to move over to the trim.flows folder and run trim.flows. As described in [Schloss et al. 2009](http://aem.asm.org/content/75/23/7537.long) we will allow one mismatch to the barcode sequence, two mismatches to the primer sequence, no ambiguous base calls, and a maximum homopolymer length of 8 nt. Also, by default, mothur will trim all reads to 450 flows and cull any reads with fewer than 450 flows. This results in reads that are about 270 bp. Again we parallelize this and fire everything off in batches of 130 flow files at a time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ../trim.flows\n",
    "\n",
    "> trimflows.batch\n",
    "\n",
    "for FLOW in $(ls ../sffinfo/SRR*flow)\n",
    "do\n",
    "    OLIGOS=${FLOW//sffinfo/oligos_files}\n",
    "    OLIGOS=${OLIGOS//flow/oligos}\n",
    "\n",
    "    echo \"mothur \\\"#trim.flows(flow=$FLOW, oligos=$OLIGOS, pdiffs=2, bdiffs=1, maxhomop=8, outputdir=./)\\\"\" >> trimflows.batch\n",
    "done\n",
    "\n",
    "for FLOW in $(ls ../1_sffinfo/*.mod.flow)\n",
    "do\n",
    "    OLIGOS=${FLOW//1_sffinfo/oligos_files}\n",
    "    OLIGOS=${OLIGOS//mod.flow/oligos}\n",
    "\n",
    "    echo \"mothur \\\"#trim.flows(flow=$FLOW, oligos=$OLIGOS, pdiffs=2, bdiffs=1, maxhomop=8, processors=8, outputdir=./)\\\"\" >> trimflows.batch\n",
    "done\n",
    "\n",
    "\n",
    "split -l 130 trimflows.batch trimflows_   #split trimflows.batch in to files with 130 lines each and named with trimflows_??\n",
    "\n",
    "cat > header.batch << \"_EOF_\"\n",
    "#!/bin/sh\n",
    "#PBS -l nodes=1:ppn=1\n",
    "#PBS -l walltime=500:00:00\n",
    "#PBS -l mem=1gb\n",
    "#PBS -j oe\n",
    "#PBS -m abe\n",
    "#PBS -V\n",
    "#PBS -M your@email.here\n",
    "#PBS -q first\n",
    "\n",
    "echo \"ncpus-2.pbs\"\n",
    "cat $PBS_NODEFILE\n",
    "qstat -f $PBS_JOBID\n",
    "\n",
    "cd $PBS_O_WORKDIR\n",
    "\n",
    "NCPUS=`wc -l $PBS_NODEFILE | awk '{print $1}'`\n",
    "\n",
    "_EOF_\n",
    "\n",
    "\n",
    "cat > tail.batch << \"_EOF_\"\n",
    "\n",
    "echo \"qsub working directory absolute is\"\n",
    "echo $PBS_O_WORKDIR\n",
    "exit\n",
    "_EOF_\n",
    "\n",
    "for BATCH in $(ls trimflows_*)\n",
    "do\n",
    "    cat header.batch $BATCH tail.batch > $BATCH.batch\n",
    "    qsub $BATCH.batch\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This generates files ending in \\*v13.flow, \\*v35.flow, and \\*v69.flow as well as a file ending in \\*.flow.files that corresponds to each sff file that we started with. After running trim.flows, we had 12944 \\*.flow.files files, which represented 42,768 samples. Of these \\*.flow.files files 176 had no samples in them indicating that trim.flows was unable to find any sequences with the right barcode or primer. Manual inspection of the original flow files showed that these were samples with very few sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "wc -l *files > flow.files.counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "a <-read.table(file=\"flow.files.counts\"); colnames(a) <- c(\"nsamples\", \"sample_id\")\n",
    "sum(a$nsamples)                     #total number of samples remaining\n",
    "sum(a$nsamples==0)                  #number of flow.files files without any samples\n",
    "nrow(a) - sum(a$nsamples==0)        #number of flow.files with samples in them\n",
    "b <- a[a$nsamples!=0,]\n",
    "write.table(b$sample_id, \"good.flow.files\", row.names=F, col.name=F, quote=F)    #output the names of the good files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We pressed on with the 12,768 \\*.flow.files files that had samples in them. At this point we could have separated the three 16S rRNA gene sequence regions, but we just plugged ahead with everything together. The next step was to denoise the sequences using the PyroNoise algorithm, which is implemented in mothur as shhh.flows. Investigators interested in repeating our analysis should note that this step involves a random number generator to break ties during the initial clustering step - the effects should be minor. This step may need a bit of baby siting depending on how much RAM you allocate and how much is needed for each job. Some may crap out and you'll have to figure out whether anything crapped out. We'll show how this was done at the end of the step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ../shhh.flows\n",
    "\n",
    "> shhhflows.batch\n",
    "\n",
    "for FLOWFILE in $(ls ../trim.flows/*flow.files)\n",
    "do\n",
    "    echo \"mothur \\\"#shhh.flows(file=$FLOWFILE, outputdir=./)\\\"\" >> shhhflows.batch\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "split -l 20 shhhflows.batch shhhflows_   #split trimflows.batch in to files with 20 lines each and named with shhhflows_??\n",
    "\n",
    "cat > header.batch << \"_EOF_\"\n",
    "#!/bin/sh\n",
    "#PBS -l nodes=1:ppn=1\n",
    "#PBS -l walltime=500:00:00\n",
    "#PBS -l mem=8gb\n",
    "#PBS -j oe\n",
    "#PBS -m abe\n",
    "#PBS -V\n",
    "#PBS -M your@email.here\n",
    "#PBS -q first\n",
    "\n",
    "echo \"ncpus-2.pbs\"\n",
    "cat $PBS_NODEFILE\n",
    "qstat -f $PBS_JOBID\n",
    "\n",
    "cd $PBS_O_WORKDIR\n",
    "\n",
    "NCPUS=`wc -l $PBS_NODEFILE | awk '{print $1}'`\n",
    "\n",
    "_EOF_\n",
    "\n",
    "\n",
    "cat > tail.batch << \"_EOF_\"\n",
    "\n",
    "echo \"qsub working directory absolute is\"\n",
    "echo $PBS_O_WORKDIR\n",
    "exit\n",
    "_EOF_\n",
    "\n",
    "for BATCH in $(ls shhhflows_*)\n",
    "do\n",
    "    cat header.batch $BATCH tail.batch > $BATCH.batch\n",
    "    qsub $BATCH.batch\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We found that this step required some handholding as certain files needed more RAM devoted to it and others would unexpectedly fail. To make sure everything went right we used a few tricks to make sure we had everything"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#how many shhh.fasta files were created (ignoring the indivdiual ones for each sample)?\n",
    "ls *shhh.fasta | grep -cv \"v\"\n",
    "#12766 - three short of what we found above...\n",
    "\n",
    "ls *shhh.fasta | grep -v \"v\" > shhh.files\n",
    "cut -c 9- ../trim.flows/flow.files.counts > flow.files\n",
    "cat shhh.files ../trim.flows/good.flow.files  | rev | cut -f 3- -d \".\" | rev | sort | uniq -u\n",
    "\n",
    "#F5672XE02.v13.mod\n",
    "#F5672XE02.v35.mod\n",
    "#total"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So this makes sense - the two files that end in mod were inadvertantly left out in the line with \"grep -v\" and we had a total after running the wc command before. So we're all here adn ready to move on! Running shhh.flows results in a fasta file and a name file. The name file indicates the names of the sequences that were duplicates and were original sequence after denoising. We'll have to include this in all of the commands going forward so that we can have a proper accounting of the sequences.\n",
    "\n",
    "Now we were ready to remove the barcodes and primers using trim.seqs. This step also generates a \"groups\" file, that indicate which sample each sequence belongs to. We will use many of the same parameters as we used in running trim.flows and we will use allfiles=T and flip=T. The former is so that we get a separate fasta, name, and group file for each sample in the original sff file and the latter is to return the reverse complement of the sequences since they were sequenced from the V5 region towards the V3 region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ../trim.seqs\n",
    "\n",
    "> trimseqs.batch\n",
    "\n",
    "for FASTA in $(ls ../shhh.flows/SRR??????.shhh.fasta)\n",
    "do\n",
    "    OLIGOS=${FASTA//shhh.flows/oligos_files}\n",
    "    OLIGOS=${OLIGOS//shhh.fasta/oligos}\n",
    "\n",
    "    NAMES=${FASTA//fasta/names}\n",
    "    echo \"mothur \\\"#trim.seqs(fasta=$FASTA, name=$NAMES, oligos=$OLIGOS, pdiffs=2, bdiffs=1, maxhomop=8, maxambig=0, allfiles=T, flip=T, outputdir=./)\\\"\" >> trimseqs.batch\n",
    "done\n",
    "\n",
    "for FASTA in $(ls ../shhh.flows/*mod.shhh.fasta)\n",
    "do\n",
    "    OLIGOS=${FASTA//shhh.flows/oligos_files}\n",
    "    OLIGOS=${OLIGOS//mod.shhh.fasta/oligos}\n",
    "\n",
    "    NAMES=${FASTA//fasta/names}\n",
    "    echo \"mothur \\\"#trim.seqs(fasta=$FASTA, name=$NAMES, oligos=$OLIGOS, pdiffs=2, bdiffs=1, maxhomop=8, maxambig=0, allfiles=T, flip=T, outputdir=./)\\\"\" >> trimseqs.batch\n",
    "done\n",
    "\n",
    "\n",
    "split -l 20 trimseqs.batch trimseqs_   #split trimseqs.batch in to files with 20 lines each and named with trimseqs_??\n",
    "\n",
    "cat > header.batch << \"_EOF_\"\n",
    "#!/bin/sh\n",
    "#PBS -l nodes=1:ppn=1\n",
    "#PBS -l walltime=500:00:00\n",
    "#PBS -l mem=1gb\n",
    "#PBS -j oe\n",
    "#PBS -m abe\n",
    "#PBS -V\n",
    "#PBS -M your@email.here\n",
    "#PBS -q first\n",
    "\n",
    "echo \"ncpus-2.pbs\"\n",
    "cat $PBS_NODEFILE\n",
    "qstat -f $PBS_JOBID\n",
    "\n",
    "cd $PBS_O_WORKDIR\n",
    "\n",
    "NCPUS=`wc -l $PBS_NODEFILE | awk '{print $1}'`\n",
    "\n",
    "_EOF_\n",
    "\n",
    "\n",
    "cat > tail.batch << \"_EOF_\"\n",
    "\n",
    "echo \"qsub working directory absolute is\"\n",
    "echo $PBS_O_WORKDIR\n",
    "exit\n",
    "_EOF_\n",
    "\n",
    "for BATCH in $(ls trimseqs_*)\n",
    "do\n",
    "    cat header.batch $BATCH tail.batch > $BATCH.batch\n",
    "    qsub $BATCH.batch\n",
    "done\n",
    "\n",
    "ls ls ../trim.flows/*v??.flow | cut -f 3 -d \"/\" > flow.files; wc -l flow.files   #21383\n",
    "ls *.v??.fasta > fasta.files; wc -l fasta.files                                  #21383\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Running trim.seqs returns a fasta, name, and group file for each sample. Because of how we named our barcodes in the original oligos files we now have files named like: SRR350309.shhh.phase2.SRR350309.WUGSC.700172315.v35.fasta in our directory. Everything in this file name from phase2 through v35 comes from the oligos file. At this point we will separate our data into separate folders according to the region of the gene that the data were derived from."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%batch\n",
    "\n",
    "cd ..\n",
    "\n",
    "for REGION in v13 v35 v69\n",
    "do\n",
    "    cd $REGION\n",
    "    cat ../trim.seqs/*$REGION.fasta > $REGION.fasta\n",
    "    cat ../trim.seqs/*$REGION.names > $REGION.names\n",
    "    cat ../trim.seqs/*$REGION.groups > $REGION.groups\n",
    "    cd ../\n",
    "done\n",
    "\n",
    "cut -f 2 v13/v13.groups | sort | uniq | wc -l   #7194\n",
    "cut -f 2 v35/v35.groups | sort | uniq | wc -l   #14015\n",
    "cut -f 2 v69/v69.groups | sort | uniq | wc -l   #171"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As mentioned above and indicated in running the last three commands, the v35 dataset has the most samples and that is what we based our analysis on for the paper. We'll go ahead and move into the v35 folder, but before proceeding, let's customize the reference files. To improve the analysis and reduce file sizes, we want to limit the reference files to the v35 region that was sequenced. As we are using the SILVA reference alignment ([Pruesse et al. 2007](http://nar.oxfordjournals.org/lookup/pmid?view=long&pmid=17947321)) we figured out that the primers that were used to amplify the DNA meant that a full length amplicon would start at position 6,426 in the alignment space and end at position 27,654 (the full alignment is 50,000 characters long). So we want to filter the silva.bacteria.fasta file based on those coordinates and then align the trainset9 fasta data to that region and filter it to remove the bases outside of the v35 region ([Wang et al. 2007](http://aem.asm.org/content/73/16/5261.long)). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd v35\n",
    "mothur \"#pcr.seqs(fasta=../references/silva.bacteria.fasta, start=6426, end=27654, outputdir=./, processors=8);filter.seqs(fasta=silva.bacteria.pcr.fasta, trump=., processors=8)\"\n",
    "cp silva.bacteria.pcr.filter.fasta silva.v35.fasta\n",
    "\n",
    "mothur \"#align.seqs(fasta=../references/trainset9_032012.pds.fasta, reference=../references/silva.bacteria.fasta, outputdir=./, processors=8); pcr.seqs(fasta=trainset9_032012.pds.align, taxonomy=../references/trainset9_032012.pds.tax, start=6426, end=27654, outputdir=./); degap.seqs(fasta=trainset9_032012.pds.pcr.align)\"\n",
    "cp trainset9_032012.pds.pcr.tax trainset9_032012.three.tax\n",
    "cp trainset9_032012.pds.pcr.ng.fasta trainset9_032012.three.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we are ready to run mothur with the sequences as a single dataset. What follows are the commands to be run within mothur or placed within a batch file (i.e. v35.mothur.batch; shown below) and run from the command line. Relevant references for these commands are available: aligner ([Schloss 2009](http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0008230) and [DeSantis et al. 2006](http://nar.oxfordjournals.org/content/34/suppl_2/W394.full)), pre-clustering ([Schloss et al. 2011](http://dx.plos.org/10.1371/journal.pone.0027310)), UChime ([Edgar et al. 2011](http://bioinformatics.oxfordjournals.org/content/27/16/2194.long)), and the classifier ([Schloss & Westcott 2011](http://aem.asm.org/content/77/10/3219.long) and [Wang et al. 2007](http://aem.asm.org/content/73/16/5261.long)). We'll first load the mothur magic module:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Installed mothurmagic.py. To use it, type:\n",
      "  %load_ext mothurmagic\n",
      "The mothurmagic extension is already loaded. To reload it, use:\n",
      "  %reload_ext mothurmagic\n"
     ]
    }
   ],
   "source": [
    "%install_ext https://raw.github.com/kdiverson/ipython-mothurmagic/master/mothurmagic.py\n",
    "%load_ext mothurmagic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%mothur\n",
    "\n",
    "#unique the sequences because there are duplicate sequences between samples\n",
    "unique.seqs(fasta=v35.fasta, name=v35.names)\n",
    "\n",
    "\n",
    "#align the sequences against the customized SILVA reference alignment for the v35 region\n",
    "#the user should alter the processors option to however many processors they have available\n",
    "align.seqs(fasta=current, reference=silva.v35.fasta, processors=8)\n",
    "\n",
    "\n",
    "#find the start and end coordinates for the aligned sequences\n",
    "summary.seqs(fasta=current, name=current)\n",
    "\n",
    "\n",
    "#remove the sequences that don't map to the correct region and then remove\n",
    "#any vertical gaps or columns that contain missing data\n",
    "screen.seqs(fasta=current, name=current, group=v35.groups, start=10010, end=21219)\n",
    "filter.seqs(fasta=current, vertical=T, trump=.)\n",
    "\n",
    "\n",
    "#unique again because the end trimming creates duplicate sequences\n",
    "unique.seqs(fasta=current, name=current)\n",
    "\n",
    "#pre-cluster sequences to remove residual errors and simplify the dataset.\n",
    "pre.cluster(fasta=current, group=current, name=current, diffs=2, processors=8)\n",
    "\n",
    "\n",
    "#use chimera.uchime to identify chimeras in the de novo detection mode. to do this, we\n",
    "#will process each sample separately and will will only consider a sequence as being \n",
    "#chimeric if it is flagged as chimeric in that sample. we then remove the chimeras from\n",
    "#the samples they were found in\n",
    "chimera.uchime(fasta=current, group=current, name=current, dereplicate=T, processors=8)\n",
    "remove.seqs(fasta=current, group=current, name=current, accnos=current, dups=F)\n",
    "\n",
    "\n",
    "#classify the sequences against our customized taxonomy reference based on the RDP\n",
    "#training set. we require that an assignment have a confidence score of at least 80%\n",
    "classify.seqs(fasta=current, reference=trainset9_032012.three.fasta, taxonomy=trainset9_032012.three.tax, cutoff=80, processors=8)\n",
    "\n",
    "\n",
    "#remove sequences that classify as either Chloroplasts, Mitochondria, Archaea, Eukaryota,\n",
    "#or as being unknown\n",
    "remove.lineage(fasta=current, group=current, name=current, taxonomy=current, taxon=Cyanobacteria_Chloroplast-Mitochondria-unknown-Archaea-Eukaryota)\n",
    "\n",
    "\n",
    "#assign sequences to bins / phylotypes based on their taxonomic assignment\n",
    "phylotype(taxonomy=current, name=current)\n",
    "\n",
    "\n",
    "#count the number of times each phylotype was observed in each sample at the genus level\n",
    "#(label=1)\n",
    "make.shared(list=current, group=current, label=1)\n",
    "\n",
    "\n",
    "#finally, output the taxonomic assignment for each phylotype\n",
    "classify.otu(list=current, group=current, name=current, taxonomy=current, label=1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Recall above we gave the samples some pretty funky names to keep every sequencing collection separate. We need to figure out which samples came from which subject and bodysite. We pooled datasets for each subject and bodysite where a sample was sequenced multiple times. Because WUGSC submitted their PPS data for the Phase 1 data, there is some redundancy. Now we need to remove the redundant reads with the help of a perl script (removeRedundant_WUGSC.pl). Go ahead and copy and paste the following into a file in the scripts folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#!perl\n",
    "\n",
    "use strict;\n",
    "use warnings;\n",
    "\n",
    "\n",
    "#\tfirst we want to get the names of the WUGSC groups that were sequenced in\n",
    "#\tthe pps.  we'll create the %groups hash and store the group names in the\n",
    "#\thash...\n",
    "\n",
    "my %groups;\n",
    "\n",
    "open(GROUPS, $ARGV[0]) or die;\n",
    "while(<GROUPS>){\n",
    "\tchomp(my $group = $_);\n",
    "\t$groups{$group} = 1;\n",
    "}\n",
    "\n",
    "close(GROUPS);\n",
    "\n",
    "\n",
    "#\tnext, we want to look through the shared file and find any group name that\n",
    "#\tstarts \"phase1\" and ends with the group name in the hash.  if it isn't in\n",
    "#\tthe hash then we'll keep it.  if not, we'll remove it.\n",
    "\n",
    "open(ORIG_SHARED, $ARGV[1]) or die;\n",
    "while(<ORIG_SHARED>){\n",
    "\tif($_ =~ /phase1\\..*\\.(WUGSC\\.\\S*)\\t/){\n",
    "\t\tif(!exists($groups{$1})){\n",
    "\t\t\tprint \"$_\";\n",
    "\t\t}\n",
    "\t}\n",
    "\telse{\n",
    "\t\tprint \"$_\";\n",
    "\t}\n",
    "}\n",
    "\n",
    "close(ORIG_SHARED);\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "#get the names of the samples from the shared file that came from WUGSC in the PPS phase\n",
    "#return the sequencing center (WUGSC; f=3) and the anonymized sample id (f=4)\n",
    "cut -f 2 v35.unique.good.filter.unique.precluster.unique.pick.three.wang.pick.tx.shared | grep \"pps.*WUGSC\" | cut -f 3,4 -d \".\" | sort | uniq > wugsc.pps.groups\n",
    "\n",
    "#remove any samples that start phase1 and end with the fields returned above\n",
    "perl ../scripts/removeRedundant_WUGSC.pl wugsc.pps.groups v35.unique.good.filter.unique.precluster.unique.pick.three.wang.pick.tx.shared > v35.three_phases.tx.shared\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There were a number of \"control\" samples included in the analysis from sequencing mock communities, generous donor samples, and water blanks. Unfortunately, once Phase 1 and 2 started, the sequencing centers became inconsistent in running these controls. Therefore, they don't really tell us anything other than what we've already established in our earlier efforts to build this pipeline. So we'll remove them using a perl script (removeControlSamples.pl)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "#!perl\n",
    "\n",
    "use strict;\n",
    "use warnings;\n",
    "\n",
    "open(SHARED, $ARGV[0]) or die;\n",
    "\n",
    "#the control samples all had underscores and none of the other samples did\n",
    "while(<SHARED>){\n",
    "\tif(!($_ =~ /_/)){\n",
    "\t\tprint $_;\n",
    "\t}\n",
    "}\n",
    "close(SHARED);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "perl ../scripts/removeControlSamples.pl v35.three_phases.tx.shared > v35.three_phases_nocontrol.tx.shared\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we want to figure out who the anonymized sample id corresponds to and which body site it came from as well as the visit number that the sample was taken from. To do this, it is essential that you have the dbGaP files. I'll assume that you have the following files in a folder named dbGaP that is located in the parent directory of v35 (the current directory):\n",
    "\n",
    "> phs000228.v3.pht001184.v3.p1.EMMES_HMP_Subject.MULTI.txt  \n",
    "> phs000228.v3.pht001185.v3.p1.EMMES_HMP_Sample.MULTI.txt  \n",
    "> phs000228.v3.pht001187.v3.p1.c1.EMMES_HMP_DCM.HMP.txt  \n",
    "> phs000228.v3.pht001193.v3.p1.c1.EMMES_HMP_GTV.HMP.txt  \n",
    "> phs000228.v3.pht002156.v1.p1.c1.EMMES_HMP_DSU.HMP.txt  \n",
    "> phs000228.v3.pht002157.v1.p1.c1.EMMES_HMP_DTP_DHX_DVD.HMP.txt  \n",
    "> phs000228.v3.pht002158.v1.p1.c1.EMMES_HMP_DEM_ENR.HMP.txt  \n",
    "\n",
    "Again, don't ask me for these files, you must get them through dbGaP. For this operation you will need phs000228.v3.pht001193.v3.p1.c1.EMMES_HMP_GTV.HMP.txt and our perl script (mapSamplesToMetadata.pl), which should be in the scripts folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#!perl\n",
    "\n",
    "use strict;\n",
    "use warnings;\n",
    "\n",
    "\n",
    "my %tracker;\n",
    "\n",
    "open(METADATA, $ARGV[1]) or die;\n",
    "while(<METADATA>){\n",
    "\tchomp($_);\n",
    "\t$_ =~ s/ /_/g;\n",
    "\n",
    "\tif($_ =~ /^\\d/){\n",
    "\t\tmy @line = split /\\t/, $_;\n",
    "\n",
    "\t\t$line[7] =~ s/(\\d\\d).*/$1/;\n",
    "\t\t$line[9] =~ s/.*DNA_(.*)/$1/;\n",
    "\n",
    "\t\t$tracker{$line[6]} = \"$line[13].$line[7].$line[9]\";\n",
    "\t}\n",
    "}\n",
    "close(METADATA);\n",
    "\n",
    "\n",
    "open(SAMPLE_IDS, $ARGV[0]) or die;\n",
    "while(<SAMPLE_IDS>){\n",
    "\tchomp($_);\n",
    "\tif($_ =~ /(\\d{9})/){\n",
    "\n",
    "\t\tprint \"$_\\t$tracker{$1}\\n\";\n",
    "\n",
    "\t}\n",
    "}\n",
    "close(SAMPLE_IDS);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "#get the names of all of the samples we have\n",
    "cut -f 2 v35.three_phases_nocontrol.tx.shared > sample.ids\n",
    "\n",
    "\n",
    "#map the sample.ids to the subject id, body site, and visit number\n",
    "perl ../scripts/mapSamplesToMetadata.pl sample.ids ../dbGaP/phs000228.v3.pht001193.v3.p1.c1.EMMES_HMP_GTV.HMP.txt > sample.design\n",
    "\n",
    "\n",
    "#the output of the perl script is a file that maps the sample names to more \n",
    "#meaningful names. using mothur, we can merge the frequency of each phylotype\n",
    "#for samples that need to be pooled\n",
    "mothur \"#merge.groups(shared=v35.three_phases_nocontrol.tx.shared, design=sample.design)\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At this point we have sample names that look like this:  \n",
    "> 160461578.01.Throat  \n",
    "> 158256496.02.Anterior_nares  \n",
    "> 160643649.01.Mid_vagina  \n",
    "> 159005010.02.Tongue_dorsum  \n",
    "> 159753524.01.Tongue_dorsum  \n",
    "> 160481808.01.R_Antecubital_fossa  \n",
    "> 159753524.01.Anterior_nares  \n",
    "> 160481808.01.Tongue_dorsum  \n",
    "> 159207311.01.R_Retroauricular_crease  \n",
    "> 159713063.02.Tongue_dorsum  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'd like to separate these samples now so that each body site has its own shared file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "head -n 1 v35.three_phases_nocontrol.tx.merge.shared > header.txt\n",
    "cp header.txt Anterior_nares.tx.shared\n",
    "cp header.txt Buccal_mucosa.tx.shared\n",
    "cp header.txt Hard_palate.tx.shared\n",
    "cp header.txt Keratinized_gingiva.tx.shared\n",
    "cp header.txt L_Antecubital_fossa.tx.shared\n",
    "cp header.txt L_Retroauricular_crease.tx.shared\n",
    "cp header.txt Mid_vagina.tx.shared\n",
    "cp header.txt Palatine_Tonsils.tx.shared\n",
    "cp header.txt Posterior_fornix.tx.shared\n",
    "cp header.txt R_Antecubital_fossa.tx.shared\n",
    "cp header.txt R_Retroauricular_crease.tx.shared\n",
    "cp header.txt Saliva.tx.shared\n",
    "cp header.txt Stool.tx.shared\n",
    "cp header.txt Subgingival_plaque.tx.shared\n",
    "cp header.txt Supragingival_plaque.tx.shared\n",
    "cp header.txt Throat.tx.shared\n",
    "cp header.txt Tongue_dorsum.tx.shared\n",
    "cp header.txt Vaginal_introitus.tx.shared\n",
    "\n",
    "grep \"Anterior_nares\" v35.three_phases_nocontrol.tx.merge.shared >> Anterior_nares.tx.shared\n",
    "grep \"Buccal_mucosa\" v35.three_phases_nocontrol.tx.merge.shared >> Buccal_mucosa.tx.shared\n",
    "grep \"Hard_palate\" v35.three_phases_nocontrol.tx.merge.shared >> Hard_palate.tx.shared\n",
    "grep \"Keratinized_gingiva\" v35.three_phases_nocontrol.tx.merge.shared >> Keratinized_gingiva.tx.shared\n",
    "grep \"L_Antecubital_fossa\" v35.three_phases_nocontrol.tx.merge.shared >> L_Antecubital_fossa.tx.shared\n",
    "grep \"L_Retroauricular_crease\" v35.three_phases_nocontrol.tx.merge.shared >> L_Retroauricular_crease.tx.shared\n",
    "grep \"Mid_vagina\" v35.three_phases_nocontrol.tx.merge.shared >> Mid_vagina.tx.shared\n",
    "grep \"Palatine_Tonsils\" v35.three_phases_nocontrol.tx.merge.shared >> Palatine_Tonsils.tx.shared\n",
    "grep \"Posterior_fornix\" v35.three_phases_nocontrol.tx.merge.shared >> Posterior_fornix.tx.shared\n",
    "grep \"R_Antecubital_fossa\" v35.three_phases_nocontrol.tx.merge.shared >> R_Antecubital_fossa.tx.shared\n",
    "grep \"R_Retroauricular_crease\" v35.three_phases_nocontrol.tx.merge.shared >> R_Retroauricular_crease.tx.shared\n",
    "grep \"Saliva\" v35.three_phases_nocontrol.tx.merge.shared >> Saliva.tx.shared\n",
    "grep \"Stool\" v35.three_phases_nocontrol.tx.merge.shared >> Stool.tx.shared\n",
    "grep \"Subgingival_plaque\" v35.three_phases_nocontrol.tx.merge.shared >> Subgingival_plaque.tx.shared\n",
    "grep \"Supragingival_plaque\" v35.three_phases_nocontrol.tx.merge.shared >> Supragingival_plaque.tx.shared\n",
    "grep \"Throat\" v35.three_phases_nocontrol.tx.merge.shared >> Throat.tx.shared\n",
    "grep \"Tongue_dorsum\" v35.three_phases_nocontrol.tx.merge.shared >> Tongue_dorsum.tx.shared\n",
    "grep \"Vaginal_introitus\" v35.three_phases_nocontrol.tx.merge.shared >> Vaginal_introitus.tx.shared\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition several of the body sites will be combined to facilitate their analysis (i.e. left and right antecubital fossa and retroauricular crease and the vaginal introitus, mid vagina, and posterior fornix)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "cp header.txt Vagina.tx.shared\n",
    "cp header.txt Retroauricular_crease.tx.shared\n",
    "cp header.txt Antecubital_fossa.tx.shared\n",
    "\n",
    "grep \"Vaginal_introitus\" v35.three_phases_nocontrol.tx.merge.shared >> Vagina.tx.shared\n",
    "grep \"Mid_vagina\" v35.three_phases_nocontrol.tx.merge.shared >> Vagina.tx.shared\n",
    "grep \"Posterior_fornix\" v35.three_phases_nocontrol.tx.merge.shared >> Vagina.tx.shared\n",
    "grep \"L_Antecubital_fossa\" v35.three_phases_nocontrol.tx.merge.shared >> Antecubital_fossa.tx.shared\n",
    "grep \"R_Antecubital_fossa\" v35.three_phases_nocontrol.tx.merge.shared >> Antecubital_fossa.tx.shared\n",
    "grep \"L_Retroauricular_crease\" v35.three_phases_nocontrol.tx.merge.shared >> Retroauricular_crease.tx.shared\n",
    "grep \"R_Retroauricular_crease\" v35.three_phases_nocontrol.tx.merge.shared >> Retroauricular_crease.tx.shared\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All of our next steps will depend on our ability to subsample the data to a common number of sequences. To figure out what our threshold should be, we will calculate the number of reads per sample and bodysite."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "mothur \"#summary.single(shared=Anterior_nares.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=Buccal_mucosa.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=Hard_palate.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=Keratinized_gingiva.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=L_Antecubital_fossa.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=L_Retroauricular_crease.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=Mid_vagina.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=Palatine_Tonsils.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=Posterior_fornix.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=R_Antecubital_fossa.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=R_Retroauricular_crease.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=Saliva.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=Stool.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=Subgingival_plaque.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=Supragingival_plaque.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=Throat.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=Tongue_dorsum.tx.shared, calc=nseqs);\"\n",
    "mothur \"#summary.single(shared=Vaginal_introitus.tx.shared, calc=nseqs);\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So let's use some R to figure out what percentage of the samples at each body site will be kept for various thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "threshold <- 1000\t#change this to see how the fraction of samples changes\n",
    "\n",
    "summary.files <- c(\"Anterior_nares.tx.groups.summary\", \"Buccal_mucosa.tx.groups.summary\", \n",
    "\t\t\t\t\"Hard_palate.tx.groups.summary\", \"Keratinized_gingiva.tx.groups.summary\",\n",
    "\t\t\t\t\"L_Antecubital_fossa.tx.groups.summary\", \"L_Retroauricular_crease.tx.groups.summary\",\n",
    "\t\t\t\t\"Mid_vagina.tx.groups.summary\", \"Palatine_Tonsils.tx.groups.summary\",\n",
    "\t\t\t\t\"Posterior_fornix.tx.groups.summary\", \"R_Antecubital_fossa.tx.groups.summary\",\n",
    "\t\t\t\t\"R_Retroauricular_crease.tx.groups.summary\", \"Saliva.tx.groups.summary\",\n",
    "\t\t\t\t\"Stool.tx.groups.summary\", \"Subgingival_plaque.tx.groups.summary\",\n",
    "\t\t\t\t\"Supragingival_plaque.tx.groups.summary\", \"Throat.tx.groups.summary\",\n",
    "\t\t\t\t\"Tongue_dorsum.tx.groups.summary\", \"Vaginal_introitus.tx.groups.summary\")\n",
    "\t\n",
    "percent <- numeric()\n",
    "\n",
    "for(file in summary.files){\n",
    "\tdata <- read.table(file=file, header=T)\n",
    "\tpercent[file] <- sum(data$nseqs >= threshold)/nrow(data)\n",
    "}\n",
    "\n",
    "percent\n",
    "hist(percent)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you play around with a few thresholds and look at the percentages you'll see that the antecubital foss and vaginal samples have a small number of reads in general. But if you start increasing the threshold above 1000 we start to lose a good number of samples from the other bodysites. We chose to go with 1000 because it gave us a good number of samples and reads per body site. With this information, we subsampled all of the bodysites to 1000 reads per sample. Any samples with fewer than 1000 reads were removed from the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%mothur\n",
    "sub.sample(shared=Antecubital_fossa.tx.shared, size=1000)\n",
    "sub.sample(shared=Anterior_nares.tx.shared, size=1000)\n",
    "sub.sample(shared=Buccal_mucosa.tx.shared, size=1000)\n",
    "sub.sample(shared=Hard_palate.tx.shared, size=1000)\n",
    "sub.sample(shared=Keratinized_gingiva.tx.shared, size=1000)\n",
    "sub.sample(shared=Palatine_Tonsils.tx.shared, size=1000)\n",
    "sub.sample(shared=Retroauricular_crease.tx.shared, size=1000)\n",
    "sub.sample(shared=Saliva.tx.shared, size=1000)\n",
    "sub.sample(shared=Stool.tx.shared, size=1000)\n",
    "sub.sample(shared=Subgingival_plaque.tx.shared, size=1000)\n",
    "sub.sample(shared=Supragingival_plaque.tx.shared, size=1000)\n",
    "sub.sample(shared=Throat.tx.shared, size=1000)\n",
    "sub.sample(shared=Tongue_dorsum.tx.shared, size=1000)\n",
    "sub.sample(shared=Vagina.tx.shared, size=1000)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assigning samples from each body site to community types\n",
    "Now we'd like to assign the samples in each body site to community types. We do this using the get.communitytype function in mothur ([Holmes et al. 2012](http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0030126)). This is a function that utilizes a random number generator and so there may be run-to-run variation. In our experience, with complicated datasets such as these, there is very little variation in the number of community types or the assignments. The label of the community type is most likely to vary between runs (e.g. in one run a group of samples might be labeled as \"Partition_1\" and in another \"Partition_3\"). We ran the everything through get.communitytype five times in separate folders:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir dmm_analysis\n",
    "cd dmm_analysis\n",
    "mkdir rep1 rep2 rep3 rep4 rep5\n",
    "\n",
    "\n",
    "cat > header.batch << \"_EOF_\"\n",
    "#!/bin/sh\n",
    "#PBS -l nodes=1:ppn=1\n",
    "#PBS -l walltime=500:00:00\n",
    "#PBS -l mem=1gb\n",
    "#PBS -j oe\n",
    "#PBS -m abe\n",
    "#PBS -V\n",
    "#PBS -M your@email.here\n",
    "#PBS -q first\n",
    "\n",
    "echo \"ncpus-2.pbs\"\n",
    "cat $PBS_NODEFILE\n",
    "qstat -f $PBS_JOBID\n",
    "\n",
    "cd $PBS_O_WORKDIR\n",
    "\n",
    "NCPUS=`wc -l $PBS_NODEFILE | awk '{print $1}'`\n",
    "\n",
    "_EOF_\n",
    "\n",
    "\n",
    "cat > tail.batch << \"_EOF_\"\n",
    "\n",
    "echo \"qsub working directory absolute is\"\n",
    "echo $PBS_O_WORKDIR\n",
    "exit\n",
    "_EOF_\n",
    "\n",
    "\n",
    "\n",
    "for REP in rep1 rep2 rep3 rep4 rep5\n",
    "do\n",
    "    cd $REP\n",
    "\n",
    "    for SHARED in $(ls ../../*subsample.shared)\n",
    "    do\n",
    "\t\tSITE=${SHARED//..\\/..\\//}\n",
    "\t    SITE=${SITE//.tx.1.subsample.shared/}\n",
    "\n",
    "        cat ../header.batch > $REP.$SITE.batch\n",
    "        echo \"mothur \\\"#get.communitytype(shared=$SHARED, outputdir=./)\\\"\" >> $REP.$SITE.batch\n",
    "        cat ../tail.batch >> $REP.$SITE.batch\n",
    "        \n",
    "        qsub $REP.$SITE.batch\n",
    "        \n",
    "    done\n",
    "    cd ../\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we needed to go through, find the number of community types that gave the best Laplace value for each body site and move the files to a single folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "bodysites <- c(\"Antecubital_fossa\", \"Anterior_nares\", \"Buccal_mucosa\", \"Hard_palate\", \"Keratinized_gingiva\", \"Palatine_Tonsils\", \"Retroauricular_crease\", \"Saliva\", \"Stool\", \"Subgingival_plaque\", \"Supragingival_plaque\", \"Throat\", \"Tongue_dorsum\", \"Vagina\")\n",
    "\n",
    "bs.scores <- matrix(rep(NA, rep(14*3)), ncol=3)\n",
    "colnames(bs.scores) <- c(\"min.iteration\", \"min.k\", \"min.laplace\")\n",
    "rownames(bs.scores) <- bodysites\n",
    "\n",
    "dir.create(\"dmm_best\")\n",
    "\n",
    "for(bs in bodysites){\n",
    "\treps <- c(\"rep1\", \"rep2\", \"rep3\", \"rep4\", \"rep5\")\n",
    "\tscores <- matrix(rep(NA, 5*10), ncol=5)\n",
    "\tcolnames(scores) <- reps\n",
    "\trownames(scores) <- 1:10\n",
    "\n",
    "\tfor(r in reps){\n",
    "\t\tfilename <- paste(\"dmm_analysis/\", r,\"/\",bs, \".tx.1.subsample.1.dmm.mix.fit\", sep=\"\")\n",
    "\t\tdata <- read.table(file=filename, header=T, row.names=1)\n",
    "\t\tscores[1:nrow(data),r] <- data$Laplace\n",
    "\t}\n",
    "\t\n",
    "\tmin.index <- which.min(scores)\n",
    "\tmin.k <- (min.index) %% 10\n",
    "\tmin.index <- min.index - min.k\n",
    "\tmin.iteration <- (min.index / 10) + 1\n",
    "\t\n",
    "\tc(min.k, min.iteration, scores[min.k, min.iteration])\n",
    "\n",
    "\tbs.scores[bs, ] <- c(min.iteration, min.k, scores[min.k, min.iteration])\n",
    "\t\n",
    "\tsource.filename <- list.files(path=paste(\"dmm_analysis/\", colnames(scores)[min.iteration], sep=\"\"), pattern=glob2rx(paste(bs, \".*\", sep=\"\")))\n",
    "\twrite(paste(\"dmm_analysis/\",colnames(scores)[min.iteration], \"/\", source.filename, sep=\"\"), \"\")\n",
    "\tfile.copy(paste(\"dmm_analysis/\",colnames(scores)[min.iteration], \"/\", source.filename, sep=\"\"), to=\"dmm_best/\")\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The get.communitytype outputs a number of files that we will use. For now, let's worry about the design file, which will tell us which community type each sample belongs to. Let's get the body specific design files for those bodysites that we pooled"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd dmm_best\n",
    "grep \"L_Retroauricular_crease\" Retroauricular_crease.tx.1.subsample.1.dmm.mix.design > L_Retroauricular_crease.tx.1.subsample.1.dmm.mix.design\n",
    "grep \"R_Retroauricular_crease\" Retroauricular_crease.tx.1.subsample.1.dmm.mix.design > R_Retroauricular_crease.tx.1.subsample.1.dmm.mix.design\n",
    "grep \"L_Antecubital_fossa\" Antecubital_fossa.tx.1.subsample.1.dmm.mix.design > L_Antecubital_fossa.tx.1.subsample.1.dmm.mix.design \n",
    "grep \"R_Antecubital_fossa\" Antecubital_fossa.tx.1.subsample.1.dmm.mix.design > R_Antecubital_fossa.tx.1.subsample.1.dmm.mix.design \n",
    "grep \"Vaginal_introitus\" Vagina.tx.1.subsample.1.dmm.mix.design > Vaginal_introitus.tx.1.subsample.1.dmm.mix.design\n",
    "grep \"Mid_vagina\" Vagina.tx.1.subsample.1.dmm.mix.design > Mid_vagina.tx.1.subsample.1.dmm.mix.design\n",
    "grep \"Posterior_fornix\" Vagina.tx.1.subsample.1.dmm.mix.design > Posterior_fornix.tx.1.subsample.1.dmm.mix.design\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One last thing we'd like to do is to create separate files where each file is a table with the subject along the rows and the body site is along the columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "body.design <- c(\"Anterior_nares.tx.1.subsample.1.dmm.mix.design\", \"Buccal_mucosa.tx.1.subsample.1.dmm.mix.design\",\n",
    "\t\t\t\t \"Hard_palate.tx.1.subsample.1.dmm.mix.design\", \"Keratinized_gingiva.tx.1.subsample.1.dmm.mix.design\",\n",
    "\t\t\t\t \"L_Antecubital_fossa.tx.1.subsample.1.dmm.mix.design\", \"L_Retroauricular_crease.tx.1.subsample.1.dmm.mix.design\",\n",
    "\t\t\t\t \"Mid_vagina.tx.1.subsample.1.dmm.mix.design\", \"Palatine_Tonsils.tx.1.subsample.1.dmm.mix.design\",\n",
    "\t\t\t\t \"Posterior_fornix.tx.1.subsample.1.dmm.mix.design\", \"R_Antecubital_fossa.tx.1.subsample.1.dmm.mix.design\",\n",
    "\t\t\t\t \"R_Retroauricular_crease.tx.1.subsample.1.dmm.mix.design\", \"Saliva.tx.1.subsample.1.dmm.mix.design\",\n",
    "\t\t\t\t \"Stool.tx.1.subsample.1.dmm.mix.design\", \"Subgingival_plaque.tx.1.subsample.1.dmm.mix.design\",\n",
    "\t\t\t\t \"Supragingival_plaque.tx.1.subsample.1.dmm.mix.design\", \"Throat.tx.1.subsample.1.dmm.mix.design\",\n",
    "\t\t\t\t \"Tongue_dorsum.tx.1.subsample.1.dmm.mix.design\", \"Vaginal_introitus.tx.1.subsample.1.dmm.mix.design\")\n",
    "\t\t\t\t \n",
    "subjects <- character()\n",
    "\n",
    "#i know this isn't efficient...\n",
    "for(design in body.design){\n",
    "\td <- read.table(file=design)\n",
    "\ts <- gsub(\"^(\\\\d*).*\", \"\\\\1\", d$V1)\n",
    "\tsubjects <- c(subjects, s)\n",
    "}\n",
    "\n",
    "subjects <- sort(unique(subjects))\n",
    "nsubjects <- length(subjects)\t#should be ~300\n",
    "\n",
    "visit1 <- matrix(rep(NA, nsubjects*18), nrow=nsubjects)\n",
    "rownames(visit1) <- subjects\n",
    "colnames(visit1) <- gsub(\"^([^.]*).*\", \"\\\\1\", body.design)\n",
    "\n",
    "visit2 <- visit1\n",
    "visit3 <- visit1\n",
    "\n",
    "for(design in body.design){\t\n",
    "\td <- read.table(file=design)\n",
    "\ts <- gsub(\"^(\\\\d*).*\", \"\\\\1\", d$V1)\t\t\t\t#get the subject id\n",
    "\tv <- gsub(\"^\\\\d*.(\\\\d\\\\d).*\", \"\\\\1\", d$V1)\t\t#get the visit number\n",
    "\tb <- gsub(\"^\\\\d*.\\\\d\\\\d.(.*)\", \"\\\\1\", d$V1)\t\t#get the visit number\n",
    "\n",
    "\tvisit1[as.character(s[v == \"01\"]), b] <- as.character(d[v == \"01\", \"V2\"])\n",
    "\tvisit2[as.character(s[v == \"02\"]), b] <- as.character(d[v == \"02\", \"V2\"])\n",
    "\tvisit3[as.character(s[v == \"03\"]), b] <- as.character(d[v == \"03\", \"V2\"])\n",
    "}\n",
    "\n",
    "sum.na <- function(x){\n",
    "\treturn(sum(is.na(x)))\n",
    "}\n",
    "\n",
    "keep <- apply(visit1, 1, sum.na) != 18\t#get rid of subject ids that only have NAs\n",
    "visit1 <- visit1[keep,]\n",
    "\n",
    "keep <- apply(visit2, 1, sum.na) != 18\t#get rid of subject ids that only have NAs\n",
    "visit2 <- visit2[keep,]\n",
    "\n",
    "keep <- apply(visit3, 1, sum.na) != 18\t#get rid of subject ids that only have NAs\n",
    "visit3 <- visit3[keep,]\n",
    "\n",
    "write.table(visit1, \"community_types.v1.txt\", quote=F)\n",
    "write.table(visit2, \"community_types.v2.txt\", quote=F)\n",
    "write.table(visit3, \"community_types.v3.txt\", quote=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir ../../analysis\n",
    "mv community_types.v* ../../analysis\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clinical metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***A lot*** of clinical metadata were collected on the 300 subjects in this study. Unfortunately, you always think of more metadata that you would like to have after you start to look at the data. Another issue with clinical metadata, especially in the case of the HMP dataset was that because the subjects were so young and healthy, there often isn't a lot of variation and the metadata don't turn out to be as interesting as originally hoped. For instance, if only a few women have had a baby, it is difficult to see whether having a baby has affected her microbiome. Ditto for a bunch of other variables. Again, all of the clinical metadata are available through dbGaP and unfortunately, you have to get them there. There is a process to getting the data that is tedious but not horrible. But whatever you do, don't ask me for them. I'll say no. Finally, once you get the dbGaP files, unencrypt them and throw them into the dbgap folder that we made way back at the top of this notebook.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ../../dbGaP\n",
    "ls *txt\n",
    "\n",
    "#should produce...\n",
    "#phs000228.v3.pht001184.v3.p1.EMMES_HMP_Subject.MULTI.txt\n",
    "#phs000228.v3.pht001185.v3.p1.EMMES_HMP_Sample.MULTI.txt\n",
    "#phs000228.v3.pht001187.v3.p1.c1.EMMES_HMP_DCM.HMP.txt\n",
    "#phs000228.v3.pht001193.v3.p1.c1.EMMES_HMP_GTV.HMP.txt\n",
    "#phs000228.v3.pht002156.v1.p1.c1.EMMES_HMP_DSU.HMP.txt\n",
    "#phs000228.v3.pht002157.v1.p1.c1.EMMES_HMP_DTP_DHX_DVD.HMP.txt\n",
    "#phs000228.v3.pht002158.v1.p1.c1.EMMES_HMP_DEM_ENR.HMP.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before we do science, we need to sift through the clinical metadata and see what of it is useful and what isn't. Recall that you can go to the [dbGaP project site](http://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs000228.v3.p1) to figure out what the various files contain and what the variable names mean. Because some of these txt files contain apostrophes that were wrecking havoc with R, I used TextWrangeler to zap all of them away. A smarter person could probalby have done it with sed. Good for you. The first thing we'll do is recreate the data in Extended Data Table 1 from the paper.\n",
    "\n",
    "Three of these files contain useful categorical metadata that we'll read in, sort, and merge into one table. Also, because patients provided the data at different visits in some of the files, we will have to be careful about picking the right visit (VISNO==\"01\")..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "categ.1 <-read.table(file=\"phs000228.v3.pht002157.v1.p1.c1.EMMES_HMP_DTP_DHX_DVD.HMP.txt\", header=T, sep=\"\\t\",\n",
    "                     na.string=c(\"\", NA))\n",
    "categ.1 <- categ.1[categ.1$VISNO == \"01\",]\n",
    "rownames(categ.1) <- categ.1$RANDSID\n",
    "categ.1 <- categ.1[,-c(ncol(categ.1))]\t#the RANDSID column is the last one\n",
    "categ.1 <- categ.1[sort(rownames(categ.1)),]\n",
    "\n",
    "\n",
    "categ.2 <-read.table(file=\"phs000228.v3.pht002158.v1.p1.c1.EMMES_HMP_DEM_ENR.HMP.txt\", header=T, sep=\"\\t\", row.names=2,\n",
    "                     na.string=c(\"\", NA))\n",
    "categ.2 <- categ.2[sort(rownames(categ.2)),]\n",
    "\n",
    "duplicates <- intersect(colnames(categ.1), colnames(categ.2)) #the columns that are in both files\n",
    "categ.2 <- categ.2[,-which(colnames(categ.2) %in% duplicates)]\t#get rid of the duplicates\n",
    "sum(rownames(categ.1) == rownames(categ.2)) == 300\t#make sure the rows are in the same order\n",
    "\n",
    "categorical <- cbind(categ.1, categ.2)\n",
    "\n",
    "\n",
    "#using textwrangler I first zapped all of the apostrophes\n",
    "categ.3 <- read.table(file=\"phs000228.v3.pht002156.v1.p1.c1.EMMES_HMP_DSU.HMP.txt\", header=T, sep=\"\\t\", na.string=c(\"\", NA))\n",
    "rownames(categ.3) <- categ.3$RANDSID\n",
    "categ.3 <- categ.3[,-c(ncol(categ.3))]\t#the RANDSID column is the last column\n",
    "setdiff(rownames(categ.3), rownames(categorical))\n",
    "dim(categ.3)\t#there weren't 300 subjects here and so some are missing\n",
    "\n",
    "missing <- rownames(categorical)[!(rownames(categorical) %in% rownames(categ.3))]\n",
    "categ.3[as.character(missing),] <- rep(NA, ncol(categ.3))\n",
    "categ.3 <- categ.3[sort(rownames(categ.3)),]\n",
    "duplicates <- intersect(colnames(categorical), colnames(categ.3)) #the columns that are in both files\n",
    "categ.3 <- categ.3[,-which(colnames(categ.3) %in% duplicates)]\t#get rid of the duplicates\n",
    "sum(rownames(categ.1) == rownames(categ.3)) == 300\t#make sure the rows are in the same order\n",
    "\n",
    "categorical <- cbind(categorical, categ.3)\n",
    "\n",
    "ncats <- ncol(categorical)\t#196\n",
    "npeople <- nrow(categorical)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So now we have our massive table of 300 subjects by 196 metadata variables. Now we will convert \"vague\" answers to \"NA\" and then identify those metadata columns that were NA for every subject:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "for(i in colnames(categorical)){\n",
    "\tcategorical[grepl(\"Dont know/remember\",categorical[,i]),i] <- NA\n",
    "\tcategorical[grepl(\"NotEvaluated\",categorical[,i]),i] <- NA\n",
    "\tcategorical[grepl(\"Forgot\",categorical[,i]),i] <- NA\n",
    "}\n",
    "\n",
    "#there's a lot of crap in here, let's figure out how to remove the variables that are all NAs\n",
    "sum.na <- function(x) sum(is.na(x) | x == \"\")\n",
    "na.count <- apply(categorical, 2, sum.na)\n",
    "notall.na <- na.count <= 250\t#empirically determined number\n",
    "\n",
    "#things with more than 250 NAs - ethnicities, tobacco usage, medical history, some drug usage\n",
    "\n",
    "categorical <- categorical[,notall.na]\n",
    "ncats <- ncol(categorical)\t#101\n",
    "    \n",
    "summary(categorical)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So we effectively reduced the amount of metadata in half. With the output of the summary command we see the various metadata fields. Looking through the dbGaP site we get a sense of what they are. A couple things pop out.  First, many of the variables are numerically coded categoritcal varaibles (e.g. SITE) and some are the same thing but coded with a string (e.g. SITE_C). We only need one of each of these. Second, some variables aren't useful (e.g. PROT). Third, some variables are continuous (e.g. AGEENR) and will be dealt with separately. Finally, some variables merely indicate whether a sample was taken at that bodysite (e.g. DVDINTRO_C). Here are the variables we decided to remove and why...\n",
    "\n",
    "\n",
    "**Variables to remove: redundant / non-informative**  \n",
    "> SITE - redundant with SITE_C  \n",
    "> DVDEDLVL_C - redundant with DVDEDLVL  \n",
    "> DVDINSRH - redundant with DVDINSRH_C  \n",
    "> DVDINSRD - redundant with DVDINSRD_C  \n",
    "> DVDOCPTN - redundant with DVDOCPTN_C  \n",
    "> DVDTOBC - redundant with DVDTOBC_C  \n",
    "> DVDPGRES - Pregnant? no one was pregnant  \n",
    "> DVDPGRES_C - Pregnant? no one was pregnant  \n",
    "> START - starting quarter of the study  \n",
    "> GENDER - redundnat with GENDER_C  \n",
    "> ETHNIC - redundant with ETHNIC_C  \n",
    "> AMERIND - american indian (only 1)  \n",
    "> AMERIND_C - american indian (only 1)  \n",
    "> ASIAN - redundant with ASIAN_C  \n",
    "> PACIFC - pacific islander (none)  \n",
    "> PACIFC_C - pacific islander (none)  \n",
    "> WHITE - redundant with WHITE_C  \n",
    "> BRTHCTRY - redundant with BRTHCTRY_C  \n",
    "> MOMCTRY - redundant with MOMCTRY_C  \n",
    "> DADCTRY - redundant with DADCTRY_C  \n",
    "> DSUSCRN - redundant with DSUSCRN_C  \n",
    "> DSUCON - redundant with DSUCON_C  \n",
    "> DSUDIET_C - redundant with DSUDIET  \n",
    "> DSUBIR - redunant with DSUBIR_C  \n",
    "> DSUBFED - redundant with DSUBFED_C  \n",
    "> BLACK - redundant with BLACK_C  \n",
    "> DVDVIRRT / DVDVIRRT_C - vaginal or vulval irritation - only 8 yes\n",
    "\n",
    "\n",
    "**not useful**  \n",
    "> dbGaP.SubjID - dbGaP Subject ID  \n",
    "> PROT - Protocol  \n",
    "> PROTSEG - Protocol   \n",
    "> protseg - Protocol  \n",
    "> VISNO - Visit number  \n",
    "> study_day - day of the study  \n",
    "> dsuotrpd - time since last oral treatment - eh.  \n",
    "> dsuvspd - time since enrollment  \n",
    "> IDSUZIP - zip code  \n",
    "> DSUSCRN_C - screened more than once  \n",
    "> DSUCON_C - subject contact  \n",
    "\n",
    "\n",
    "**Variables to remove: continuous variables**  \n",
    "> DVDINTPH - vagina introitus pH  \n",
    "> DVDPFPH - posterior fornix pH  \n",
    "> DVDTMPF - temperature in F  \n",
    "> DVDTMPC - temperature in C  \n",
    "> AGEENR - age  \n",
    "\n",
    "\n",
    "**Variables to remove: indicate whether various body sites were sampled**  \n",
    "> DVDBLOOD / DVDBLOOD_C - blood  \n",
    "> DVDSERUM / DVDSERUM_C - serum  \n",
    "> DVDSTOOL / DVDSTOOL_C - stool  \n",
    "> DVDSAL / DVDSAL_C - saliva  \n",
    "> DVDGING / DVDGING_C - gingiva  \n",
    "> DVDTONG / DVDTONG_C - tongue  \n",
    "> DVDPTON / DVDPTON_C - tonsils  \n",
    "> DVDHPAL / DVDHPAL_C - hard palate  \n",
    "> DVDTHRO / DVDTHRO_C - throat   \n",
    "> DVDBUCC / DVDBUCC_C - buccal mucosa  \n",
    "> DVDSUPRA / DVDSUPRA_C - supragingival plaque  \n",
    "> DVDSUB / DVDSUB_C - subgingial plaque  \n",
    "> DVDRACR / DVDRACR_C - right retroauricular crease  \n",
    "> DVDRACL / DVDRACL_C - left retroauricular crease  \n",
    "> DVDACRR / DVDACRR_C - right antecuibtal fossa  \n",
    "> DVDACRL / DVDACRL_C - left antecuibtal fossa  \n",
    "> DVDNASAL / DVDNASAL_C - anterior nares  \n",
    "> DVDINTRO / DVDINTRO_C - vaginal introitus  \n",
    "> DVDPFORN / DVDPFORN_C - posterior fornix  \n",
    "> DVDMIDVG / DVDMIDVG_C - mid vagina  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "to.remove <- c(\"dbGaP.SubjID\", \"PROT\", \"PROTSEG\", \"SITE_C\", \"protseg\", \"VISNO\",\n",
    "\t\t\t   \"study_day\", \"DVDEDLVL_C\", \"DVDINSRH\", \"DVDINSRD\", \"DVDOCPTN\",\n",
    "\t\t\t   \"DVDTOBC\", \"DVDPGRES\", \"DVDPGRES_C\", \"START\", \"GENDER\", \"ETHNIC\",\n",
    "\t\t\t   \"AMERIND\", \"AMERIND_C\", \"ASIAN\", \"PACIFC\", \"PACIFC_C\", \"WHITE\",\n",
    "\t\t\t   \"BRTHCTRY\", \"MOMCTRY\", \"DADCTRY\", \"DSUSCRN\", \"DSUCON\", \"BLACK\",\n",
    "\t\t\t   \"DSUDIET_C\", \"DSUBIR\", \"dsuotrpd\", \"dsuvspd\", \"DSUZIP\",\n",
    "\t\t\t   \"DVDINTPH\", \"DVDPFPH\", \"DVDTMPF\", \"DVDTMPC\", \"AGEENR\", \"DSUBFED\",\n",
    "\t\t\t   \"DVDBLOOD\", \"DVDBLOOD_C\", \"DVDSERUM\", \"DVDSERUM_C\", \"DVDSTOOL\",\n",
    "\t\t\t   \"DVDSTOOL_C\", \"DVDSAL\", \"DVDSAL_C\", \"DVDGING\", \"DVDGING_C\",\n",
    "\t\t\t   \"DVDTONG\", \"DVDTONG_C\", \"DVDPTON\", \"DVDPTON_C\", \"DVDHPAL\",\n",
    "\t\t\t   \"DVDHPAL_C\", \"DVDTHRO\", \"DVDTHRO_C \", \"DVDBUCC\", \"DVDBUCC_C\",\n",
    "\t\t\t   \"DVDSUPRA\", \"DVDSUPRA_C\", \"DVDSUB\", \"DVDSUB_C\", \"DVDRACR\",\n",
    "\t\t\t   \"DVDRACR_C\", \"DVDRACL\", \"DVDRACL_C\", \"DVDACRR\", \"DVDACRR_C\",\n",
    "\t\t\t   \"DVDACRL\", \"DVDACRL_C\", \"DVDNASAL\", \"DVDNASAL_C\", \"DVDVIRRT\",\n",
    "\t\t\t   \"DVDVIRRT_C\", \"DVDINTRO\", \"DVDINTRO_C\", \"DVDPFORN\", \"DVDPFORN_C\",\n",
    "\t\t\t   \"DVDMIDVG\", \"DVDMIDVG_C\", \"DVDTHRO_C\", \"DSUSCRN_C\", \"DSUCON_C\")\n",
    "\t\t\t   \n",
    "categorical <- categorical[,-which(colnames(categorical) %in% to.remove)]\n",
    "\n",
    "dim(categorical) #17 categories remain\n",
    "summary(categorical)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So we're left with 17 variables at this point Looking at the output of the summary command we see that many of them are now binary variables:  \n",
    "\n",
    "> SITE - city of origin - Houston / StLouis  \n",
    "> GENDER_C - gender - Male / Female  \n",
    "> DVDTOBC_C - tobacco user - yes / no  \n",
    "> BLACK_C - yes / no  \n",
    "> WHITE_C - yes /no  \n",
    "> DSUBFED_C - Were you breastfed as a child? (don't know was pooled with NA)  \n",
    "> DVDINSRH_C - health insurance?  \n",
    "> DVDINSRD_C - dental insurance?  \n",
    "\n",
    "and others have numerous levels without much heft to any of them. Therefore we re-keyed the other variables to make them binary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "#DVDEDLVL_C - education level - 8 different levels\n",
    "#turn it into BS or higher\n",
    "categorical$DVDEDLVL <- categorical$DVDEDLVL == 8 | categorical$DVDEDLVL == 9 | categorical$DVDEDLVL == 10 |\n",
    "                        categorical$DVDEDLVL == 11 \n",
    "\n",
    "\n",
    "#DVDOCPTN_C - Occupation - 8 different levels\n",
    "#turn it into Student or not\n",
    "categorical$DVDOCPTN_C <- categorical$DVDOCPTN_C == \"42 Student\"\n",
    "\n",
    "\n",
    "#BRTHCTRY_C, MOMCTRY_C, DADCTRY_C - birth country of subject and parents\n",
    "#14 different levels - turn into whether or not \"Canada/US\"\n",
    "categorical$BRTHCTRY_C <- categorical$BRTHCTRY_C == \"Canada/US\"\n",
    "categorical$MOMCTRY_C <- categorical$MOMCTRY_C == \"Canada/US\"\n",
    "categorical$DADCTRY_C <- categorical$DADCTRY_C == \"Canada/US\"\n",
    "\n",
    "\n",
    "#DSUDIET - four levels of meat consumption - two with meat and two without\n",
    "#turn it into meat or not\n",
    "categorical$DSUDIET <- categorical$DSUDIET == 1 | categorical$DSUDIET == 2\n",
    "\n",
    "\n",
    "#DSUBIR_C - whether subject had given birth never, once or more than once\n",
    "#turn it into whether the subject had given birth\n",
    "categorical$DSUBIR_C <- categorical$DSUBIR_C == \"More than once\" | categorical$DSUBIR_C == \"Once\"\n",
    "\n",
    "\n",
    "#ETHNIC_C - Hispanic / latino / spanish?\n",
    "categorical$ETHNIC_C <- categorical$ETHNIC_C == \"Hispanic or Latino or Spanish\"\n",
    "\n",
    "summary(categorical)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When we looked through the medication data we noticed that a number of people were chronically on medications. So we decided to make this a categorical varible. We found that the most commonly used medications were the following:  \n",
    "\n",
    " - 13=Contraceptives (oral, implant, injectable)(N=102)\n",
    " - 21=Vitamins, minerals, food supplements (N=54)\n",
    " - 9=Antidepressants/mood-altering drugs (N=37)\n",
    " - 8=Antibiotics/(anti)-infectives, parasitics, microbials (N=33)\n",
    " - 10=Antihistamines/Decongestants (N=28)\n",
    " \n",
    "To get going, we want to get the start and stop dates for each subject's medication usage as well as the dates for each subject's visit:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "medication <- read.table(file=\"phs000228.v3.pht001187.v3.p1.c1.EMMES_HMP_DCM.HMP.txt\", header=T, sep=\"\\t\", na.string=c(\"\", NA))\n",
    "\n",
    "sort(summary(as.factor(medication$DCMCODE_C)))\n",
    "\n",
    "#the most commonly used medications in the study...\n",
    "#13=Contraceptives (oral, implant, injectable) \t\t\t\t102\n",
    "#21=Vitamins, minerals, food supplements      \t\t\t\t54\n",
    "#9=Antidepressants/mood-altering drugs \t\t\t\t\t\t37\n",
    "#8=Antibiotics/(anti)-infectives, parasitics, microbials    33\n",
    "#10=Antihistamines/Decongestants\t\t\t\t\t\t    28 \n",
    "\n",
    "#pull out thse five medication classes\n",
    "medication <- medication[,c(\"RANDSID\", \"DCMCODE\", \"MSTART\", \"MSTOP\")]\n",
    "popular.user <- medication$DCMCODE == 13 | medication$DCMCODE == 21 | medication$DCMCODE == 9 |\n",
    "                    medication$DCMCODE == 8 | medication$DCMCODE == 10\n",
    "medication <- medication[popular.user,] #only want those lines from the file with these medications\n",
    "medication <- medication[order(medication$RANDSID),]   #make sure everything is in the right order\n",
    "\n",
    "\n",
    "#figure out the study date that each person visited\n",
    "metadata <- read.table(file=\"phs000228.v3.pht002157.v1.p1.c1.EMMES_HMP_DTP_DHX_DVD.HMP.txt\", header=T, sep=\"\\t\",\n",
    "                       na.string=c(\"\", NA))\n",
    "normalvisit <- metadata$VISNO == \"01\" | metadata$VISNO == \"02\" | metadata$VISNO == \"03\"\n",
    "metadata <- metadata[normalvisit,]  #only get those rows that corresponded to a normal visit\n",
    "subjects <- levels(as.factor(metadata$RANDSID))\n",
    "subjects == sort(subjects)   #make sure everything is in the right order\n",
    "\n",
    "\n",
    "#now we want to populate a table with the date for each person's visits\n",
    "visit.dates <- matrix(rep(NA, length(subjects)*3), ncol=3) #start with everyone at NA\n",
    "rownames(visit.dates) <- subjects #subject ids in the rows\n",
    "colnames(visit.dates) <- c(\"Visit1\", \"Visit2\", \"Visit3\")  #visits in the columns\n",
    "\n",
    "\n",
    "#loop through and fill in the study dates for each person's first, second, and third visit\n",
    "for(r in rownames(metadata)){\n",
    "\tif(metadata[r, \"VISNO\"] == \"01\"){\n",
    "\t\tvisit.dates[as.character(metadata[r,\"RANDSID\"]), \"Visit1\"] = metadata[r, \"study_day\"]\n",
    "\t}\n",
    "\telse if(metadata[r, \"VISNO\"] == \"02\"){\n",
    "\t\tvisit.dates[as.character(metadata[r,\"RANDSID\"]), \"Visit2\"] = metadata[r, \"study_day\"]\n",
    "\t}\n",
    "\telse if(metadata[r, \"VISNO\"] == \"03\"){\n",
    "\t\tvisit.dates[as.character(metadata[r,\"RANDSID\"]), \"Visit3\"] = metadata[r, \"study_day\"]\n",
    "\t}\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that at this point if someone didn't have a second or third visit, they will be listed as NA. Now we want to figure out whether a visit corresponded to the subjects being on the medications. We defined subjects that last took the medication within 30 days of a visit as \"being on the medication\" for that visit. Then we defined chronic usage as being on the medication for all of that subject's visits. With these criteria the time from taking the medication didn't have a big effect on the number of chronic users."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "#this function asks whether the subject was on the medication at their visit\n",
    "getDrugVisitData <- function(drug.metadata){\n",
    "\n",
    "\tdrug <- matrix(rep(NA, length(subjects)*3), ncol=3)\n",
    "\tcolnames(drug) <- c(\"Visit1\", \"Visit2\", \"Visit3\")\n",
    "\trownames(drug) <- rownames(visit.dates)\n",
    "\tdrug[!is.na(visit.dates)] <- FALSE  #if the subject visited, then we assume they're negative to start\n",
    "\tdrug[is.na(visit.dates)] <- NA      #if they didn't visit, we don't care\n",
    "\n",
    "\n",
    "\tvisitwindow <- 30\t#the person discontinued the drug in the last 30 days\n",
    "\n",
    "\tfor(r in rownames(drug.metadata)){\n",
    "\t\ts <- drug.metadata[as.character(r), \"RANDSID\"]\n",
    "\t\tstart <- drug.metadata[as.character(r), \"MSTART\"]\n",
    "\t\tstop <- drug.metadata[as.character(r), \"MSTOP\"]\n",
    "\t\tif(is.na(stop))\tstop = 100000   #if the stop date is NA then assume that they were on the medication through the end\n",
    "                                        #of the study\n",
    "\t\n",
    "            \n",
    "        #these work like so... - subject is assumed false initially and if they're subsequently changed to true, then we can't\n",
    "        #change them to false. if the start date is before the visit date adn the stop date is earlier than the stop date plus\n",
    "        #the visitwindow (30 days) then we consider them true. otherwise, keep them as false.\n",
    "            \n",
    "\t\tif(drug[as.character(s), \"Visit1\"] == FALSE & start <= visit.dates[as.character(s), \"Visit1\"] &\n",
    "           stop+visitwindow >= visit.dates[as.character(s), \"Visit1\"] & !is.na(visit.dates[as.character(s), \"Visit1\"])){\n",
    "\t\t\tdrug[as.character(s), \"Visit1\"] = TRUE\n",
    "\t\t} else if(drug[as.character(s), \"Visit1\"] == FALSE & !is.na(visit.dates[as.character(s), \"Visit1\"]))\n",
    "\t\t\tdrug[as.character(s), \"Visit1\"] = FALSE\n",
    "\n",
    "\t\tif(drug[as.character(s), \"Visit2\"] == FALSE & start <= visit.dates[as.character(s), \"Visit2\"] &\n",
    "           stop+visitwindow >= visit.dates[as.character(s), \"Visit2\"] & !is.na(visit.dates[as.character(s), \"Visit2\"])){\n",
    "\t\t\tdrug[as.character(s), \"Visit2\"] = TRUE\n",
    "\t\t} else if(drug[as.character(s), \"Visit2\"] == FALSE & !is.na(visit.dates[as.character(s), \"Visit2\"]))\n",
    "\t\t\tdrug[as.character(s), \"Visit2\"] = FALSE\n",
    "\n",
    "\t\tif(drug[as.character(s), \"Visit3\"] == FALSE & start <= visit.dates[as.character(s), \"Visit3\"] &\n",
    "           stop+visitwindow >= visit.dates[as.character(s), \"Visit3\"] & !is.na(visit.dates[as.character(s), \"Visit3\"])){\n",
    "\t\t\tdrug[as.character(s), \"Visit3\"] = TRUE\n",
    "\t\t} else if(drug[as.character(s), \"Visit3\"] == FALSE & !is.na(visit.dates[as.character(s), \"Visit3\"]))\n",
    "\t\t\tdrug[as.character(s), \"Visit3\"] = FALSE\n",
    "\t}\n",
    "\t\n",
    "    #send back a table with visits as columns, subjects as rows, and values indicate whether the subject was on the medication\n",
    "\treturn(drug)\n",
    "}\n",
    "\n",
    "            \n",
    "#this function defines whether the subject was a chronic user or not\n",
    "getChronic <- function(drug){\n",
    "\tchronic <- rep(FALSE, nrow(drug))\n",
    "    \n",
    "    #subject had to be on medication for all of their visits to be considered a chronic user\n",
    "\tchronic[((drug[,\"Visit1\"] == T & is.na(drug[,\"Visit2\"]) & is.na(drug[,\"Visit3\"])) |\n",
    "\t\t\t (drug[,\"Visit1\"] == T & drug[,\"Visit2\"] == T & is.na(drug[,\"Visit3\"])) |\n",
    "\t\t\t (drug[,\"Visit1\"] == T & drug[,\"Visit2\"] == T & drug[,\"Visit3\"] == T))] <- T\n",
    "\n",
    "    #returns a vector indicating whether each subject was chronic\n",
    "\treturn(chronic)\n",
    "\t\n",
    "}\n",
    "\n",
    "\n",
    "contra <- getDrugVisitData(medication[medication$DCMCODE == 13,])\n",
    "vit_min <- getDrugVisitData(medication[medication$DCMCODE == 21,])\n",
    "antidep <- getDrugVisitData(medication[medication$DCMCODE == 9,])\n",
    "antibiot <- getDrugVisitData(medication[medication$DCMCODE == 8,])\n",
    "antihist <- getDrugVisitData(medication[medication$DCMCODE == 10,])\n",
    "\n",
    "contra.chronic <- getChronic(contra)\n",
    "vit_min.chronic <- getChronic(vit_min)\n",
    "antidep.chronic <- getChronic(antidep)\n",
    "antihist.chronic <- getChronic(antihist)\n",
    "antibiot.chronic <- getChronic(antibiot)\n",
    "contra.chronic[categorical$GENDER_C==\"Male\"] <- NA\n",
    "\n",
    "            \n",
    "sum(contra.chronic, na.rm=T); sum(apply(contra, 1, sum, na.rm=T) > 0)  #72; 87\n",
    "sum(vit_min.chronic); sum(apply(vit_min, 1, sum, na.rm=T) > 0)         #34; 44\n",
    "sum(antidep.chronic); sum(apply(antidep, 1, sum, na.rm=T) > 0)         #22; 31\n",
    "sum(antihist.chronic); sum(apply(antihist, 1, sum, na.rm=T) > 0)       #9; 22\n",
    "sum(antibiot.chronic); sum(apply(antibiot, 1, sum, na.rm=T) > 0)       #2; 11\n",
    "#not enough antibiotic.chronic subjects to do anything\n",
    "\n",
    "#finally, append chronic contraception, vitamin/mineral, antidepressents, and antihistamine users to categorical table\n",
    "categorical <- cbind(categorical, contra.chronic, vit_min.chronic, antidep.chronic, antihist.chronic)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There's one more piece of categorical data we'd like to get - the subjects' BMI expressed as a categorical varaible."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cont <- read.table(file=\"phs000228.v3.pht002157.v1.p1.c1.EMMES_HMP_DTP_DHX_DVD.HMP.txt\", header=T, sep=\"\\t\",\n",
    "                   na.string=c(\"\", NA))\n",
    "cont <- cont[cont$VISNO == \"00\",]  #BMI was recorded based on the intake visit\n",
    "rownames(cont) <- cont$RANDSID\n",
    "cont <- cont[order(rownames(cont)),]\n",
    "\n",
    "bmi <- cont[,\"DTPBMI\"] #this is the only variable we need from the table\n",
    "\n",
    "BMI_C <- character()\n",
    "BMI_C[bmi<18.5] <- \"underweight\"\n",
    "BMI_C[bmi>=18.5 & bmi < 25] <- \"normal\"\n",
    "BMI_C[bmi>=25 & bmi < 30] <- \"overweight\"\n",
    "BMI_C[bmi>=30] <- \"obese\"\n",
    "BMI_C <- factor(BMI_C)\n",
    "rownames(cont) == rownames(categorical) #double check that our rows are in the correct order\n",
    "\n",
    "\n",
    "#let's append the BMI_C varaiable to the end of categorical\n",
    "categorical <- cbind(categorical, BMI_C)\n",
    "for(c in colnames(categorical)){\tcategorical[,c] <- factor(categorical[,c])\t}\n",
    "\n",
    "\n",
    "#output the table to a file\n",
    "write.table(categorical, \"categorical.metadata\", quote=F)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With that last command we have outputted the categorical data to a file. Let's move on to the continuous data. We found the following continuous data via dbGaP:  \n",
    "\n",
    "> DVDINTPH - pH - vaginal introitus  \n",
    "> DVDPFPH - pH - posterior fornix  \n",
    "> DTPPULSE - pulse  \n",
    "> DTPBMI - BMI  \n",
    "> DTPSYSTL - diastolic pressure  \n",
    "> DTPDIAST - systolic pressure  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Get continuous data...\n",
    "cont.1 <- read.table(file=\"phs000228.v3.pht002157.v1.p1.c1.EMMES_HMP_DTP_DHX_DVD.HMP.txt\", sep=\"\\t\",\n",
    "                     header=T, na.string=c(NA, \"\"))\n",
    "\n",
    "#some variables were taken at the intake appointment\n",
    "continuous <- cont.1[cont.1$VISNO==\"00\",c(\"DTPPULSE\",\"DTPBMI\",\"DTPSYSTL\",\"DTPDIAST\", \"RANDSID\")]\n",
    "rownames(continuous) <- continuous$RANDSID\n",
    "\n",
    "#define a continuous table variable\n",
    "continuous <- continuous[order(rownames(continuous)),]\n",
    "\n",
    "\n",
    "\n",
    "#vaginal pH was taken at each visit. we created a new file - vaginal_ph.txt - that contains\n",
    "#the two pH values for each woman at each visit:\n",
    "cont.1$VISNO <- gsub(\"^(\\\\d\\\\d).*\", \"\\\\1\", cont.1$VISNO)\n",
    "pH <- cont.1[cont.1$VISNO == \"01\" | cont.1$VISNO == \"02\" | cont.1$VISNO == \"03\",]\n",
    "pH <- pH[,c(\"RANDSID\", \"VISNO\", \"SITE\", \"DVDINTPH\", \"DVDPFPH\")]\n",
    "pH$VISNO <- as.numeric(gsub(\"^0\", \"\", pH$VISNO))\n",
    "pH <- pH[!is.na(pH$DVDINTPH) | !is.na(pH$DVDPFPH),]\n",
    "write.table(pH, file=\"vaginal_ph.txt\", quote=F, row.names=F, sep=\"\\t\")\n",
    "\n",
    "\n",
    "#let's add the subjects' ages\n",
    "cont.2 <- read.table(file=\"phs000228.v3.pht002158.v1.p1.c1.EMMES_HMP_DEM_ENR.HMP.txt\", sep=\"\\t\", header=T, na.string=c(NA, \"\"))\n",
    "rownames(cont.2) <- cont.2$RANDSID\n",
    "cont.2 <- cont.2[order(rownames(cont.2)),]\n",
    "rownames(cont.2) == rownames(continuous)  #make sure the rows are in the right order\n",
    "#The following was the only continuous data avaialble in this metadata file\n",
    "#AGEENR - Age in years\n",
    "AGEENR <- cont.2$AGEENR\n",
    "continuous <- cbind(continuous, AGEENR) \n",
    "\n",
    "#remove the RANDSID column from the table\n",
    "continuous <- continuous[,-which(colnames(continuous) %in% c(\"RANDSID\"))]\n",
    "\n",
    "#write continuous data out to a file\n",
    "write.table(continuous, \"continuous.metadata\", quote=F)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At this point all of the categorical and continuous data have been identified and stored to a file. We have enverything that was used to generate Extended Data Table 1:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "summary(categorical)\n",
    "\n",
    "#count bfed \"not remember\"\n",
    "bfed <- categ.3$DSUBFED_C\n",
    "bfed[bfed == \"\"] <- NA\n",
    "summary(factor(bfed))\n",
    "\n",
    "#count transients\n",
    "sum(apply(contra, 1, sum, na.rm=T) > 0)-sum(contra.chronic, na.rm=T); \n",
    "sum(apply(vit_min, 1, sum, na.rm=T) > 0)-sum(vit_min.chronic, na.rm=T); \n",
    "sum(apply(antidep, 1, sum, na.rm=T) > 0)-sum(antidep.chronic, na.rm=T); \n",
    "sum(apply(antihist, 1, sum, na.rm=T) > 0)-sum(antihist.chronic, na.rm=T); \n",
    "\n",
    "summary(continuous)\n",
    "\n",
    "pH.summary <- pH[,c(\"DVDINTPH\", \"DVDPFPH\")]\n",
    "apply(pH.summary, 2, summary)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's copy these metadata files to our analysis folder and head over there to do our analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "cp categorical.metadata continuous.metadata vaginal_ph.txt ../analysis/\n",
    "cd ../analysis/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis\n",
    "Now we have the sequence data and clinical metadata curated and into a form that we can handle. Let's do some analysis. We have several tasks ahead of us...  \n",
    "\n",
    "* Describe community types\n",
    "* Identify associations between body sites and clinical metadata\n",
    "* Identify associations between body sites\n",
    "* Quantify rate of change between community types for each body site\n",
    "* Effect of time on stability of community types\n",
    "* Validate DMM models as a community type approach relative to the PAM-based approach\n",
    "\n",
    "\n",
    "##Describe community types\n",
    "First we wanted to inspect the DMM model fits and see the top taxa in each community type. This analysis was performed using the data found in the dmm_best folder. Let's generate the types of plots shown in Figure 1A and 1B. First we'll make the line plot of the Laplace fit metric vs. the number of community types (e.g. Figure 1A). For fun we'll bold and color in red the x-axis label that corresponds to the minimum Laplace metric. These will be outputted as pdfs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "dmm_best_path <- \"../v35/dmm_best/\"\n",
    "fit.files <- c(\"Antecubital_fossa.tx.1.subsample.1.dmm.mix.fit\", \"Anterior_nares.tx.1.subsample.1.dmm.mix.fit\",\n",
    "\t\t\t\t\"Buccal_mucosa.tx.1.subsample.1.dmm.mix.fit\", \"Hard_palate.tx.1.subsample.1.dmm.mix.fit\",\n",
    "\t\t\t\t\"Keratinized_gingiva.tx.1.subsample.1.dmm.mix.fit\", \"Palatine_Tonsils.tx.1.subsample.1.dmm.mix.fit\",\n",
    "\t\t\t\t\"Retroauricular_crease.tx.1.subsample.1.dmm.mix.fit\", \"Saliva.tx.1.subsample.1.dmm.mix.fit\",\n",
    "\t\t\t\t\"Stool.tx.1.subsample.1.dmm.mix.fit\", \"Subgingival_plaque.tx.1.subsample.1.dmm.mix.fit\",\n",
    "\t\t\t\t\"Supragingival_plaque.tx.1.subsample.1.dmm.mix.fit\", \"Throat.tx.1.subsample.1.dmm.mix.fit\",\n",
    "\t\t\t\t\"Tongue_dorsum.tx.1.subsample.1.dmm.mix.fit\", \"Vagina.tx.1.subsample.1.dmm.mix.fit\")\n",
    "\n",
    "for(bs in fit.files){\n",
    "\tpdf.file <- paste(bs, \".pdf\", sep=\"\")\n",
    "\tfit <- read.table(file=paste(dmm_best_path, bs, sep=\"\"), header=T)\n",
    "\n",
    "\tpdf(file=pdf.file)\n",
    "\t\n",
    "\tbs <- gsub(\"([^.]*).*\", \"\\\\1\", bs)\n",
    "\tbs <- gsub(\"_\", \" \", bs)\n",
    "\t\n",
    "\tpar(mar=c(5, 5, 0.5, 0.5))\n",
    "\tplot(fit$Laplace~fit$K, xlab=\"Number of Dirichlet\\nComponents\", ylab=paste(\"Model Fit\\n(\",bs, \")\", sep=\"\"), type=\"l\")\n",
    "\tpoints(fit$Laplace~fit$K, pch=19)\n",
    "\t\n",
    "\taxis(1, at=which.min(fit$Laplace), label=which.min(fit$Laplace), col.axis=\"red\", font=2)\n",
    "\t\n",
    "\tdev.off()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's make the box plots that show the relative abundance for the top 5 genera in each community type for each body site (e.g. Figure 1B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "dmm_best_path <- \"../v35/dmm_best/\"\n",
    "\n",
    "sites <- c(\"Antecubital_fossa.tx.1.subsample.1.dmm.mix.design\", \"Anterior_nares.tx.1.subsample.1.dmm.mix.design\",\n",
    "\t\t\"Buccal_mucosa.tx.1.subsample.1.dmm.mix.design\", \"Hard_palate.tx.1.subsample.1.dmm.mix.design\",\n",
    "\t\t\"Keratinized_gingiva.tx.1.subsample.1.dmm.mix.design\", \"Palatine_Tonsils.tx.1.subsample.1.dmm.mix.design\",\n",
    "\t\t\"Retroauricular_crease.tx.1.subsample.1.dmm.mix.design\", \"Saliva.tx.1.subsample.1.dmm.mix.design\",\n",
    "\t\t\"Stool.tx.1.subsample.1.dmm.mix.design\", \"Subgingival_plaque.tx.1.subsample.1.dmm.mix.design\",\n",
    "\t\t\"Supragingival_plaque.tx.1.subsample.1.dmm.mix.design\", \"Throat.tx.1.subsample.1.dmm.mix.design\",\n",
    "\t\t\"Tongue_dorsum.tx.1.subsample.1.dmm.mix.design\", \"Vagina.tx.1.subsample.1.dmm.mix.design\")\n",
    "\n",
    "#need to get the last level that was given a classification for each phylotype\n",
    "taxonomy <- read.table(file=\"../v35/v35.unique.good.filter.unique.precluster.unique.pick.three.wang.pick.tx.1.cons.taxonomy\",\n",
    "                       header=T, row.names=1)\n",
    "taxonomy$Taxonomy <- gsub(\"\\\\(\\\\d*\\\\)\", \"\", taxonomy$Taxonomy)\n",
    "taxonomy$Taxonomy <- gsub(\"unclassified;\", \"\", taxonomy$Taxonomy)\n",
    "taxonomy$Taxonomy <- gsub(\";$\", \"\", taxonomy$Taxonomy)\n",
    "taxonomy$Taxonomy <- gsub(\".*;\", \"\", taxonomy$Taxonomy)\n",
    "taxonomy$Taxonomy <- gsub('\"', \"\", taxonomy$Taxonomy)\n",
    "\n",
    "for(bs in sites){\n",
    "\tdesign.fname <- bs #need the design file to know which sample belongs to each community type\n",
    "   \tdesign <- read.table(file=paste(dmm_best_path, design.fname, sep=\"\"))\n",
    "    \n",
    "\tshared.fname <- gsub(\"1.dmm.mix.design\", \"shared\", bs) #need the relative abundance data to know\n",
    "\tshared <- read.table(file=paste(\"../v35/\", shared.fname, sep=\"\"), header=T, row.names=2)#the abundance of each\n",
    "\tshared <- shared[,-c(1,2)]                         #OTU in each sample\n",
    "\n",
    "    summary.fname <- gsub(\"design\", \"summary\", bs) #need the summary file to know the top 5 OTUs\n",
    "\tsummary <- read.table(file=paste(dmm_best_path, summary.fname, sep=\"\"), header=T, row.names=1)#for defining each commmunity type\n",
    "    \n",
    "    site <- gsub(\"^([^.]*).*\", \"\\\\1\", design.fname)\n",
    "\tsite <- gsub(\"_\", \" \", site)\n",
    "\t\n",
    "\tntypes <- length(levels(design$V2)) #get the number of community types\n",
    "\ttop.otus <- summary[1:5,]\n",
    "\totu.ids <- rownames(top.otus) #get the names of the top 5 phylotypes\n",
    "\totu.tax <- taxonomy[otu.ids,\"Taxonomy\"] #get their taxonomic affiliation\n",
    "\n",
    "\totu.shared <- shared[,otu.ids]/1000\t#get the relative abundance\n",
    "\trownames(otu.shared) == design$V1\t#check they're in the right order\n",
    "\n",
    "\tpdf(file=paste(summary.fname, \".pdf\", sep=\"\"))  #let's save these as pdfs\n",
    "\tpar(mar=c(10, 4, 0.5, 0.5))\n",
    "\tplot(x=c(0,5*(ntypes+2)), y=c(0,1), type=\"n\", axes=F, xlab=\"\", ylab=\"Relative abundance (%)\")\n",
    "\n",
    "\twidth = 0.5\n",
    "\tleft <- 1\n",
    "\tclrs <- rainbow(ntypes)\n",
    "\n",
    "\tfor(i in otu.ids){\n",
    "        #calculate the intra quartile ranges\n",
    "\t\tiqr <- aggregate(otu.shared[,i], by=list(design$V2), quantile, probs=c(0.025, 0.25, 0.50, 0.75, 0.975))\n",
    "\t\n",
    "        #error bars represent the 95% confindence interval\n",
    "\t\tarrows(x0=left:(left+ntypes-1), x1=left:(left+ntypes-1), y0=iqr$x[,\"50%\"], y1=iqr$x[,\"97.5%\"], angle=90,\n",
    "               length=0.05, lwd=2)\n",
    "\t\tarrows(x0=left:(left+ntypes-1), x1=left:(left+ntypes-1), y0=iqr$x[,\"50%\"], y1=iqr$x[,\"2.5%\"], angle=90,\n",
    "               length=0.05, lwd=2)\n",
    "\n",
    "        #rectangles represent the 50% confidence interval with a line in the middle for the median\n",
    "        rect(xleft=(left:(left+ntypes-1))-width, xright=(left:(left+ntypes-1))+width, ytop=iqr$x[,\"75%\"],\n",
    "             ybottom=iqr$x[,\"50%\"], col=clrs, lwd=3)\n",
    "\t\trect(xleft=(left:(left+ntypes-1))-width, xright=(left:(left+ntypes-1))+width, ytop=iqr$x[,\"25%\"],\n",
    "             ybottom=iqr$x[,\"50%\"], col=clrs, lwd=3)\n",
    "\n",
    "        #bounce down the x-axis for the next phylotype\n",
    "\t\tleft <- left + ntypes + 2\n",
    "\t}\n",
    "\n",
    "\taxis(2, label=seq(0,100,20), at=seq(0,1,0.2), las=2)\n",
    "\tstart <- (ntypes+1) / 2\n",
    "\tincrement <- ntypes + 2\n",
    "\n",
    "    #label the phylotypes on the x-axis\n",
    "\taxis(1, label=otu.tax, at=c(start, start+increment, start+2*increment, start+3*increment, start+4*increment),\n",
    "         cex=0.5, las=2)\n",
    "\tbox()\n",
    "\n",
    "    #get a pretty legend\n",
    "\tlegend(\"topright\", legend=paste(\"Community Type\", LETTERS[1:ntypes]), fill=clrs, pt.cex=1.5, cex=1, bty=\"n\")\n",
    "\tmtext(1, line=8.5, text=site, font=2)  #tell us what the body site was\n",
    "\tdev.off()\n",
    "}\n",
    "#expect some warning messages: these indicate that the median for some phylotypes was zero and so\n",
    "#the bottom rectangle for the IQR is zero. whatever."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identify associations between body sites and clinical metadata\n",
    "### Categorical metadata\n",
    "In attempting to deal with the community type data we have a small difficulty that we have up to 3 community type assignments for each person at each body site but the clinical metadata did not change. Our strategy was to use an iterative procedure. Our basic method for each body site was the following. First, we randomly picked one visit from each subject and used Fisher's exact test to test for an association with each of the clinical variables. We corrected for multiple comparisons using the Benjamini-Hochberg algorithm. Second, we again randomly selected one visit from each subject and tested. We did a total of 1000 iterations and quantified the median P-value and the fraction of iterations that yielded a significant test. The output from this section is discussed in the manuscript and provdied in SourceDataFigure3.xlsx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "totalIters <- 1000\n",
    "\n",
    "#Set up the community type data - recall we generated these above to describe the community type\n",
    "#each subject at each visit\n",
    "v1<-read.table(file=\"community_types.v1.txt\", header=T, row.names=1)\n",
    "v2<-read.table(file=\"community_types.v2.txt\", header=T, row.names=1)\n",
    "v3<-read.table(file=\"community_types.v3.txt\", header=T, row.names=1)\n",
    "\n",
    "\n",
    "#get the collection of the ids for all 300 subjects\n",
    "randsids <- union(rownames(v1), union(rownames(v2), rownames(v3)))\n",
    "\n",
    "\n",
    "#Set up the categorical data\n",
    "categ <- read.table(file=\"categorical.metadata\", header=T)\n",
    "categ.good <- categ[randsids,]\t#only want those metadata that we have microbiome data for\n",
    "\n",
    "variables <- colnames(categ.good)\n",
    "bodysites <- colnames(v1)\n",
    "\n",
    "#this table will hold the fraction of iterations that a bodysite had a significant test\n",
    "sig.rate <- matrix(rep(0, length(variables)*length(bodysites)), nrow=length(variables))\n",
    "rownames(sig.rate) <- variables\n",
    "colnames(sig.rate) <- bodysites\n",
    "\n",
    "\n",
    "#this table will hold the median P-value across the iterations for each bodysite\n",
    "medians <- matrix(rep(0, length(variables)*length(bodysites)), nrow=length(variables))\n",
    "rownames(medians) <- variables\n",
    "colnames(medians) <- bodysites\n",
    "\n",
    "for(i in bodysites){    \n",
    "\tprint(i)\n",
    "\ttimes.sig <- rep(0, length(variables))  #set up the counting vector for the number of sig iters\n",
    "\tnames(times.sig) <- variables\n",
    "\ttotal <- 0\n",
    "\n",
    "\tsig <- rep(NA, length(variables))       #\n",
    "\tnames(sig) <- variables\n",
    "\n",
    "    #the record of the community type for each subject (rows) at each visit (column)\n",
    "\tvisits <- data.frame(matrix(rep(NA, length(randsids)*3), ncol=3))\n",
    "\tcolnames(visits) <- c(\"v1\", \"v2\", \"v3\")\n",
    "\trownames(visits) <- randsids\n",
    "\n",
    "\tvisits[rownames(v1), \"v1\"] <- as.character(v1[,i])\n",
    "\tvisits[rownames(v2), \"v2\"] <- as.character(v2[,i])\n",
    "\tvisits[rownames(v3), \"v3\"] <- as.character(v3[,i])\n",
    "\n",
    "    \n",
    "    #set up the table of p-values for each body site and iteration\n",
    "\tmedian.iter <- matrix(rep(NA, length(variables)*totalIters), nrow=length(variables))\n",
    "\trownames(median.iter) <- variables\n",
    "\tcolnames(median.iter) <- 1:totalIters\n",
    "    \n",
    "\tfor(iter in 1:totalIters){  #start looping over the iterations\n",
    "        iter <- 1\n",
    "\t\ttype <- as.character()\n",
    "\t\tfor(j in randsids){     #for each subject, randomly select the community type at one visit\n",
    "\t\t\ttype[j] <- sample(na.omit(as.character(visits[j,])))[1] #omits those visits that didn't happen\n",
    "\t\t}\n",
    "\n",
    "\t\tfor(j in variables){  #now loop through the categorical variables\n",
    "            contingency <- table(type, categ.good[,j]) #make a contingency table\n",
    "\n",
    "\t\t\tif(sum(apply(contingency, 2, sum)>10) > 1){ #if either value of the metadata was seen less than 10 times, move on\n",
    "\t\t\t\tif(ncol(contingency)>2 | nrow(contingency) > 2){ \t#if we have more than 2 communitytypes or variable levels\n",
    "\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t\t#we'll simulate the p-values\n",
    "\t\t\t\t\tsig[j] <- fisher.test(contingency, simulate.p.value=T, B=1e5)$p.value\n",
    "\t\t\t\t} else{\n",
    "\t\t\t\t\tsig[j] <- fisher.test(contingency)$p.value\n",
    "\t\t\t\t}\n",
    "\t\t\t} else{\n",
    "\t\t\t\tsig[j] <- NA\n",
    "\t\t\t}\t\t\n",
    "\t\t}\n",
    "\n",
    "\t\tadjusted <- p.adjust(sig, method=\"BH\")\n",
    "\t\tcorrect <- adjusted < 0.05\n",
    "\n",
    "\t\ttimes.sig <- correct + times.sig\n",
    "\t\ttotal <- total + 1\n",
    "\t\tmedian.iter[, iter] <- sig\n",
    "\t}\n",
    "\tsig.rate[,i] <- times.sig / total\t#calculate the fraction of significant tests\n",
    "\tmedians[,i] <- apply(median.iter, 1, median)\t#calculate the median p-value\n",
    "}\n",
    "\n",
    "write.table(sig.rate, file=\"categorical.sig.rate\", quote=F, sep=\"\\t\")\n",
    "write.table(medians, file=\"categorical.medians\", quote=F, sep=\"\\t\")\n",
    "\n",
    "interesting <- apply(sig.rate, 1, max, na.rm=T) > 0.5  #cut through the data to see which data had any signficance\n",
    "sig.rate[interesting,]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the following cell we see that there are 9 bodysite/metadata combinations that were significant in more than 75% of the iterations. We inadvertantly left out the observation that there is a significant association between Asians and their community type. We go ahead and make bar plots representing the contingency tables (e.g. Figure 1C and D and Extended Figures 2 and 3) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "threshold <- 0.75   #what's the minimum fraction of iterations that need to be significant\n",
    "totalIters <- 100   #number of iterations to calculate the average contingency table for\n",
    "                    #we dont need a lot of precision here\n",
    "\n",
    "#we read back in the categorical significance rates but will ignore the SITE data becuase\n",
    "#of the confounding described above and chronic Antihistamine usage since all the tests\n",
    "#resuled in NAs.\n",
    "sig.rate <- read.table(file=\"categorical.sig.rate\", header=T)\n",
    "sig.rate <- sig.rate[-which(rownames(sig.rate) %in% c(\"SITE\", \"antihist.chronic\")),]\n",
    "\n",
    "#Let's get the most frequently significant bodysites and metadata\n",
    "row.max <- apply(sig.rate, 1, max, na.rm=T) > threshold\n",
    "col.max <- apply(sig.rate, 2, max, na.rm=T) > threshold\n",
    "most.sig <- sig.rate[row.max, col.max]\n",
    "#there were 8 combinations\n",
    "\n",
    "\n",
    "#we're essentially going to recreate the radonomized contingency tables like we did to \n",
    "#run the Fisher's exact test\n",
    "v1<-read.table(file=\"community_types.v1.txt\", header=T, row.names=1)\n",
    "v2<-read.table(file=\"community_types.v2.txt\", header=T, row.names=1)\n",
    "v3<-read.table(file=\"community_types.v3.txt\", header=T, row.names=1)\n",
    "good.ids <- unique(c(rownames(v1), rownames(v2), rownames(v3)))\n",
    "\n",
    "categ <- read.table(file=\"categorical.metadata\", header=T)\n",
    "categ.good <- categ[good.ids,]\n",
    "\n",
    "for(md in rownames(most.sig)){      #scan over the metadata in most.sig\n",
    "\tfor(bs in colnames(most.sig)){  #and scan over the bodysites in most.sig\n",
    "\n",
    "\t\tif(!is.na(sig.rate[md, bs] ) & sig.rate[md, bs] > threshold){#if it isn't an NA and\n",
    "                                                                     #is over our threshold...\n",
    "            #let's set up a visits matrix like before\n",
    "            visits <- data.frame(matrix(rep(NA, length(good.ids)*3), ncol=3))\n",
    "\t\t\tcolnames(visits) <- c(\"v1\", \"v2\", \"v3\")\n",
    "\t\t\trownames(visits) <- good.ids\n",
    "\t\t\tvisits[rownames(v1), \"v1\"] <- as.character(v1[,bs])\n",
    "\t\t\tvisits[rownames(v2), \"v2\"] <- as.character(v2[,bs])\n",
    "\t\t\tvisits[rownames(v3), \"v3\"] <- as.character(v3[,bs])\n",
    "\t\t\t\n",
    "            #get the possible community types that were observed at the bodysite \n",
    "\t\t\ttypes <- levels(factor(c(visits[,1], visits[,2], visits[,3])))\n",
    "\t\t\tntypes <- length(types)\n",
    "\t\t\t\n",
    "            #get the possible values for the metadata\n",
    "\t\t\tlevels <- levels(as.factor(categ.good[,md]))\n",
    "\t\t\tnlevels <- length(levels)\n",
    "            \n",
    "            #initialize the contingency table\n",
    "\t\t\tcontingency <- as.table(matrix(rep(0, ntypes*nlevels), nrow=ntypes))\n",
    "\t\t\trownames(contingency) <- types\n",
    "\t\t\tcolnames(contingency) <- levels\n",
    "\t\t\t\n",
    "            #let it rip...\n",
    "\t\t\tfor(iter in 1:totalIters){\n",
    "\t\t\t\ttype <- as.character()\n",
    "\t\t\t\tfor(k in good.ids){\t#four nested for loops!\n",
    "\t\t\t\t\ttype[k] <- sample(na.omit(as.character(visits[k,])))[1]\n",
    "\t\t\t\t}\n",
    "\t\t\t\ttype <- factor(type)\n",
    "\t\t\t\tcontingency <- contingency + table(type, categ.good[,md])\n",
    "\t\t\t}\n",
    "            \n",
    "            #add up the contingency tables\n",
    "\t\t\tcontingency <- contingency / totalIters    #get the proportions of each community type\n",
    "\t\t\tp.contingency <- prop.table(contingency, 2)#in each metadata class\n",
    "\n",
    "            #make a pretty barplot and save it as a pdf\n",
    "\t\t\tpdf(file=paste(bs, \".\", md, \".pdf\", sep=\"\")) \n",
    "\t\t\tpar(mar=c(6,5, 0.5,0.5))\n",
    "\t\t\tbarplot(as.matrix(t(p.contingency)), beside=T, yaxt=\"n\", names=LETTERS[1:nrow(p.contingency)], \n",
    "                    legend.text=colnames(p.contingency), args.legend=c(cex=1.2), ylim=c(0,max(p.contingency)+0.05))\n",
    "\n",
    "\t\t\taxis(2, at=seq(0,max(p.contingency+0.05),0.10), label=seq(0,max(p.contingency+0.05)*100, 10), las=1)\n",
    "\n",
    "\t\t\tmtext(side=1, line=2.5, text=paste(gsub(\"_\", \" \", bs), \"Community Type\"), cex=1.2)\n",
    "\t\t\tmtext(side=2, line=2.5, text=\"Percentage of Samples\", cex=1.2)\n",
    "\t\t\tmtext(side=1, line=4, text=md)\n",
    "\t\t\tbox()\n",
    "\t\t\tdev.off()\n",
    "\t\t}\n",
    "\t\t\n",
    "\t}\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "totalIters <- 1000\n",
    "\n",
    "#Set up the community type data\n",
    "v1<-read.table(file=\"community_types.v1.txt\", header=T, row.names=1)\n",
    "v2<-read.table(file=\"community_types.v2.txt\", header=T, row.names=1)\n",
    "v3<-read.table(file=\"community_types.v3.txt\", header=T, row.names=1)\n",
    "\n",
    "randsids <- union(rownames(v1), union(rownames(v2), rownames(v3)))\n",
    "\n",
    "\n",
    "#Set up the continuous data\n",
    "cont <- read.table(file=\"continuous.metadata\", header=T, na.string=c(\"\", NA))\n",
    "\n",
    "#recall that we need to bring in the pH values from each visit - let's just initialize it for now\n",
    "cont <- cbind(cont, \"DVDINTPH\"=rep(NA, nrow(cont)), \"DVDPFPH\"=rep(NA, nrow(cont)))\n",
    "\n",
    "cont.good <- cont[randsids,]\n",
    "variables <- colnames(cont.good)[2:ncol(cont.good)]\n",
    "bodysites <- colnames(v1)\n",
    "\n",
    "#now we'll get that pH data\n",
    "ph.file <- read.table(file=\"vaginal_ph.txt\", header=T)\n",
    "\n",
    "#get the subject ids that we have pH data for\n",
    "v1.names <- as.character(ph.file[ph.file$VISNO==1,1])\n",
    "v2.names <- as.character(ph.file[ph.file$VISNO==2,1])\n",
    "v3.names <- as.character(ph.file[ph.file$VISNO==3,1])\n",
    "ph.names <- union(v1.names, union(v2.names, v3.names)) #get the unique list\n",
    "\n",
    "#let's make a table of vaginal introitus pH values for each woman at each visit\n",
    "ph.int <- data.frame(\"visit1\"=rep(NA, length(ph.names)), \"visit2\"=rep(NA, length(ph.names)),\n",
    "                     \"visit3\"=rep(NA, length(ph.names)), row.names=ph.names)\n",
    "ph.int[v1.names, \"visit1\"] <- ph.file[ph.file$VISNO==1, \"DVDINTPH\"]\n",
    "ph.int[v2.names, \"visit2\"] <- ph.file[ph.file$VISNO==2, \"DVDINTPH\"]\n",
    "ph.int[v3.names, \"visit3\"] <- ph.file[ph.file$VISNO==3, \"DVDINTPH\"]\n",
    "\n",
    "#let's make a table of posterior fornix pH values for each woman at each visit\n",
    "ph.pf <- data.frame(\"visit1\"=rep(NA, length(ph.names)), \"visit2\"=rep(NA, length(ph.names)),\n",
    "                    \"visit3\"=rep(NA, length(ph.names)), row.names=ph.names)\n",
    "ph.pf[v1.names, \"visit1\"] <- ph.file[ph.file$VISNO==1, \"DVDPFPH\"]\n",
    "ph.pf[v2.names, \"visit2\"] <- ph.file[ph.file$VISNO==2, \"DVDPFPH\"]\n",
    "ph.pf[v3.names, \"visit3\"] <- ph.file[ph.file$VISNO==3, \"DVDPFPH\"]\n",
    "\n",
    "\n",
    "#set up a table with the frequency of significant tests for each variable / bodysite combinatino\n",
    "sig.rate <- matrix(rep(0, (length(variables))*length(bodysites)), nrow=(length(variables)))\n",
    "rownames(sig.rate) <- c(variables)\n",
    "colnames(sig.rate) <- bodysites\n",
    "\n",
    "#set up a table with the median p-value for each variable / bodysite combinatino\n",
    "medians <- matrix(rep(0, (length(variables))*length(bodysites)), nrow=(length(variables)))\n",
    "rownames(medians) <- c(variables)\n",
    "colnames(medians) <- bodysites\n",
    "\n",
    "\n",
    "for(i in bodysites){\n",
    "\tprint(i)\n",
    "\ttimes.sig <- rep(0, length(variables))\n",
    "\tnames(times.sig) <- variables\n",
    "\ttotal <- 0\n",
    "\n",
    "\tsig <- rep(NA, length(variables))\n",
    "\tnames(sig) <- variables\n",
    "\n",
    "\tvisits <- data.frame(matrix(rep(NA, length(randsids)*3), ncol=3))\n",
    "\tcolnames(visits) <- c(\"v1\", \"v2\", \"v3\")\n",
    "\trownames(visits) <- randsids\n",
    "\n",
    "\tvisits[rownames(v1), \"v1\"] <- as.character(v1[,i])\n",
    "\tvisits[rownames(v2), \"v2\"] <- as.character(v2[,i])\n",
    "\tvisits[rownames(v3), \"v3\"] <- as.character(v3[,i])\n",
    "\n",
    "\tmedian.iter <- matrix(rep(NA, length(variables)*totalIters), nrow=length(variables))\n",
    "\trownames(median.iter) <- variables\n",
    "\t\n",
    "\tfor(iter in 1:totalIters){\n",
    "\t\ttype <- character()\n",
    "\n",
    "\t\tfor(j in randsids){\n",
    "\t\t\tx<-1:3\n",
    "\t\t\t\n",
    "\t\t\trand.vis <- sample(x)\n",
    "\t\t\tif(!is.na(visits[j,rand.vis[1]])){\n",
    "\t\t\t\trand.vis <- rand.vis[1]\n",
    "\t\t\t} else if(!is.na(visits[j,rand.vis[2]])){\n",
    "\t\t\t\trand.vis <- rand.vis[2]\n",
    "\t\t\t} else {\n",
    "\t\t\t\trand.vis <- rand.vis[3]\n",
    "\t\t\t}\t\t\n",
    "\t\t\ttype[j] <- visits[j,rand.vis]\n",
    "\t\t\tcont.good[j, \"DVDINTPH\"] <- ph.int[j, rand.vis]#fill in the pH values for that visit\n",
    "\t\t\tcont.good[j, \"DVDPFPH\"] <- ph.pf[j, rand.vis] \n",
    "\t\t}\n",
    "\n",
    "\t\tfor(j in variables){  #do the anova\n",
    "\t\t\tsig[j] <- anova(lm(cont.good[,j]~factor(type)))$Pr[1]\n",
    "\t\t}\n",
    "\n",
    "\t\tadjusted <- p.adjust(sig, method=\"BH\") #correct for multiple comparisons\n",
    "\t\tcorrect <- adjusted < 0.05  #significant?\n",
    "\n",
    "\t\ttimes.sig <- correct + times.sig #add 1 if significant\n",
    "\t\ttotal <- total + 1\n",
    "\t\tmedian.iter[, iter] <- sig #keep track of p-values\n",
    "\t}\n",
    "\tsig.rate[,i] <- times.sig / total #get rate of significance\n",
    "\tmedians[,i] <- apply(median.iter, 1, median) #get median p-values\n",
    "}\n",
    "\n",
    "write.table(sig.rate, file=\"continuous.sig.rate\", quote=F, sep=\"\\t\")\n",
    "write.table(medians, file=\"continuous.medians\", quote=F, sep=\"\\t\")\n",
    "\n",
    "interesting <- apply(sig.rate, 1, max, na.rm=T) > 0.5  #cut through the data to see which data had any signficance\n",
    "sig.rate[interesting,]  #maybe not all that interesting afterall\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There were weak associations between vaginal pH and the vaginal community types"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Continuous metadata\n",
    "Now, let's use a similar procedure to look at the coninuous data. Instead of using Fisher's exact test, we'll use ANOVA and we'll use the pH values that were obtained from that day's sampling."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "rm(list=ls())\n",
    "#set up the community type data\n",
    "#first we get the subject ids\n",
    "categ <- read.table(file=\"categorical.metadata\", header=T, row.names=1)\n",
    "randsids <- rownames(categ)\n",
    "\n",
    "#next we read in the community type assignments\n",
    "v1<-read.table(file=\"community_types.v1.txt\", header=T, row.names=1)\n",
    "v2<-read.table(file=\"community_types.v2.txt\", header=T, row.names=1)\n",
    "v3<-read.table(file=\"community_types.v3.txt\", header=T, row.names=1)\n",
    "\n",
    "#now we build the visit tables for the three vaginal sites - intialize\n",
    "vi.types <- matrix(rep(NA, length(randsids)*3), nrow=length(randsids))\n",
    "rownames(vi.types) <- randsids\n",
    "colnames(vi.types) <- c(\"visit1\", \"visit2\", \"visit3\")\n",
    "mv.types <- vi.types\n",
    "pf.types <- vi.types\n",
    "pf.ph <- vi.types\n",
    "int.ph <- vi.types\n",
    "\n",
    "#now we build the visit tables for the three vaginal sites - fill\n",
    "vi.types[rownames(v1), \"visit1\"] <- as.character(v1[,\"Vaginal_introitus\"])\n",
    "vi.types[rownames(v2), \"visit2\"] <- as.character(v2[,\"Vaginal_introitus\"])\n",
    "vi.types[rownames(v3), \"visit3\"] <- as.character(v3[,\"Vaginal_introitus\"])\n",
    "\n",
    "mv.types[rownames(v1), \"visit1\"] <- as.character(v1[,\"Mid_vagina\"])\n",
    "mv.types[rownames(v2), \"visit2\"] <- as.character(v2[,\"Mid_vagina\"])\n",
    "mv.types[rownames(v3), \"visit3\"] <- as.character(v3[,\"Mid_vagina\"])\n",
    "\n",
    "pf.types[rownames(v1), \"visit1\"] <- as.character(v1[,\"Posterior_fornix\"])\n",
    "pf.types[rownames(v2), \"visit2\"] <- as.character(v2[,\"Posterior_fornix\"])\n",
    "pf.types[rownames(v3), \"visit3\"] <- as.character(v3[,\"Posterior_fornix\"])\n",
    "\n",
    "\n",
    "#now we'll get that pH data \n",
    "ph.file <- read.table(file=\"vaginal_ph.txt\", header=T)\n",
    "ph1 <- ph.file[ph.file$VISNO==1, c(\"RANDSID\", \"DVDINTPH\", \"DVDPFPH\")]\n",
    "ph2 <- ph.file[ph.file$VISNO==2, c(\"RANDSID\", \"DVDINTPH\", \"DVDPFPH\")]\n",
    "ph3 <- ph.file[ph.file$VISNO==3, c(\"RANDSID\", \"DVDINTPH\", \"DVDPFPH\")]\n",
    "\n",
    "rownames(ph1) <- ph1$RANDSID; ph1 <- ph1[,-1]\n",
    "rownames(ph2) <- ph2$RANDSID; ph2 <- ph2[,-1]\n",
    "rownames(ph3) <- ph3$RANDSID; ph3 <- ph3[,-1]\n",
    "\n",
    "#introitus pH\n",
    "int.ph[rownames(ph1),\"visit1\"] <- ph1[, \"DVDINTPH\"]\n",
    "int.ph[rownames(ph2),\"visit2\"] <- ph2[, \"DVDINTPH\"]\n",
    "int.ph[rownames(ph3),\"visit3\"] <- ph3[, \"DVDINTPH\"]\n",
    "\n",
    "#posterior pH\n",
    "pf.ph[rownames(ph1),\"visit1\"] <- ph1[, \"DVDPFPH\"]\n",
    "pf.ph[rownames(ph2),\"visit2\"] <- ph2[, \"DVDPFPH\"]\n",
    "pf.ph[rownames(ph3),\"visit3\"] <- ph3[, \"DVDPFPH\"]\n",
    "\n",
    "#concatenate the commmunity type data from the three visits into a single vector\n",
    "vi <- c(vi.types[,\"visit1\"], vi.types[,\"visit2\"], vi.types[,\"visit3\"])\n",
    "mv <- c(mv.types[,\"visit1\"], mv.types[,\"visit2\"], mv.types[,\"visit3\"])\n",
    "pf <- c(pf.types[,\"visit1\"], pf.types[,\"visit2\"], pf.types[,\"visit3\"])\n",
    "\n",
    "#concatenate the pH data from the three visits into a single vector\n",
    "intph <- c(int.ph[,\"visit1\"], int.ph[,\"visit2\"], int.ph[,\"visit3\"])\n",
    "pfph <- c(pf.ph[,\"visit1\"], pf.ph[,\"visit2\"], pf.ph[,\"visit3\"])\n",
    "\n",
    "#build the stripcharts...\n",
    "pdf(file=\"Vagina_introitus.DVDINTPH.pdf\")\n",
    "\tpar(mar=c(5,4,0.5,0.5))\n",
    "\tgood <- !is.na(vi) & !is.na(intph)\n",
    "\tstripchart(intph[good]~factor(vi[good]), pch=19, vertical=T, method=\"jitter\", xlab=\"\", ylab=\"\",\n",
    "                               ylim=c(3,7), yaxt=\"n\", xaxt=\"n\", col=\"red\")\n",
    "\ttitle(ylab=\"Vaginal Introitus pH\")\n",
    "\tmtext(1, line=3, text=\"Vagina Introitus Community Type\")\n",
    "\tave <- aggregate(intph[good], by=list(factor(vi[good])), mean)\n",
    "\tsegments(x0=seq(0.8,4.8,1), x1=seq(1.2,5.2, 1), y0=ave$x, lwd=4)\n",
    "\taxis(1, at=1:5, label=paste(LETTERS[1:5]))\n",
    "\taxis(2, las=2)\n",
    "dev.off()\n",
    "\n",
    "pdf(file=\"Vagina_introitus.DVDPFPH.pdf\")\n",
    "\tpar(mar=c(5,4,0.5,0.5))\n",
    "\tgood <- !is.na(vi) & !is.na(pfph)\n",
    "\tstripchart(pfph[good]~factor(vi[good]), pch=19, vertical=T, method=\"jitter\", xlab=\"\", ylab=\"\",\n",
    "                               ylim=c(3,7), yaxt=\"n\", xaxt=\"n\", col=\"red\")\n",
    "\ttitle(ylab=\"Posterior Fornix pH\")\n",
    "\tmtext(1, line=3, text=\"Vagina Introitus Community Type\")\n",
    "\tave <- aggregate(pfph[good], by=list(factor(vi[good])), mean)\n",
    "\tsegments(x0=seq(0.8,4.8,1), x1=seq(1.2,5.2, 1), y0=ave$x, lwd=4)\n",
    "\taxis(1, at=1:5, label=paste(LETTERS[1:5]))\n",
    "\taxis(2, las=2)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "pdf(file=\"Mid_vagina.DVDINTPH.pdf\")\n",
    "\tpar(mar=c(5,4,0.5,0.5))\n",
    "\tgood <- !is.na(mv) & !is.na(intph)\n",
    "\tstripchart(intph[good]~factor(mv[good]), pch=19, vertical=T, method=\"jitter\", xlab=\"\", ylab=\"\",\n",
    "                               ylim=c(3,7), yaxt=\"n\", xaxt=\"n\", col=\"red\")\n",
    "\ttitle(ylab=\"Vaginal Introitus pH\")\n",
    "\tmtext(1, line=3, text=\"Mid Vagina Community Type\")\n",
    "\tave <- aggregate(intph[good], by=list(factor(mv[good])), mean)\n",
    "\tsegments(x0=seq(0.8,4.8,1), x1=seq(1.2,5.2, 1), y0=ave$x, lwd=4)\n",
    "\taxis(1, at=1:5, label=paste(LETTERS[1:5]))\n",
    "\taxis(2, las=2)\n",
    "dev.off()\n",
    "\n",
    "pdf(file=\"Mid_vagina.DVDPFPH.pdf\")\n",
    "\tpar(mar=c(5,4,0.5,0.5))\n",
    "\tgood <- !is.na(mv) & !is.na(pfph)\n",
    "\tstripchart(pfph[good]~factor(mv[good]), pch=19, vertical=T, method=\"jitter\", xlab=\"\", ylab=\"\",\n",
    "                               ylim=c(3,7), yaxt=\"n\", xaxt=\"n\", col=\"red\")\n",
    "\ttitle(ylab=\"Posterior Fornix pH\")\n",
    "\tmtext(1, line=3, text=\"Mid Vagina Community Type\")\n",
    "\tave <- aggregate(pfph[good], by=list(factor(mv[good])), mean)\n",
    "\tsegments(x0=seq(0.8,4.8,1), x1=seq(1.2,5.2, 1), y0=ave$x, lwd=4)\n",
    "\taxis(1, at=1:5, label=paste(LETTERS[1:5]))\n",
    "\taxis(2, las=2)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "pdf(file=\"Posterior_fornix.DVDINTPH.pdf\")\n",
    "\tpar(mar=c(5,4,0.5,0.5))\n",
    "\tgood <- !is.na(pf) & !is.na(intph)\n",
    "\tstripchart(intph[good]~factor(pf[good]), pch=19, vertical=T, method=\"jitter\", xlab=\"\", ylab=\"\",\n",
    "                               ylim=c(3,7), yaxt=\"n\", xaxt=\"n\", col=\"red\")\n",
    "\ttitle(ylab=\"Vaginal Introitus pH\")\n",
    "\tmtext(1, line=3, text=\"Posterior Fornix Community Type\")\n",
    "\tave <- aggregate(intph[good], by=list(factor(pf[good])), mean)\n",
    "\tsegments(x0=seq(0.8,4.8,1), x1=seq(1.2,5.2, 1), y0=ave$x, lwd=4)\n",
    "\taxis(1, at=1:5, label=paste(LETTERS[1:5]))\n",
    "\taxis(2, las=2)\n",
    "dev.off()\n",
    "\n",
    "pdf(file=\"Posterior_fornix.DVDPFPH.pdf\")\n",
    "\tpar(mar=c(5,4,0.5,0.5))\n",
    "\tgood <- !is.na(pf) & !is.na(pfph)\n",
    "\tstripchart(pfph[good]~factor(pf[good]), pch=19, vertical=T, method=\"jitter\", xlab=\"\", ylab=\"\",\n",
    "                               ylim=c(3,7), yaxt=\"n\", xaxt=\"n\", col=\"red\")\n",
    "\ttitle(ylab=\"Posterior Fornix pH\")\n",
    "\tmtext(1, line=3, text=\"Posterior Fornix Community Type\")\n",
    "\tave <- aggregate(pfph[good], by=list(factor(pf[good])), mean)\n",
    "\tsegments(x0=seq(0.8,4.8,1), x1=seq(1.2,5.2, 1), y0=ave$x, lwd=4)\n",
    "\taxis(1, at=1:5, label=paste(LETTERS[1:5]))\n",
    "\taxis(2, las=2)\n",
    "dev.off()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The SITE variable\n",
    "The variable that was most frequently detected as being sigificantly associated with the various community types was SITE - the city that the subject was recruited from - Houston and St. Louis. This was also seen in the original Nature paper from 2012. As described way back at the beginning, the SITE varaible was perfectly confounded with where the DNA extraction, PCR and sequencing were done. Although it's entirely possible that people in Houston and St Louis have very different microbiomes, we wouldn't want to rest on this dataset for that comparison. Let's see whether any of our other metadata are associated with the SITE variable. Let's start with the categorical data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "categ <- read.table(file=\"categorical.metadata\", header=T)\n",
    "site <- categ$SITE\n",
    "categ <- categ[,-1]\n",
    "\n",
    "p.values <- rep(0, ncol(categ))\n",
    "names(p.values) <- colnames(categ)\n",
    "\n",
    "for(md in colnames(categ)){\n",
    "\tp.values[md] <- fisher.test(site, categ[,md])$p.value\n",
    "}\n",
    "\n",
    "p.adjust(sort(p.values), method=\"BH\")\n",
    "p.adjust(sort(p.values), method=\"BH\") < 0.05\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What we glean from this is that their mom and dad's country of origin, whether they were Asian, white, or hispanic, whether they chronically use vitamins and mineral supplements, whether they had dental insurance, and their education level were significantly associated with the city of origin. This may cause us to not over interpret the result relating the vaginal community type to education level; however, considering that association was among the strongest observed, it is hard to discount.  \n",
    "  \n",
    "Now we want to turn to the continuous metadata:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cont <- read.table(file=\"continuous.metadata\", header=T)\n",
    "rownames(cont) == rownames(categ)\n",
    "\n",
    "p.values <- rep(0, ncol(cont))\n",
    "names(p.values) <- colnames(cont)\n",
    "\n",
    "for(md in colnames(cont)){\n",
    "\tp.values[md] <- anova(lm(cont[,md]~factor(site)))$Pr[1]\n",
    "}\n",
    "\n",
    "p.adjust(p.values, method=\"BH\")\n",
    "p.adjust(p.values, method=\"BH\") < 0.05\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For some reason the diastolic blood pressure was significantly different between both cities. None of these variable were significantly associated with our community types so we'll move on."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identify associations between body sites\n",
    "Next we wanted to know whether the community type at one bodysite was more likely to be associated with the community type at a different bodysite. While we expected things like the left and right antecubital fossa to be associated, we were also interested in associations from disparate types of bodysites. Here we again used a method very similar to what we did for associating communitytypes with categorical metadata. An important difference is that instead of picking a random visit for each bodysite, we picked a random visit for each person and used all of the bodysite data for that person from that visit. We then iterated a bunch. These data became the basis for the heatmap in Figure 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "totalIters <- 1000\n",
    "\n",
    "#Set up the community type data\n",
    "v1<-read.table(file=\"community_types.v1.txt\", header=T, row.names=1)\n",
    "v2<-read.table(file=\"community_types.v2.txt\", header=T, row.names=1)\n",
    "v3<-read.table(file=\"community_types.v3.txt\", header=T, row.names=1)\n",
    "\n",
    "randsids <- unique(c(rownames(v1), rownames(v2), rownames(v3)))\n",
    "bodysites <- colnames(v1)\n",
    "\n",
    "#make a table to keep track of when a subject visited. we will want\n",
    "#to take a person's entire microbiome, not just individual sites\n",
    "visits <- matrix(as.numeric(rep(NA, length(randsids)*3)), ncol=3)\n",
    "rownames(visits) <- randsids\n",
    "colnames(visits) <- c(\"v1\", \"v2\", \"v3\")\n",
    "visits[rownames(v1),1] <- 1\n",
    "visits[rownames(v2),2] <- 2\n",
    "visits[rownames(v3),3] <- 3\n",
    "\n",
    "\n",
    "total <- 0\n",
    "times.sig <- matrix(rep(0, 18*18), ncol=18)\n",
    "rownames(times.sig) <- bodysites\n",
    "colnames(times.sig) <- bodysites\n",
    "\n",
    "median.iter <- array(NA, dim=c(18,18,totalIters))\n",
    "\n",
    "for(iter in 1:totalIters){  #this is going to take awhile (triple loop!)\n",
    "\tif(iter %% 10 == 0){\n",
    "\t\tprint(iter)\n",
    "\t}\n",
    "\t\n",
    "\t#create a random visit across all individuals\n",
    "\trandom.visit <- matrix(ncol=18, nrow=0)\n",
    "\tcolnames(random.visit) <- bodysites\n",
    "\n",
    "\tfor(i in randsids){#for each individual, randomly pick one of their visits...\n",
    "\t\tvisit <- as.numeric(sample(na.omit(as.character(visits[i,])))[1])\n",
    "\t\n",
    "\t\t#and once we've got that random visit, take all of their bodysite\n",
    "\t\t#data for that visit\n",
    "\t\tif(visit == 1){\n",
    "\t\t\trandom.visit <- rbind(random.visit, v1[i,])\n",
    "\t\t} else if(visit ==2){\n",
    "\t\t\trandom.visit <- rbind(random.visit, v2[i,])\n",
    "\t\t} else if(visit == 3){\n",
    "\t\t\trandom.visit <- rbind(random.visit, v3[i,])\n",
    "\t\t} else {\n",
    "\t\t\trandom.visit <- rbind(random.visit, rep(NA, length(bodysites)))\n",
    "\t\t}\n",
    "\t}\n",
    "\n",
    "\t#let's build a p-value matrix for each bodysite vs. each bodysite\n",
    "\tp.values <- matrix(rep(NA, 18*18), ncol=18)\n",
    "\t\n",
    "\tfor(i in 2:18){\n",
    "\t\tfor(j in 1:i){\n",
    "\t\t\tif(j<i){ #no need to do everything twice...\n",
    "\t\t\t\tcontingency <- table(random.visit[,i], random.visit[,j])\n",
    "\t\t\t\tnrows <- sum(apply(contingency, 1, sum)>0)\n",
    "\t\t\t\tncols <- sum(apply(contingency, 2, sum)>0)\n",
    "\t\t\t\t\n",
    "\t\t\t\tif(nrows == 1 | ncols == 1){#not sure this would happen, but...\n",
    "\t\t\t\t\tp.values[i,j] <- NA\n",
    "\t\t\t\t}\n",
    "\t\t\t\telse if(nrows > 2 | ncols > 2){ #if it's more than 2x2, bootstrap\n",
    "\t\t\t\t\tp.values[i,j] <- fisher.test(contingency, simulate.p.value=T, B=1e5)$p.value\n",
    "\t\t\t\t} else{\t#but if it's 2x2 just do the normal method\n",
    "\t\t\t\t\tp.values[i,j] <- fisher.test(contingency)$p.value\n",
    "\t\t\t\t}\n",
    "\t\t\t}\t\t\n",
    "\t\t}\n",
    "\t}\n",
    "\n",
    "\tadjusted <- p.adjust(p.values, method=\"BH\")\t#adjust the p.values\n",
    "\tcorrect <- adjusted < 0.05\t#significant?\n",
    "\n",
    "\ttimes.sig <- correct + times.sig #increment if significant\n",
    "\ttotal <- total + 1\t\t\t\t\t#increment number of times through\n",
    "\tmedian.iter[, , iter] <- p.values\t#keep track of p.values\n",
    "}\n",
    "\n",
    "sig.rate <- times.sig / total\t\t#get the fraction significant\n",
    "sig.rate <- as.matrix(as.dist(sig.rate, upper=T))#make the matrix symmetric\n",
    "diag(sig.rate) <- 1\t\t\t\t\t\t\t\t#make the diagonal 1\n",
    "rownames(sig.rate) <- bodysites\t#clean things up a bit\n",
    "colnames(sig.rate) <- bodysites #clean thigns up a bit\n",
    "write.table(sig.rate, file=\"body_site.sig.rate\", quote=F, sep=\"\\t\")#output!\n",
    "\n",
    "#ditto on the median p-values\n",
    "medians <- as.matrix(as.dist(apply(median.iter, c(1,2), median)))\n",
    "diag(medians) <- 1e-6\n",
    "rownames(medians) <- bodysites\n",
    "colnames(medians) <- bodysites\n",
    "write.table(medians, file=\"body_site.medians\", quote=F, sep=\"\\t\")\n",
    "\n",
    "\n",
    "#now to build the heatmap of p-values...\n",
    "bsite_ordered <- c(\t\t#want to sort the bodysites to make sense\n",
    "\t4,\t#\"Keratinized_gingiva\",\n",
    "\t15,\t#\"Supragingival_plaque\",\n",
    "\t14,\t#\"Subgingival_plaque\",\n",
    "\t2,\t#\"Buccal_mucosa\",\n",
    "\t3,\t#\"Hard_palate\",\n",
    "\t17,\t#\"Tongue_dorsum\",\n",
    "\t16,\t#\"Saliva\",\n",
    "\t8,\t#\"Palatine_tonsils\",\n",
    "\t16,\t#\"Throat\",\n",
    "\t13,\t#\"Stool\",\n",
    "\t1,\t#\"Anterior_nares\",\n",
    "\t5,\t#\"L_Antecubital_fossa\",\n",
    "\t10,\t#\"R_Antecubital_fossa\",\n",
    "\t6,\t#\"L_Retroauricular_crease\",\n",
    "\t11,\t#\"R_Retroauricular_crease\",\n",
    "\t18,\t#\"Vaginal_introitus\",\n",
    "\t7,\t#\"Mid_vagina\",\n",
    "\t9\t#\"Posterior_fornix\"\n",
    ")\n",
    "\n",
    "#scale the values between 1e-5 and 1\n",
    "scaled.values <- 7 - ceiling(as.matrix((-log10(medians[bsite_ordered,bsite_ordered]))))\n",
    "scaled.values[scaled.values == 0] <- 1\n",
    "scaled.values[is.na(scaled.values)] <- 6\n",
    "\n",
    "pdf(file=\"body_site.pdf\")\n",
    "heatmap(scaled.values, Rowv=NA, Colv=NA, scale=\"none\", \n",
    "\tlabRow=gsub(\"_\", \" \", colnames(medians)[bsite_ordered]),\n",
    "    labCol=gsub(\"_\", \" \", colnames(medians)[bsite_ordered]),margins=c(10,10), revC=T,\n",
    "    col=c(heat.colors(5, alpha=0.8), \"white\"), font=2)\n",
    "\n",
    "par(xpd=T)\t#let's write anywhere!\n",
    "#loc<-locator()\n",
    "loc <- list(\"x\"=0.83, \"y\"=0.1)\n",
    "\n",
    "legend(loc, legend=c(expression(paste(\"P<10\"^\"-1\")), expression(paste(\"P<10\"^\"-2\")),\n",
    "\t\t\texpression(paste(\"P<10\"^\"-3\")), expression(paste(\"P<10\"^\"-4\")),\n",
    "\t\t\texpression(paste(\"P<10\"^\"-5\"))), fill=rev(heat.colors(5, alpha=0.8)))\n",
    "par(xpd=F)\t#no more writing anywhere :(\n",
    "dev.off()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quantify rate of change between community types for each body site\n",
    "Now we turned to the question of whether some bodysites were more stable than others. Since the subjects didn't come in on a regular interval, it was difficult to quantify true rates of change or to model the change. We tried several things all of which pointed to the fact that time alone could not explain the change. So we just present the rates of change for each community type and body site (see Figure 3). First we'll start by generating transition matrices for each bodysite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "v1 <- read.table(file=\"community_types.v1.txt\", sep=\" \", header=T)\n",
    "v2 <- read.table(file=\"community_types.v2.txt\", sep=\" \", header=T)\n",
    "v3 <- read.table(file=\"community_types.v3.txt\", sep=\" \", header=T)\n",
    "\n",
    "bodysites <- colnames(v1)\n",
    "\n",
    "for(bs in bodysites){#for each bodysite\n",
    "\tv1.samples <- rownames(v1)#get the the randsid\n",
    "\tv1.parts <- as.character(v1[, grepl(bs, colnames(v1))])#get the community type\n",
    "\tv1.samples <- v1.samples[!is.na(v1.parts)]\t#exclude those sampes where we don't have a\n",
    "\tv1.parts <- v1.parts[!is.na(v1.parts)]\t    #community type\n",
    "\t\n",
    "\tv2.samples <- rownames(v2)\n",
    "\tv2.parts <- as.character(v2[, grepl(bs, colnames(v2))])\n",
    "\tv2.samples <- v2.samples[!is.na(v2.parts)]\t\n",
    "\tv2.parts <- v2.parts[!is.na(v2.parts)]\t\n",
    "\n",
    "\tv3.samples <- rownames(v3)\n",
    "\tv3.parts <- as.character(v3[, grepl(bs, colnames(v3))])\n",
    "\tv3.samples <- v3.samples[!is.na(v3.parts)]\t\n",
    "\tv3.parts <- v3.parts[!is.na(v3.parts)]\n",
    "    \n",
    "    #now we know the samples that have community types\n",
    "\n",
    "\tsamples <- unique(sort(c(v1.samples, v2.samples, v3.samples))) #the list of samples with types\n",
    "\tvisits <- data.frame(v1=rep(NA, length(samples)), v2=rep(NA, length(samples)),\n",
    "                         v3=rep(NA, length(samples)), row.names=samples)#initialize a frame\n",
    "\t\n",
    "\tvisits[(v1.samples), 1] <- v1.parts #fill the data frame with community type data\n",
    "\tvisits[(v2.samples), 2] <- v2.parts #from each visit\n",
    "\tvisits[(v3.samples), 3] <- v3.parts\n",
    "\n",
    "\tfirst <- c(visits[,1], visits[,2]) #create a vector for visits 1 and 2\n",
    "\tsecond <- c(visits[,2], visits[,3])#create a vector for visits 2 and 3\n",
    "\t#now we can compare first and second to see whether a subject changed between visits 1 and 2\n",
    "    #and between visits 2 and 3\n",
    "    \n",
    "\t#need to make sure that we have all the community types in the rows and columns\n",
    "\ttypes <- levels(factor(c(first, second)))\t\n",
    "\tntypes <- length(types)\n",
    "\ttransition <- matrix(rep(0, ntypes*ntypes), nrow=ntypes)\n",
    "\trownames(transition) <- types\n",
    "\tcolnames(transition) <- types\n",
    "\t\n",
    "\tt <- table(first, second)\n",
    "\ttransition[rownames(t), colnames(t)] <- t #now we should have a square transition matrix\n",
    "\t\n",
    "    #output\n",
    "\twrite.table(transition, paste(bs,\".transition.matrix\", sep=\"\"), sep=\"\\t\", quote=F)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have all of the transition matrices for each bodysite, let's generate a bubble plot indicating the stability of each bodysite along with its relative frequency (e.g. Figure 3A). Along the x-axis we'll plot the stability of each community type, along the y-axis we'll plot the bodysite (bunched into general types) and the icon will be sized proportionally to the fraction of samples in that community type. A vertical line will be drawn for each bodysite to indicate the average community type stability across all subjects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "#need to read in the transition matrix and calculate the stability values and fraction\n",
    "#of subjects\n",
    "getStabProp <- function(bodysite){\n",
    "\tdata <- read.table(file=bodysite, header=T)\n",
    "\tparts <- unique(c(colnames(data), rownames(data)))\n",
    "\n",
    "    #get the proportions - what fraction of the rows go to columns\n",
    "\tprop <- prop.table(as.matrix(data), 1) \n",
    "\t\n",
    "\t#define stability as 1 - the proportion of change\n",
    "    stability <- 100*(1-diag(prop)) #get stability values\n",
    "    \n",
    "    #what fraction of people are in each community type\n",
    "\tproportion <- rowSums(data)/sum(data)\t\n",
    "\tdata.frame(row.names=rownames(data), stability=stability, proportion=proportion)\n",
    "}\n",
    "\n",
    "\n",
    "#need to scale the size of the blip to be proportional to the proportion of people in\n",
    "#that community type\n",
    "getCEX <- function(proportion){\n",
    "\tmax.cex=5\n",
    "\tmin.cex=0.25\n",
    "\tcex <- (max.cex-min.cex)*proportion+min.cex\n",
    "\treturn(cex)\n",
    "}\n",
    "\n",
    "\n",
    "ll=0.25  #length of the line indicating the average above the y-position\n",
    "jitter =0.1 #amount to jitter points if there are duplicate stability values\n",
    "\n",
    "#need to plot the stability data along a line for each chunk of bodysites. If there are\n",
    "#duplicate stability values for a bodysite, need to jitter the position on the y-axis\n",
    "plotStability <- function(site.vector, pos){\n",
    "\n",
    "\tfor(site in site.vector){#do on all of the bodysites within a bodytype chunk\n",
    "\n",
    "\t\tdata <- getStabProp(site)#get the stability / proportion data\n",
    "\t\n",
    "\t\typos <- rep(pos, nrow(data))#set the default position on the y-axis\n",
    "        \n",
    "        #if there are duplicate stability values jitter them to either position on the y-axis\n",
    "\t\typos[duplicated(data$stability)] <- ypos[duplicated(data$stability)] + jitter\n",
    "\t\typos[duplicated(data$stability, fromLast=T)] <- ypos[duplicated(data$stability, fromLast=T)] - jitter\n",
    "\n",
    "        #plot the blips on the line sized in proportion to relative abundance\n",
    "\t\tpoints(x=data$stability, y=ypos, pch=21, bg=\"gray\", cex=getCEX(data$proportion))\n",
    "\n",
    "        #calculate the average stability for the bodysite\n",
    "        avg <- sum(data$stability*data$proportion)\n",
    "        \n",
    "        #make a vertical line at the average\n",
    "\t\tlines(x=c(avg, avg), y=c(pos-ll, pos+ll), col=\"black\", lwd=3)\n",
    "\n",
    "        #decrement the position on the y-axis\n",
    "        pos <- pos-1\n",
    "\t}\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "header.line = 11  #position left of the y-axis for the body type heading\n",
    "indent.line = 9   #position left of the y-axis for the bodysite heading\n",
    "\n",
    "pdf(\"bodysite.change.rate.pdf\", height=11, width=8.5)\n",
    "par(mar=c(5, header.line+0.5, 0.5, 0.5)) #need padding on the right\n",
    "\n",
    "plot(x=c(0,100), y=c(1,27), type=\"n\", main=\"\", xlab=\"% Change\", ylab=\"\", axes=F)\n",
    "axis(1)\n",
    "box()\n",
    "\n",
    "\n",
    "#gastrointestinal\n",
    "sites <- c(\"Stool.transition.matrix\")\n",
    "names <- gsub(\".transition.matrix\", \"\", sites)\n",
    "names <- gsub(\"_\", \" \", names)\n",
    "mtext(side=2, line=header.line, at=27, text=\"Gastrointestinal\", las=1, adj=0)\n",
    "mtext(side=2, line=indent.line, at=26, text=\"Stool\", las=1, adj=0)\n",
    "abline(h=26, col=\"gray\")#make horizontal gray line \n",
    "\n",
    "start <- 26\n",
    "plotStability(sites, start)\n",
    "\n",
    "\n",
    "#oral\n",
    "sites <- c(\"Supragingival_plaque.transition.matrix\", \"Palatine_Tonsils.transition.matrix\", \n",
    "\t\t\t\"Subgingival_plaque.transition.matrix\", \"Throat.transition.matrix\",\n",
    "\t\t\t\"Tongue_dorsum.transition.matrix\", \"Keratinized_gingiva.transition.matrix\", \n",
    "\t\t\t\"Buccal_mucosa.transition.matrix\", \"Hard_palate.transition.matrix\", \"Saliva.transition.matrix\")\n",
    "names <- gsub(\".transition.matrix\", \"\", sites)\n",
    "names <- gsub(\"_\", \" \", names)\n",
    "mtext(side=2, line=header.line, at=24, text=\"Oral\", las=1, adj=0)\n",
    "mtext(side=2, line=indent.line, at=15:23, text=rev(names), las=1, adj=0)\n",
    "abline(h=15:23, col=\"gray\")\n",
    "\n",
    "start <- 23\n",
    "plotStability(sites, start)\n",
    "\n",
    "\n",
    "#pulmonary\n",
    "sites <- c(\"Anterior_nares.transition.matrix\")\n",
    "names <- gsub(\".transition.matrix\", \"\", sites)\n",
    "names <- gsub(\"_\", \" \", names)\n",
    "mtext(side=2, line=header.line, at=13, text=\"Pulmonary\", las=1, adj=0)\n",
    "mtext(side=2, line=indent.line, at=12, text=rev(names), las=1, adj=0)\n",
    "abline(h=12, col=\"gray\")\n",
    "\n",
    "start <- 12\n",
    "plotStability(sites, start)\n",
    "\n",
    "\n",
    "#skin\n",
    "sites <- c(\"R_Retroauricular_crease.transition.matrix\", \"L_Retroauricular_crease.transition.matrix\",\n",
    "\t\t\t\"R_Antecubital_fossa.transition.matrix\", \"L_Antecubital_fossa.transition.matrix\")\n",
    "names <- gsub(\".transition.matrix\", \"\", sites)\n",
    "names <- gsub(\"_\", \" \", names)\n",
    "mtext(side=2, line=header.line, at=10, text=\"Skin\", las=1, adj=0)\n",
    "mtext(side=2, line=indent.line, at=6:9, text=rev(names), las=1, adj=0)\n",
    "abline(h=6:9, col=\"gray\")\n",
    "\n",
    "start <- 9\n",
    "plotStability(sites, start)\n",
    "\n",
    "\n",
    "#vagina\n",
    "sites <- c(\"Vaginal_introitus.transition.matrix\",\"Mid_vagina.transition.matrix\", \n",
    "\t\t\t\"Posterior_fornix.transition.matrix\")\n",
    "names <- gsub(\".transition.matrix\", \"\", sites)\n",
    "names <- gsub(\"_\", \" \", names)\n",
    "mtext(side=2, line=header.line, at=4, text=\"Vagina\", las=1, adj=0)\n",
    "mtext(side=2, line=indent.line, at=1:3, text=rev(names), las=1, adj=0)\n",
    "abline(h=1:3, col=\"gray\")\n",
    "\n",
    "start <- 3\n",
    "plotStability(sites, start)\n",
    "\n",
    "\n",
    "#make a legend. for Figure 3A we separated the points a bit so they were easier to see\n",
    "legend(x=80, y=13, legend=c(\"1%\", \"10%\", \"25%\", \"50%\", \"75%\"), pch=21, pt.bg=\"gray\",\n",
    "                    bg=\"white\", pt.cex=getCEX(c(0.01, 0.10, 0.25, 0.50, 0.75)))\n",
    "\n",
    "dev.off()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we'd like to graphically depict the transition matrices as we did in Figure 3B. These will allow us to see how frequently one community type changes into another or stays the same. This took a fair amount of messing with parameters to get the formatting to look decent..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "install.packages(\"diagram\")\n",
    "library(diagram)\n",
    "\n",
    "#set the layouts to use in plotmat for different number of community types\n",
    "getPos <- function(n) {\n",
    "\tif(n==2)\t\t{\t\treturn(c(2))\t\t}\t\t\t\t\t\n",
    "\telse if(n==3)\t{\t\treturn(c(1,2))\t\t}\t\t\t\t\t\n",
    "\telse if(n==4)\t{\t\treturn(c(2,2))\t\t}\t\t\t\t\t\n",
    "\telse if(n==5)\t{\t\treturn(c(2,1, 2))\t}\t\t\t\t\t\n",
    "\telse if(n==6)\t{\t\treturn(c(2,2,2))\t}\n",
    "\telse if(n==7)\t{\t\treturn(c(2,3,2))\t}\n",
    "}\n",
    "\n",
    "        \n",
    "#describe the curvature when there are different numbers of community types\n",
    "getCurves <- function(n) {\n",
    "\tif(n==2){\n",
    "\t\treturn(matrix(c(0,\t0.1,\n",
    "\t\t\t\t\t\t0.1,\t0), nrow=2))\n",
    "\t}\t\t\t\t\t\n",
    "\telse if(n==3){\n",
    "\t\treturn(matrix(c(0,\t0.08,\t0.08,\n",
    "\t\t\t\t\t\t0.08,\t0,\t0.08,\n",
    "\t\t\t\t\t\t0.08,\t0.08,\t0), nrow=3))\n",
    "\t}\t\t\t\t\t\n",
    "\telse if(n==4){\n",
    "\t\treturn(matrix(c(0.00,\t0.08,\t0.08,\t0.65,\n",
    "\t\t\t\t\t\t0.08,\t0.00,\t0.65,\t0.08,\n",
    "\t\t\t\t\t\t0.08,\t0.65,\t0.00,\t0.08,\n",
    "\t\t\t\t\t\t0.65,\t0.08,\t0.08,\t0.00), nrow=4))\n",
    "\t}\t\t\t\t\t\n",
    "\telse if(n==5){\n",
    "\t\treturn(matrix(c(0,\t0.05,\t0.1,\t0.05,\t0.6,\n",
    "\t\t\t\t\t\t0.05,\t0,\t0.1,\t0.6,\t0.05,\n",
    "\t\t\t\t\t\t0.1,\t0.1,\t0,\t0.1,\t0.1,\n",
    "\t\t\t\t\t\t0.05,\t0.6,\t0.1,\t0,\t0.05,\n",
    "\t\t\t\t\t\t0.6,\t0.05,\t0.1,\t0.05,\t0), nrow=5))\n",
    "\t}\n",
    "\telse if(n==6){\n",
    "\t\treturn(matrix(c(0,\t0.05,\t0.05,\t0.1,\t0.2,\t0.15,\n",
    "\t\t\t\t\t\t0.05,\t0,\t0.1,\t0.05,\t0.15,\t0.2,\n",
    "\t\t\t\t\t\t0.05,\t0.1,\t0,\t0.05,\t0.05,\t0.1,\n",
    "\t\t\t\t\t\t0.1,\t0.05,\t0.05,\t0,\t0.1,\t0.05,\n",
    "\t\t\t\t\t\t0.2,\t0.15,\t0.05,\t0.1,\t0,\t0.05,\n",
    "\t\t\t\t\t\t0.15,\t0.2,\t0.1,\t0.05,\t0.05,\t0), nrow=6))\n",
    "\t}\t\t\t\t\t\n",
    "\telse if(n==7){\n",
    "\t\treturn(matrix(c(0,\t0.15,\t0.15,\t0.15,\t0.1,\t0.1,\t0.25,\n",
    "\t\t\t\t\t\t0.15,\t0,\t0.1,\t0.15,\t0.15,\t0.25,\t0.1,\n",
    "\t\t\t\t\t\t0.15,\t0.1,\t0,\t0.15,\t0.25,\t0.15,\t0.1,\n",
    "\t\t\t\t\t\t0.15,\t0.15,\t0.15,\t0,\t0.15,\t0.15,\t0.15,\n",
    "\t\t\t\t\t\t0.1,\t0.15,\t0.25,\t0.15,\t0,\t0.1,\t0.15,\n",
    "\t\t\t\t\t\t0.1,\t0.25,\t0.15,\t0.15,\t0.1,\t0,\t0.15,\n",
    "\t\t\t\t\t\t0.25,\t0.1,\t0.1,\t0.15,\t0.15,\t0.15,\t0), nrow=7))\n",
    "\t}\t\t\t\t\t\n",
    "}\n",
    "\n",
    "        \n",
    "#describe on which side (left/right) and how big self arrows should be\n",
    "getShiftX <- function(n) {\n",
    "\tif(n==2)\t\t{\treturn(c(-0.09, 0.09))\t\t\t\t\t\t\t\t\t}\t\n",
    "\telse if(n==3)\t{\treturn(c(0,-0.09,0.09))\t\t\t\t\t\t\t\t\t}\t\n",
    "\telse if(n==4)\t{\treturn(c(-0.09, 0.09, -0.09, 0.09))\t\t\t\t\t\t}\n",
    "\telse if(n==5)\t{\treturn(c(-0.09, 0.09, 0.09, -0.09, 0.09))\t\t\t\t}\n",
    "\telse if(n==6)\t{\treturn(c(-0.09, 0.09, -0.09, 0.09, -0.09, 0.09))\t\t}\n",
    "\telse if(n==7)\t{\treturn(c(-0.09, 0.09, -0.09, 0.09, 0.09,-0.09, 0.09))\t}\n",
    "}\n",
    "\n",
    "\n",
    "#describe on which side (top/bottom) and how big self arrows should be\n",
    "getShiftY <- function(n) {\n",
    "\tif(n==2)\t\t{ return(0) }\n",
    "\telse if(n==3)\t{ return(c(0.09,-0.04,-0.04)) }\n",
    "\telse if(n==4)\t{ return(0) }\n",
    "\telse if(n==5)\t{ return(0) }\n",
    "\telse if(n==6)\t{ return(0) }\n",
    "\telse if(n==7)\t{ return(0) }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "matrices <- c(\"Anterior_nares.transition.matrix\", \"Buccal_mucosa.transition.matrix\",\n",
    "\t\t\t\"Hard_palate.transition.matrix\", \"Keratinized_gingiva.transition.matrix\",\n",
    "\t\t\t\"L_Antecubital_fossa.transition.matrix\", \"L_Retroauricular_crease.transition.matrix\",\n",
    "\t\t\t\"Mid_vagina.transition.matrix\", \"Palatine_Tonsils.transition.matrix\",\n",
    "\t\t\t\"Posterior_fornix.transition.matrix\", \"R_Antecubital_fossa.transition.matrix\",\n",
    "\t\t\t\"R_Retroauricular_crease.transition.matrix\", \"Saliva.transition.matrix\",\n",
    "\t\t\t\"Stool.transition.matrix\", \"Subgingival_plaque.transition.matrix\",\n",
    "\t\t\t\"Supragingival_plaque.transition.matrix\", \"Throat.transition.matrix\",\n",
    "\t\t\t\"Tongue_dorsum.transition.matrix\", \"Vaginal_introitus.transition.matrix\")\n",
    "\n",
    "for(m in matrices){#go through all of the bodysites\n",
    "\n",
    "\ttrans <- read.table(file=m, header=T)        #want to know the fraction of\n",
    "\ttrans <- 100*prop.table(as.matrix(trans), 1) #community type row that become community type column\t\n",
    "\ttrans <- format(trans, nsmall=1, digits=1)   #try to make the numbers pretty\n",
    "\tntypes <- nrow(trans) #get the number of types\n",
    "\n",
    "\tbs <- gsub(\"^([^.]*).*\", \"\\\\1\", m) #get the bodysite\n",
    "\tbs <- gsub(\"_\", \" \", bs)\n",
    "\t\n",
    "    #plot the transition state diagram\n",
    "\tpdf(file=gsub(\"matrix\", \"pdf\", m))\n",
    "\tplotmat(\n",
    "\t\tt(trans),\n",
    "\t\tpos = getPos(ntypes),\n",
    "\t\tcurve = getCurves(ntypes),\n",
    "\t\tself.shiftx = getShiftX(ntypes),\n",
    "\t\tself.shifty = getShiftY(ntypes),\n",
    "\n",
    "\t\tname = LETTERS[1:ntypes],\n",
    "\t\tlwd = 1.5,\n",
    "\t\tbox.lwd = 2,\n",
    "\t\tcex.txt = 0.75,\n",
    "\t\tbox.cex = 1.2,\n",
    "\t\tbox.size = 0.06,\n",
    "\t\tarr.length = 0.5,\n",
    "\t\tbox.type = \"circle\",\n",
    "\t\tbox.prop = 1,\n",
    "\t\tshadow.size = 0,\n",
    "\t\tself.cex = 0.6,\n",
    "\t\tmy = -0.075,\n",
    "\t\tmx = -0.01,\n",
    "\t\trelsize = 0.95,\n",
    "\t\tcex.main=1.5,\n",
    "\t\tbox.col=\"gray\",\n",
    "\t\ttxt.col=\"black\",\n",
    "\t\tarr.col=\"black\",\n",
    "\t\tdtext=0.5,\n",
    "\t\tabsent=-0.1,\n",
    "\t\tmain = bs, \n",
    "        latex=T\n",
    "\t)\n",
    "\tdev.off()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Effect of time on stability of community types\n",
    "Since the community types were not fixed for each person and some community types were not stable, we next endeavored to determine whether we could associate changes in community types with the duration between samplings. As we mention in the paper, it was not possible to identify any significant relationships."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "#get the community type data for each visit\n",
    "v1 <- read.table(file=\"community_types.v1.txt\", sep=\" \", header=T)\n",
    "v2 <- read.table(file=\"community_types.v2.txt\", sep=\" \", header=T)\n",
    "v3 <- read.table(file=\"community_types.v3.txt\", sep=\" \", header=T)\n",
    "bodysites <- colnames(v1)\n",
    "\n",
    "#get the dates of the various visits\n",
    "md <- read.table(file=\"../dbGaP/phs000228.v3.pht002157.v1.p1.c1.EMMES_HMP_DTP_DHX_DVD.HMP.txt\", header=T, sep=\"\\t\", na.string=c(NA, \"\"))\n",
    "md1 <- md[md$VISNO == \"01\", c(\"RANDSID\", \"study_day\")]\n",
    "md2 <- md[md$VISNO == \"02\", c(\"RANDSID\", \"study_day\")]\n",
    "md3 <- md[md$VISNO == \"03\", c(\"RANDSID\", \"study_day\")]\n",
    "\n",
    "randsids <- levels(factor(md$RANDSID))\n",
    "dates <- matrix(rep(NA, length(randsids)*3), nrow=length(randsids))\n",
    "rownames(dates) <- randsids\n",
    "colnames(dates) <- c(\"visit1\", \"visit2\", \"visit3\")\n",
    "\n",
    "#create a dates matrix so we know when each visit occured\n",
    "dates[as.character(md1$RANDSID), \"visit1\"] <- md1$study_day\n",
    "dates[as.character(md2$RANDSID), \"visit2\"] <- md2$study_day\n",
    "dates[as.character(md3$RANDSID), \"visit3\"] <- md3$study_day\n",
    "\n",
    "\n",
    "for(bs in bodysites){#go through the bodysites\n",
    "\ttypes <- matrix(rep(NA, length(randsids)*3), ncol=3) #create a types matrix for the bodysite so we know the type\n",
    "\tcolnames(types) <- c(\"visit1\", \"visit2\", \"visit3\")   #at each visit\n",
    "\trownames(types) <- randsids\n",
    "\t\n",
    "\ttypes[rownames(v1), \"visit1\"] <- v1[,bs] #fill the matirx\n",
    "\ttypes[rownames(v2), \"visit2\"] <- v2[,bs]\n",
    "\ttypes[rownames(v3), \"visit3\"] <- v3[,bs]\n",
    "\t\n",
    "    #figure out who had the same community type at visits 1 and 2 and at visits 2 and 3\n",
    "\tv12.same <- !is.na(types[,\"visit1\"]) & !is.na(types[,\"visit2\"]) & types[,\"visit1\"] == types[,\"visit2\"]\n",
    "\tv23.same <- !is.na(types[,\"visit2\"]) & !is.na(types[,\"visit3\"]) & types[,\"visit2\"] == types[,\"visit3\"]\n",
    "\n",
    "    #get the number of days between visits that had the same community type\n",
    "    dates.same <- c(dates[v12.same, \"visit2\"] - dates[v12.same, \"visit1\"], dates[v23.same, \"visit3\"] -\n",
    "                    dates[v23.same, \"visit2\"])\t\n",
    "\t\n",
    "\n",
    "    #figure out who had different community type at visits 1 and 2 and at visits 2 and 3\n",
    "    v12.different <- !is.na(types[,\"visit1\"]) & !is.na(types[,\"visit2\"]) & types[,\"visit1\"] != types[,\"visit2\"]\n",
    "\tv23.different <- !is.na(types[,\"visit2\"]) & !is.na(types[,\"visit3\"]) & types[,\"visit2\"] != types[,\"visit3\"]\n",
    "\n",
    "    #get the number of days between visits that had different community types\n",
    "    dates.different <- c(dates[v12.different, \"visit2\"] - dates[v12.different, \"visit1\"], dates[v23.different, \"visit3\"] -\n",
    "                         dates[v23.different, \"visit2\"])\t\n",
    "\n",
    "    \n",
    "    #test for a significant difference in the days between visits with different community types vs. those that were the same\n",
    "\tpooled.dates <- c(dates.same, dates.different)\n",
    "\tcategory <- c(rep(\"Same\\nCommunity Type\", length(dates.same)), rep(\"Different\\nCommunity Type\", length(dates.different)))\n",
    "\tp.value <- format(wilcox.test(pooled.dates~category)$p.value, digits=3)\n",
    "\n",
    "\t\n",
    "    #plot the data\n",
    "    pdf(file=paste(bs, \".time.pdf\", sep=\"\"))\n",
    "\tpar(mar=c(5, 8, 0.5, 0.5))\n",
    "\tstripchart(pooled.dates~category, method=\"jitter\", yaxt=\"n\", xaxt=\"n\", ylab=\"\", xlab=\"Days between samplings\",\n",
    "               xlim=c(0,max(md3$study_day)))\n",
    "\taxis(1)\n",
    "\taxis(2, at=c(1,2), label=levels(factor(category)), tick=F, las=2)\n",
    "\ttext(x=450, y=2.18, label=paste(bs, \"\\n(P=\", p.value, \")\", sep=\"\"), pos=2)\n",
    "\tdev.off()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Running this and looking at the resulting pdfs makes it clear that none of the times between samplimgs for the community types that changed and those that didn't change were significantly different. The P-value in the upper right corner of each plot is from a non-parametric Wilcox Test."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Validate DMM models as a community type approach relative to the PAM-based approach\n",
    "Nearly every study to perform community typing has used partitioning around the medoid with distances between samples as the basis for identifying community types. Metrics (e.g. CH and Silouette) are then used to pick the optimal number of community types. For some reason [Koren et al. 2013](http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1002863) did not use the DMM approach eventhough it was published well before their study. We did our best to replicate their methods of simulating a community with four types and evaluate the PAM and DMM-based approach. We also, simulated a community with only one community type to look at the likelihood of falsely detecting a community type. Finally, we generated PAM-based community types for each bodysite and evaluated their Laplace score for those types. These results are summarized in the Supplemental Data section of our paper. \n",
    "\n",
    "### Simulating communities with one and four community types\n",
    "The above mentioned [Koren et al. 2013](http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1002863) study simulated four community types. Here's their methods:  \n",
    "\n",
    "> \"We generated a synthetic dataset of 100 communities each containing 3,000 “sequences” belonging to 500 mock OTUs. For each synthetic community, 90% (2,700 sequences, or OTU observations) was drawn from the same randomly generated lognormal abundance distribution (shared across all communities) and the remaining 10% (300 sequences) drawn from one of four unique lognormal distributions, forcing the data into four clusters. We then applied the enterotyping methods as described above.\"\n",
    "\n",
    "This was a bit vague (shape parameters?). Plus it didn't include a negative control, which would be a single community type. Plus they didn't do the DMM analysis. So we attempated to replicate and improve upon their analysis with two sets of imulations:\n",
    "\n",
    " * **Single community type.**  We constructed a truncated lognormal distribution with a normal mean of 1.25 and a normal standard deviation of 3.00 and a maximum richness of 500.  For each of 100 simulated communities we drew 3,000 random integers (i.e. sequences) from this distribution where each integer represented a specific OTU.  \n",
    " * **Four community types.**  We used the same truncated lognormal distribution described for the single cluster simulation to draw 2,700 integers sequences for each of 25 samples.  We then permuted the distribution and drew an additional 300 integers from the permuted distribution and added those counts to non-permuted distribution.  We repeated this procedure three additional times so that we had effectively simulated sampling 3,000 sequences from 100 samples representing 4 community types.  \n",
    " \n",
    "In the following code chunk we'll generate the **quad**ruple simulated data and the single simulated data and output them as shared files so we can analyze them using the PAM and DMM-based approaches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ..\n",
    "mkdir simulation\n",
    "cd simulation/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR: Cell magic `%%R` not found.\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "getTruncatedLogNormal <- function(nseqs, limit, mu, sd){\n",
    "    mu <- 1.25\n",
    "    sd <- 3.00\n",
    "    \n",
    "\tdist <- round(rlnorm(nseqs*10, meanlog=mu, sdlog=sd))+1   #pull 10-fold more sequences than we need\n",
    "\tdist <- dist[dist < limit]                                #in case some go over limit (maximum richness)\n",
    "\tdist <- dist[0:nseqs]                                     #return the first nsequences\n",
    "}\n",
    "\n",
    "\n",
    "iterations <- 100\n",
    "\n",
    "lower <- 1\n",
    "upper <- 500\n",
    "\n",
    "frac.z <- 0.90                       #fraction to draw from the common distribution - z = common distribution\n",
    "nseqs <- 3000                        #number of sequences we want to draw\n",
    "ncommunities <- 100                  #number of communities to sample from\n",
    "samples_per <- ncommunities * 0.25   #we want each community to have 25% of the samples\n",
    "\n",
    "a.sample <- sample(1:upper)          #here is the permutation of the ordering of the OTUs that corresponds\n",
    "b.sample <- sample(1:upper)          #to each of the four community types. this way the \"extra\" distribution\n",
    "c.sample <- sample(1:upper)          #will be the same shaped distribution, just in a different order\n",
    "d.sample <- sample(1:upper)\n",
    "\n",
    "for(iter in 1:iterations){\n",
    "    #we're going to fill each of the 4 community types with data\n",
    "    \n",
    "\ta<-matrix(nrow=samples_per, ncol=upper)  \n",
    "\tfor(i in 1:samples_per){   #now looping over each of the 25 samples...\n",
    "\t\ta.dist <- getTruncatedLogNormal(nseqs*(1-frac.z), upper, mu, sd) #the a component draws from the LN distro and reorders\n",
    "\t\tz.dist <- getTruncatedLogNormal(nseqs*frac.z, upper, mu, sd)     #the z component is the regular ordered LN distro\n",
    "\t\ta.table <- table(c(a.sample[a.dist], z.dist))                    #get the number of times each OTU/integer shows up\n",
    "\t\tx <- rep(0, upper); names(x) <-1:500       #make sure everything is on the same OTU naming scheme\n",
    "\t\tx[names(a.table)] <- a.table\n",
    "\t\ta[i,] <- x     #viola - the a component of the shared file - repeat for b, c, and d\n",
    "\t}\n",
    "\n",
    "\tb<-matrix(nrow=samples_per, ncol=upper)\n",
    "\tfor(i in 1:samples_per){\n",
    "\t\tb.dist <- getTruncatedLogNormal(nseqs*(1-frac.z), upper, mu, sd)\n",
    "\t\tz.dist <- getTruncatedLogNormal(nseqs*frac.z, upper, mu, sd)\n",
    "\t\tb.table <- table(c(b.sample[b.dist], z.dist))\n",
    "\t\tx <- rep(0, upper); names(x) <-1:500\n",
    "\t\tx[names(b.table)] <- b.table\n",
    "\t\tb[i,] <- x\n",
    "\t}\n",
    "\n",
    "\tc<-matrix(nrow=samples_per, ncol=upper)\n",
    "\tfor(i in 1:samples_per){\n",
    "\t\tc.dist <- getTruncatedLogNormal(nseqs*(1-frac.z), upper, mu, sd)\n",
    "\t\tz.dist <- getTruncatedLogNormal(nseqs*frac.z, upper, mu, sd)\n",
    "\t\tc.table <- table(c(c.sample[c.dist], z.dist))\n",
    "\t\tx <- rep(0, upper); names(x) <-1:500\n",
    "\t\tx[names(c.table)] <- c.table\n",
    "\t\tc[i,] <- x\n",
    "\t}\n",
    "\n",
    "\td<-matrix(nrow=samples_per, ncol=upper)\n",
    "\tfor(i in 1:samples_per){\n",
    "\t\td.dist <- getTruncatedLogNormal(nseqs*(1-frac.z), upper, mu, sd)\n",
    "\t\tz.dist <- getTruncatedLogNormal(nseqs*frac.z, upper, mu, sd)\n",
    "\t\td.table <- table(c(d.sample[d.dist], z.dist))\n",
    "\t\tx <- rep(0, upper); names(x) <-1:500\n",
    "\t\tx[names(d.table)] <- d.table\n",
    "\t\td[i,] <- x\n",
    "\t}\n",
    "\n",
    "\tshared <- rbind(a, b, c, d)  #make shared file\n",
    "\n",
    "\tshared <- shared[, apply(shared, 2, sum) > 0]  #remove columns that only have zeros\n",
    "\tnotus <- ncol(shared) #build the shared file...\n",
    "\tlabel <- rep(1, ncommunities)  #the first column has a common label\n",
    "    \n",
    "    #the second column has the sample name - here we indicate the community type for each\n",
    "\tgroups <- c(paste(1:samples_per, \".A\", sep=\"\"),\n",
    "                paste(1:samples_per, \".B\", sep=\"\"),\n",
    "                paste(1:samples_per, \".C\", sep=\"\"),\n",
    "                paste(1:samples_per, \".D\", sep=\"\"))\n",
    "\tnotus.col <- rep(notus, ncommunities)  # the third column has the number of OTUs\n",
    "\tfull.shared <- cbind(label, groups, notus.col, shared)  #...build the shared file\n",
    "\tcolnames(full.shared) <- c(\"label\", \"Group\", \"numOTUs\", 1:notus)\n",
    "\twrite.table(file=paste(\"quad_\", iter, \".shared\", sep=\"\"), full.shared, row.names=F, quote=F, sep=\"\\t\")\n",
    "    #done!\n",
    "}\n",
    "\n",
    "\n",
    "#generate the single case...\n",
    "frac.z <- 1.0000  #now we want to build a shared file where there's only one community type\n",
    "for(iter in 1:iterations){\n",
    "\tz<-matrix(nrow=ncommunities, ncol=upper)\n",
    "\tfor(i in 1:ncommunities){\n",
    "\t\tz.dist <- getTruncatedLogNormal(nseqs*frac.z, upper, mu, sd)  #again draw from our lognormal disribution\n",
    "\t\tz.table <- table(z.dist)                  #everything else is pretty much the same...\n",
    "\t\tx <- rep(0, upper); names(x) <-1:500\n",
    "\t\tx[names(z.table)] <- z.table\n",
    "\t\tz[i,] <- x\n",
    "\t}\n",
    "\tshared <- z\n",
    "\tshared <- shared[, apply(shared, 2, sum) > 0]\n",
    "\tnotus <- ncol(shared)\n",
    "\tlabel <- rep(1, ncommunities)\n",
    "\tgroups <- c(paste(1:ncommunities, \".A\", sep=\"\"))\n",
    "\tnotus.col <- rep(notus, ncommunities)\n",
    "\n",
    "\tfull.shared <- cbind(label, groups, notus.col, shared)\n",
    "\tcolnames(full.shared) <- c(\"label\", \"Group\", \"numOTUs\", 1:notus)\n",
    "\n",
    "\twrite.table(file=paste(\"single_\", iter, \".shared\", sep=\"\"), full.shared, row.names=F, quote=F, sep=\"\\t\")\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we've built the single and quadruple community type shared files we want to know whether we can detect those community types. First we'll start with the PAM-based approach. As we mention in the supplement, the Silhouette Index (SI) is prefered over the Calinski-Harabasz (CH) statistic because it had a more objective basis. They asserted that SI values below 0.25 have a lack of substantial structure to the clusters. Let's test this out..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "require(cluster)\n",
    "require(fpc)\n",
    "\n",
    "n.simulations <- 100\n",
    "\n",
    "dist.JSD <- function(inMatrix, pseudocount=0.000001, ...) {\n",
    "\t\n",
    "\tKLD <- function(x,y) sum(x *log(x/y))\n",
    "\t\n",
    "#nb:\tthe line below is from the MetaHit group, which is actually the \n",
    "#\t\troot JSD, not the straight up JSD\n",
    "#\tJSD <- function(x,y) sqrt(0.5 * KLD(x, (x+y)/2) + 0.5 * KLD(y, (x+y)/2))\n",
    "\tJSD <- function(x,y) (0.5 * KLD(x, (x+y)/2) + 0.5 * KLD(y, (x+y)/2))\n",
    "\t\n",
    "\tmatrixColSize <- length(colnames(inMatrix))\n",
    "\tmatrixRowSize <- length(rownames(inMatrix))\n",
    "\tcolnames <- colnames(inMatrix)\n",
    "\tresultsMatrix <- matrix(0, matrixColSize, matrixColSize)\n",
    "\n",
    "\t#I'm not sure I \"believe\" in JSD, but it's what all the enterotype papers\n",
    "\t#have used so I use it here. There seems to be a funkiness here in that\n",
    "\t#you have to add a very small number to the relative abundances since you \n",
    "\t#can easily have a divide by zero. Sketchy for such sparse data matrices. IMHO.\n",
    "\tinMatrix = apply(inMatrix,1:2,function(x) ifelse (x==0,pseudocount,x))\n",
    "\n",
    "\tfor(i in 1:matrixColSize) {\n",
    "\t\tfor(j in 1:matrixColSize) { \n",
    "\t\t\tif(j > i){\t#no need to do everything twice.\n",
    "\t\t\t\tresultsMatrix[i,j] <- JSD(as.vector(inMatrix[,i]),\n",
    "\t\t\t\t\t\t\t\t\t\t\tas.vector(inMatrix[,j]))\n",
    "\t\t\t\tresultsMatrix[j,i] <- resultsMatrix[i,j]\n",
    "\t\t\t}\n",
    "\t\t}\n",
    "\t}\n",
    "\tcolnames -> colnames(resultsMatrix) -> rownames(resultsMatrix)\n",
    "\tas.dist(resultsMatrix)->resultsMatrix\n",
    "\tattr(resultsMatrix, \"method\") <- \"dist\"\n",
    "\treturn(resultsMatrix) \n",
    "}\n",
    "\n",
    "#use the built in pam clustering algoirthm\n",
    "pam.clustering <- function(x,k) { # x is a distance matrix and k the number of clusters\n",
    "\tcluster <- as.vector(pam(as.dist(x), k, diss=TRUE)$clustering)\n",
    "\treturn(cluster)\n",
    "}\n",
    "\n",
    "\n",
    "pam.approach <- function(prefix){\n",
    "    max.ch <- vector()\n",
    "    max.silhouette <- vector()\n",
    "    \n",
    "    write(paste(\"simulation\", \"k_CH\", \"max_CH\", \"k_Silhouette\", \"max_Silhoutte\"),\n",
    "          paste(prefix, \"enterotype_output.txt\", sep=\"\"), append=F)\n",
    "    \n",
    "    for(i in 1:n.simulations){\n",
    "    \n",
    "        #read back in the simulated shared file\n",
    "        data <- read.table(paste(prefix, i, \".shared\", sep=\"\"), header=T, row.names=2)\n",
    "        data <- data[,-c(1,2)]\n",
    "        data <- data / apply(data, 1, sum)\n",
    "        data <- t(data)#need OTUs in rows and samples in columns\n",
    "    \n",
    "        data.dist <- dist.JSD(data)#get the distance matrix\n",
    "    \n",
    "        ch <- vector()\n",
    "        width <- vector()\n",
    "    \n",
    "        for (k in 2:10) { #calcualte the ch and silhouette for k=2:10 (can't do k=1)\n",
    "            data.cluster <- pam.clustering(data.dist, k)\n",
    "            stats <- cluster.stats(data.dist, data.cluster)\n",
    "            ch[k] <- stats$ch\n",
    "            width[k] <- stats$avg.silwidth\n",
    "        }\n",
    "        \n",
    "        max.ch[i] <- which.max(ch)\t#get the maximum\n",
    "        max.silhouette[i] <- which.max(width) #get the maximum\n",
    "    \n",
    "        write(paste(i, max.ch[i], max(ch, na.rm=T), max.silhouette[i], max(width, na.rm=T)),\n",
    "              paste(prefix, \"enterotype_output.txt\", sep=\"\"), append=T)\n",
    "    }\n",
    "}\n",
    "\n",
    "pam.approach(\"single_\")\n",
    "pam.approach(\"quad_\")\n",
    "\n",
    "\n",
    "single <- read.table(file=\"single_enterotype_output.txt\", header=T, row.names=1)\n",
    "single$k_CH <- factor(single$k_CH)\n",
    "single$k_Silhouette <- factor(single$k_Silhouette)\n",
    "summary(single)\n",
    "\n",
    "quadruple <- read.table(file=\"quad_enterotype_output.txt\", header=T, row.names=1)\n",
    "quadruple$k_CH <- factor(quadruple$k_CH)\n",
    "quadruple$k_Silhouette <- factor(quadruple$k_Silhouette)\n",
    "summary(quadruple)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And so we see from these final commands that the PAM-based methods can't really call a single community type case (although the scores are admittedly low) and while it can call the four community type case, the scores are pretty mediocre by Koren et al.'s standards. Next we want to run the DMM models on the community type data. Like we did for the microbiome data, we will run these in parallel to control for picking the wrong starting condition and the natural variation in the algorithm. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR: Cell magic `%%` not found.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir dmm.sim.r1 dmm.sim.r2 dmm.sim.r3 dmm.sim.r4 dmm.sim.r5\n",
    "\n",
    ">simulated.batch\n",
    "\n",
    "for SHARED in $(ls single_*.shared quad_*.shared)\n",
    "do\n",
    "    echo \"mothur \\\"#get.communitytype(shared=../$SHARED, outputdir=./)\\\"\" >> simulated.batch\n",
    "done\n",
    "\n",
    "split -l 10 simulated.batch simulated_\n",
    "\n",
    "\n",
    "cat > header.batch << \"_EOF_\"\n",
    "#!/bin/sh\n",
    "#PBS -l nodes=1:ppn=1\n",
    "#PBS -l walltime=500:00:00\n",
    "#PBS -l mem=1gb\n",
    "#PBS -j oe\n",
    "#PBS -m abe\n",
    "#PBS -V\n",
    "#PBS -M your@email.here\n",
    "#PBS -q first\n",
    "\n",
    "echo \"ncpus-2.pbs\"\n",
    "cat $PBS_NODEFILE\n",
    "qstat -f $PBS_JOBID\n",
    "\n",
    "cd $PBS_O_WORKDIR\n",
    "\n",
    "NCPUS=`wc -l $PBS_NODEFILE | awk '{print $1}'`\n",
    "\n",
    "_EOF_\n",
    "\n",
    "\n",
    "cat > tail.batch << \"_EOF_\"\n",
    "\n",
    "echo \"qsub working directory absolute is\"\n",
    "echo $PBS_O_WORKDIR\n",
    "exit\n",
    "_EOF_\n",
    "\n",
    "\n",
    "for DIR in dmm.sim.r1 dmm.sim.r2 dmm.sim.r3 dmm.sim.r4 dmm.sim.r5\n",
    "do\n",
    "    cd $DIR\n",
    "    \n",
    "    cp ../simulated_?? ./\n",
    "    for BATCH in $(ls simulated_??)\n",
    "    do\n",
    "        cat ../header.batch $BATCH ../tail.batch > $BATCH.$DIR.batch\n",
    "        qsub $BATCH.$DIR.batch\n",
    "    done\n",
    "    \n",
    "    cd ../\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that those jobs have completed we want to find the best fit for each simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "niters <- 100\n",
    "\n",
    "getNTypes <- function(prefix){\n",
    "\tntypes <- rep(NA, niters)\n",
    "\n",
    "\tfor(i in 1:niters){\n",
    "\t\tr1 <- read.table(file=paste(\"dmm.sim.r1/\", prefix, i, \".1.dmm.mix.fit\", sep=\"\"), header=T, row.names=1)\n",
    "\t\tr2 <- read.table(file=paste(\"dmm.sim.r2/\", prefix, i, \".1.dmm.mix.fit\", sep=\"\"), header=T, row.names=1)\n",
    "\t\tr3 <- read.table(file=paste(\"dmm.sim.r3/\", prefix, i, \".1.dmm.mix.fit\", sep=\"\"), header=T, row.names=1)\n",
    "\t\tr4 <- read.table(file=paste(\"dmm.sim.r4/\", prefix, i, \".1.dmm.mix.fit\", sep=\"\"), header=T, row.names=1)\n",
    "\t\tr5 <- read.table(file=paste(\"dmm.sim.r5/\", prefix, i, \".1.dmm.mix.fit\", sep=\"\"), header=T, row.names=1)\n",
    "\n",
    "\t\tr1[is.infinite(as.matrix(r1))] <- NA\n",
    "\t\tr2[is.infinite(as.matrix(r2))] <- NA\n",
    "\t\tr3[is.infinite(as.matrix(r3))] <- NA\n",
    "\t\tr4[is.infinite(as.matrix(r4))] <- NA\n",
    "\t\tr5[is.infinite(as.matrix(r5))] <- NA\n",
    "\n",
    "\t\tmaxK <- max(c(nrow(r1), nrow(r2), nrow(r3), nrow(r4), nrow(r5)))\n",
    "\t\tlaplace <- matrix(rep(NA, 5*maxK), nrow=maxK)\n",
    "\t\tcolnames(laplace) <- c(\"r1\", \"r2\", \"r3\", \"r4\", \"r5\")\n",
    "\t\trownames(laplace) <- 1:maxK\n",
    "\t\t\n",
    "\t\tlaplace[1:nrow(r1),\"r1\"] <- r1[, \"Laplace\"]\n",
    "\t\tlaplace[1:nrow(r2),\"r2\"] <- r2[, \"Laplace\"]\n",
    "\t\tlaplace[1:nrow(r3),\"r3\"] <- r3[, \"Laplace\"]\n",
    "\t\tlaplace[1:nrow(r4),\"r4\"] <- r4[, \"Laplace\"]\n",
    "\t\tlaplace[1:nrow(r5),\"r5\"] <- r5[, \"Laplace\"]\n",
    "\t\t\n",
    "\t\tlaplace[apply(is.na(laplace), 1, sum) ==5 ,] <- rep(1e6,5)\n",
    "\n",
    "\t\tcomposite <- apply(laplace, 1, min, na.rm=T)\n",
    "\t\tntypes[i] <- which.min(composite)\n",
    "\t}\n",
    "\treturn(ntypes)\n",
    "}\n",
    "\n",
    "n.singletypes <- getNTypes(\"single_\")\n",
    "table(n.singletypes)\n",
    "\n",
    "n.quadtypes <- getNTypes(\"quad_\")\n",
    "table(n.quadtypes)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As these results indicate, the DMM-based approach correctly called all of the single and quadruple community type iterations."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Applying PAM-based approach to HMP data\n",
    "Next we wanted to see how the PAM-based approach did on the observed community data. First, we assigned all of the bodysites to community types for varying numbers of clusters and found the best assignment using SI and CH. Next, we used those assignments to calculate the Laplace metric. This allowed us to compare the PAM and DMM-based assignments directly. Note that a key difference between the PAM and DMM-based approaches is that PAM-based methods require the use of a distance calculator whereas the DMM-based approach does not."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "shared <- c(\"Antecubital_fossa.tx.1.subsample.shared\", \"Anterior_nares.tx.1.subsample.shared\", \n",
    "\t\"Buccal_mucosa.tx.1.subsample.shared\", \"Hard_palate.tx.1.subsample.shared\", \n",
    "\t\"Keratinized_gingiva.tx.1.subsample.shared\", \"Palatine_Tonsils.tx.1.subsample.shared\", \n",
    "\t\"Retroauricular_crease.tx.1.subsample.shared\", \"Saliva.tx.1.subsample.shared\", \n",
    "\t\"Stool.tx.1.subsample.shared\", \"Subgingival_plaque.tx.1.subsample.shared\", \n",
    "\t\"Supragingival_plaque.tx.1.subsample.shared\", \t\"Throat.tx.1.subsample.shared\", \n",
    "\t\"Tongue_dorsum.tx.1.subsample.shared\", \"Vagina.tx.1.subsample.shared\")\n",
    "\n",
    "path <- \"../v35/\"\n",
    "\n",
    "#set up a summary table where the rows are the bodysites an the columns are the silhouette width, the number\n",
    "#of clusters based on SI, the CH index, and the number of clusters based on the CH index\n",
    "summary <- data.frame(matrix(rep(NA, 14*4), nrow=14, ncol=4))\n",
    "rownames(summary) <- gsub(\".tx.1.subsample.shared\", \"\", shared)\n",
    "colnames(summary) <- c(\"width\", \"w_clusters\", \"ch\", \"ch_clusters\")\n",
    "\t\n",
    "for(s in shared){\n",
    "\twrite(s, \"\")\n",
    "\tstub <- gsub(\"shared\", \"\", s)\n",
    "    file <- paste(path, s, sep=\"\")\n",
    "\tdata <- read.table(file, header=T, row.names=2)  #get the shared file and turn it into relative abundance data\n",
    "\tdata <- data[,-c(1,2)]\n",
    "\tdata <- data / apply(data, 1, sum) \n",
    "\tdata <- t(data) #transpose it\n",
    "\n",
    "\tdata.dist <- dist.JSD(data)  #get the distance matrix\n",
    "\n",
    "\tch <- vector()\n",
    "\twidth <- vector()\n",
    "\n",
    "\tfor (k in 2:10) {  #get the si and ch indices for 2 to 10 clusters\n",
    "\t\tdata.cluster <- pam.clustering(data.dist, k)\n",
    "\t\tstats <- cluster.stats(data.dist, data.cluster)\n",
    "\t\tch[k] <- stats$ch\n",
    "\t\twidth[k] <- stats$avg.silwidth\n",
    "\t}\n",
    "\t\n",
    "\tpdf(paste(stub, \"enterotype.pdf\", sep=\"\"), height=10, width=5)  #generate a spike plot for the SI and CH indices\n",
    "\tpar(mfrow=c(2,1))\n",
    "\tplot(ch, type=\"h\", xlab=\"k clusters\", ylab=\"CH index\")\n",
    "\tplot(width~seq(1:10), type=\"h\", xlab=\"kclusters\", ylab=\"Average Silhouette Width\")\n",
    "\tdev.off()\n",
    "\t\n",
    "\tmax.ch <- which.max(ch)  #get the number of types by CH\n",
    "\tmax.silhouette <- which.max(width)  #get the number of types by SI\n",
    "\t\n",
    "\tsummary[gsub(\".tx.1.subsample.shared\", \"\", s),] <- c(max(width, na.rm=T), max.silhouette, max(ch, na.rm=T), max.ch)#update the summary file\n",
    "\twrite.table(cbind(colnames(data), paste(\"Partition_\", pam.clustering(data.dist, max.ch), sep=\"\")), #output a design file\n",
    "                paste(stub, \"pam.ch.enterotype\", sep=\"\"), quote=F, row.names=F, col.names=F) #based on the CH index\n",
    "\n",
    "    write.table(cbind(colnames(data), paste(\"Partition_\", pam.clustering(data.dist, max.silhouette), sep=\"\")), #output a design\n",
    "                paste(stub, \"pam.sw.enterotype\", sep=\"\"), quote=F, row.names=F, col.names=F) #file based on the SI index\n",
    "}\t\n",
    "\n",
    "write.table(summary, file=\"pam.enterotype.summary\", quote=F, sep=\"\\t\") #output the summary file\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have the community types by the PAM-based approach, let's run the community type assignments into the DMM model and calculate the Laplace value for each bodysite based on the number of clusters found using the SI and CH indices. To do this, we use a homebrew C++ program that will take in a shared file and a design file. You can get our code from [GitHub](https://github.com/SchlossLab/pds_dmm):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "git clone git@github.com:SchlossLab/pds_dmm.git \n",
    "cd pds_dmm\n",
    "make\n",
    "cd ..\n",
    "\n",
    "for ENTEROTYPE in $(ls *enterotype)\n",
    "do\n",
    "    SHARED=${ENTEROTYPE//pam.*/shared}\n",
    "\tpds_dmm/pds_dmm -shared ../v35/$SHARED -design $ENTEROTYPE\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we want to synthesize everything to make a table that looks like Extended Data Table 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "body_stubs <- c(\"Antecubital_fossa\", \"Anterior_nares\", \"Buccal_mucosa\", \"Hard_palate\", \n",
    "\t\t\t\t\"Keratinized_gingiva\", \"Palatine_Tonsils\", \"Retroauricular_crease\",\n",
    "\t\t\t\t\"Saliva\", \"Stool\", \"Subgingival_plaque\", \"Supragingival_plaque\", \"Throat\", \n",
    "\t\t\t\t\"Tongue_dorsum\", \"Vagina\")\n",
    "\n",
    "sausage <- \".tx.1.subsample.\"\n",
    "\n",
    "mix_suffix <- \"1.dmm.mix.fit\"\n",
    "sw_suffix <- \"pam.sw.fit\"\n",
    "ch_suffix <- \"pam.ch.fit\"\n",
    "\n",
    "pam.summary <- read.table(file=\"pam.enterotype.summary\", header=T)\n",
    "\n",
    "extended.table.2 <- matrix(rep(NA, 8*length(body_stubs)), ncol=8)\n",
    "rownames(extended.table.2) <- body_stubs\n",
    "colnames(extended.table.2) <- c(\"SI.clusters\", \"SI.score\", \"SI.Laplace\", \"CH.clusters\",\n",
    "\t\t\t\t\t\t\t\t\"CH.score\", \"CH.Laplace\", \"DMM.clusters\", \"DMM.Laplace\")\n",
    "\n",
    "for(bs in body_stubs){\n",
    "\tdmm.fit <- read.table(file=paste(\"../v35/dmm_best/\", bs, sausage, mix_suffix, sep=\"\"), header=T)\n",
    "\tdmm.K <- which.min(dmm.fit$Laplace)\n",
    "\tdmm.Laplace <- min(dmm.fit$Laplace, na.rm=T)\n",
    "\t\n",
    "\tsw.Laplace <- read.table(file=paste(bs, sausage, sw_suffix, sep=\"\"), header=T)$Laplace\n",
    "\tch.Laplace <- read.table(file=paste(bs, sausage, ch_suffix, sep=\"\"), header=T)$Laplace\n",
    "\n",
    "\tpam <- pam.summary[bs,]\n",
    "\textended.table.2[bs,] <- c(pam$w_clusters, pam$width, sw.Laplace, pam$ch_clusters, pam$ch, ch.Laplace, dmm.K, dmm.Laplace)\n",
    "}\n",
    "\n",
    "write.table(file=\"extended_table_2.txt\", extended.table.2, quote=F)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Representing community types in ordinations\n",
    "We strongly resisted the reviewer's request to put in an ordination of the samples colored by the community types. Why? Well, because we know that many OTUs are responsible for dividing the samples amongst the community types and an ordination really only allows you to use a couple of variables. Futhermore, construction of an ordination requires filtering the data through a distance matrix, which is one of the problems with the PAM-based approach. Whatever, here we present code generating the ordinations for each bodysite with coloring for each community type. First we use mothur to calculate the JSD distances between samples. We do this by rarefying the samples to 1000 sequences 100 times. After all of the iterations we calculate the average JSD matrix:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for SHARED in $(ls ../v35/*subsample.shared)\n",
    "do\n",
    "    DIST=${SHARED//shared/jsd.1.lt.dist}\n",
    "    DIST=${DIST//..\\/v35\\//}\n",
    "\tmothur \"#dist.shared(shared=$SHARED, calc=jsd, processors=8, outputdir=./); nmds(phylip=$DIST, maxdim=2)\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we build the NMDS ordination figures and output them as PDFs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "bodysites <- c(\"Antecubital_fossa\", \"Anterior_nares\", \"Buccal_mucosa\", \"Hard_palate\", \"Keratinized_gingiva\",\n",
    "               \"Palatine_Tonsils\", \"Retroauricular_crease\", \"Saliva\", \"Stool\", \"Subgingival_plaque\",\n",
    "               \"Supragingival_plaque\", \"Throat\", \"Tongue_dorsum\", \"Vagina\")\n",
    "\n",
    "\n",
    "for(bs in bodysites){\n",
    "    axes <- read.table(file=paste(bs, \".tx.1.subsample.jsd.1.lt.nmds.axes\", sep=\"\"), header=T, row.names=1)\n",
    "    axes <- axes[order(rownames(axes)),]\n",
    "    \n",
    "    stress <- read.table(file=paste(bs, \".tx.1.subsample.jsd.1.lt.nmds.stress\", sep=\"\"), header=T)\n",
    "    iter <- which.min(stress$Stress)\n",
    "    stress.score <- format(stress[iter, \"Stress\"], digits=2)\n",
    "    r.squared <- format(stress[iter, \"Rsq\"], digits=2)\n",
    "    \n",
    "    pdf(file=paste(bs,\".ordination.pdf\", sep=\"\"), width=8.5, height=11)\n",
    "    \n",
    "    par(mfrow=c(2,1))\n",
    "    par(mar=c(1, 4, 3, 1))\n",
    "    dmm <- read.table(file=paste(\"../v35/dmm_best/\", bs, \".tx.1.subsample.1.dmm.mix.design\", sep=\"\"), row.names=1)\n",
    "    rownames(dmm)==rownames(axes)\n",
    "\n",
    "    ntypes <- length(levels(factor(dmm$V2)))\n",
    "    clrs <- rainbow(ntypes)\n",
    "    \n",
    "    plot(axes, col=clrs[dmm$V2], pch=19, cex=0.75, xaxt=\"n\", xlab=\"\", ylab=\"NMDS Axis 2\", ylim=c(-0.7,0.7), xlim=c(-0.7,0.7))\n",
    "    text(x=-0.75, y=0.68, \"DMM\", cex=1.5, font=2, pos=4)\n",
    "    legend(\"bottomleft\", legend=paste(\"Community Type\", LETTERS[1:ntypes]),\n",
    "           pch=19, col=clrs, pt.cex=1.5)\n",
    "    mtext(side=3, line=1.25, text=gsub(\"_\", \" \", bs), at=0.7, adj=1, cex=1.5, font=2)\n",
    "    \n",
    "    \n",
    "    par(mar=c(4, 4, 0, 1))\n",
    "    pam <- read.table(file=paste(bs, \".tx.1.subsample.pam.sw.enterotype\", sep=\"\"), row.names=1)\n",
    "    rownames(pam)==rownames(axes)\n",
    "    ntypes <- length(levels(factor(pam$V2)))\n",
    "    clrs <- rainbow(ntypes)\n",
    "    \n",
    "    plot(axes, col=clrs[pam$V2], pch=19, cex=0.75, ylab=\"NMDS Axis 2\", xlab=\"NMDS Axis 1\", ylim=c(-0.7,0.7), xlim=c(-0.7,0.7))\n",
    "    text(x=-0.75, y=0.68, \"PAM (SI)\", cex=1.5, font=2, pos=4)\n",
    "    legend(\"bottomleft\", legend=paste(\"Community Type\", LETTERS[1:ntypes]), pch=19, col=rainbow(ntypes), pt.cex=1.5)\n",
    "\n",
    "    mtext(side=1, line=2, at=-0.9, text=paste(\"Stress: \", stress.score, \"\\nRsquared: \", r.squared, sep=\"\"), adj=0)\n",
    "    dev.off()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Conclusions\n",
    "We hope you've found this notebook useful at providing greater details into the methods we used to characterize the full HMP 16S rRNA gene sequencing dataset. Here are the take home points from the study and our methods development / application:\n",
    "\n",
    " * Identifying community types is a useful way of categorizing community data at a bodysite and can be used as a tool to reduce the complexity of the data in order to associate these types with clinical metadata (e.g. whether one was breastfed, gender, level of education, etc.)\n",
    " * Your mouth is connected to your rectum. You wouldn't believe the number of people (including one reviewer!) that were shocked to hear that the community type in your mouth is predictive of those in feces. Again, it is important to note that the bacterial taxa themselves are not necessarily the same, but the type of community you have in one site is predictive of others.\n",
    " * Community types are not fixed and each type appears to have its own level of variation. Unfortunately, we're somewhat powerless to figure out why they change. Other studies suggest the same thing. For example [Wu et al. 2012]() found two community types (using PAM) since they were interested in diet and couldn't find a short term result concluded that it was based on long term diet changes. We just don't know.\n",
    " * We appreciate that the concept of community types has been controversial. We have demonstrated that applying the DMM-based approach is more robust than the previously-applied PAM-based approaches, which do not seem robust. This is especially the case when there is only one community type since it will not be detected by the current PAM methods.  \n",
    "\n",
    "\n",
    "Some things for the future:  \n",
    "\n",
    " * If we can bin people into community types we can begin to associate community types with cases of disease or risk of disease. This has already been done in a number of studies (see the paper for references).  \n",
    " * We need better/more temporal metadata that we can use to explain changes in community types  \n",
    " * We suspect that as we get more and more subjects we will see more and more community types. The types represent clusters of points. We know that each individual is more like themselves than anyone else. Therefore, if we sample more points from each person, we would expect them to form their own type.  \n",
    " \n",
    "\n",
    "Finally, if you find this set of methods useful, please let us know (pschloss@umich.edu)!\n",
    "\n",
    "\n",
    "# Postscript\n",
    "The data analysis that was performed in the original study used a phylotype-based approach to characterize the structure of the human microbiome. We have also been interested in using the data to generate OTUs and mothur-based shared files. We did this as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mothur \"#cluster.split(fasta=data/raw/v35.unique.good.filter.unique.precluster.unique.pick.pick.fasta, name=data/raw/v35.unique.good.filter.unique.precluster.unique.pick.pick.names, taxonomy=data/raw/v35.unique.good.filter.unique.precluster.unique.pick.three.wang.pick.taxonomy, taxlevel=5, classic=T, processors=4, cluster=T, outputdir=data/mothur/);\n",
    "        make.shared(list=data/mothur/v35.unique.good.filter.unique.precluster.unique.pick.pick.an.list, group=data/raw/v35.good.pick.pick.groups, label=0.03, outputdir=mothur);\n",
    "        classify.otu(list=data/mothur/v35.unique.good.filter.unique.precluster.unique.pick.pick.an.list, group=data/raw/v35.good.pick.pick.groups, name=data/raw/v35.unique.good.filter.unique.precluster.unique.pick.pick.names, taxonomy=data/raw/v35.unique.good.filter.unique.precluster.unique.pick.three.wang.pick.taxonomy, label=0.03, outputdir=data/mothur);\n",
    "        get.oturep(fasta=data/raw/v35.unique.good.filter.unique.precluster.unique.pick.pick.fasta, name=data/raw/v35.unique.good.filter.unique.precluster.unique.pick.pick.names, list=data/mothur/v35.unique.good.filter.unique.precluster.unique.pick.pick.an.list, label=0.03, method=abundance);\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We then went back and ran these output files through the various scripts that we did above starting at the block staring \"Recall above we gave the samples...\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
